// amis-core/esm/index.js
/**
 * amis-core v2.9.0
 * Copyright 2018-2023 fex
 */

import { __assign } from 'tslib';
import { stores, defaultOptions } from './factory.js';
export { Renderer, addSchemaFilter, clearStoresCache, extendDefaultEnv, filterSchema, getRendererByName, getRenderers, registerRenderer, resolveRenderer, unRegisterRenderer, updateEnv } from './factory.js';
import './renderers/builtin.js';
import './renderers/register.js';
import { wrapFetcher } from './utils/api.js';
export { buildApi, clearApiCache, getApiCache, isApiOutdated, isApiOutdatedWithData, isEffectiveApi, isSameApi, isValidApi, jsFetcher, jsonpFetcher, normalizeApi, normalizeApiResponseData, responseAdaptor, setApiCache, str2AsyncFunction, str2function, wrapAdaptor, wrapFetcher } from './utils/api.js';
export { attachmentAdpator } from './utils/attachmentAdpator.js';
export { autobindMethod, bind } from './utils/autobind.js';
export { ColorScale } from './utils/ColorScale.js';
export { columnsSplit } from './utils/columnsSplit.js';
export { dataMapping, resolveMapping, resolveMappingObject } from './utils/dataMapping.js';
export { DataSchema } from './utils/DataSchema.js';
export { DataScope } from './utils/DataScope.js';
export { filterDate, parseDuration, relativeValueRe } from './utils/date.js';
import { enableDebug } from './utils/debug.js';
export { DebugWrapper, debug, enableDebug, warning } from './utils/debug.js';
export { calculatePosition, getContainer, getStyleNumber, ownerDocument } from './utils/dom.js';
export { ServerError } from './utils/errors.js';
export { escapeHtml } from './utils/escapeHtml.js';
export { getExprProperties } from './utils/filter-schema.js';
import './utils/filter.js';
export { formatDuration } from './utils/formatDuration.js';
export { FormulaExec, formulaExec, isExpression, isNeedFormula, isNowFormula, registerFormulaExec, replaceExpression } from './utils/formula.js';
export { getVariable } from './utils/getVariable.js';
export { collectVariables } from './utils/grammar.js';
export { handleAction } from './utils/handleAction.js';
import { promisify } from './utils/helper.js';
import * as helper from './utils/helper.js';
export { helper as utils };
export { JSONTraverse, SkipOperation, __uri, anyChanged, autobind, bulkBindFunctions, camel, chainEvents, chainFunctions, convertArrayValueToMoment, detectPropValueChanged, difference, eachTree, everyTree, filterTree, findIndex, findObjectsWithKey, findTree, findTreeIndex, flattenTree, flattenTreeWithLeafNodes, getLevelFromClassName, getPropValue, getRange, getScrollParent, getScrollbarWidth, getTree, getTreeAncestors, getTreeDepth, getTreeParent, getWidthRate, guid, hasAbility, hasFile, hasOwnPropertyInPath, hasVisibleExpression, hashCode, immutableExtends, injectPropsToObject, isArrayChildrenModified, isBreakpoint, isClickOnInput, isDisabled, isEmpty, isMobile, isNumeric, isObjectShallowModified, isSuperDataModified, isUnfolded, isVisible, lcFirst, loadScript, makeColumnClassBuild, makeHorizontalDeeper, mapObject, mapTree, noop, normalizeNodePath, object2formData, omitControls, padArr, parseQuery, pickEventsProps, preventDefault, promisify, qsparse, qsstringify, range, removeHTMLTag, repeatCount, rmUndefined, someTree, sortArray, spliceTree, syncDataFromSuper, ucFirst, until, uuid, uuidv4, visibilityFilter } from './utils/helper.js';
export { highlight } from './utils/highlight.js';
export { generateIcon } from './utils/icon.js';
export { getImageDimensions, toDataURL } from './utils/image.js';
export { isPureVariable } from './utils/isPureVariable.js';
export { JSONSchme2AMisSchema } from './utils/json-schema-2-amis-schema.js';
export { keyToPath } from './utils/keyToPath.js';
export { makeSorter } from './utils/makeSorter.js';
export { normalizeLink } from './utils/normalizeLink.js';
export { normalizeOptions } from './utils/normalizeOptions.js';
export { cloneObject, createObject, deleteVariable, extendObject, isObject, pickValues, setVariable } from './utils/object.js';
export { offset } from './utils/offset.js';
export { offsetParent } from './utils/offsetParent.js';
export { matchOptionValue, optionValueCompare } from './utils/optionValueCompare.js';
export { position } from './utils/position.js';
export { prettyBytes } from './utils/prettyBytes.js';
export { bindEvent, createRendererEvent, dispatchEvent, getRendererEventListeners, resolveEventData } from './utils/renderer-event.js';
import { replaceText } from './utils/replaceText.js';
export { replaceText } from './utils/replaceText.js';
export { getComputedStyle, resizeSensor } from './utils/resize-sensor.js';
export { resolveVariable } from './utils/resolveVariable.js';
export { resolveVariableAndFilter } from './utils/resolveVariableAndFilter.js';
export { resolveVariableAndFilterForAsync } from './utils/resolveVariableAndFilterForAsync.js';
export { RootClose } from './utils/RootClose.js';
export { scrollPosition } from './utils/scrollPosition.js';
export { SimpleMap } from './utils/SimpleMap.js';
export { string2regExp } from './utils/string2regExp.js';
export { buildStyle } from './utils/style.js';
export { tokenize } from './utils/tokenize.js';
export { register } from './utils/tpl-builtin.js';
export { evalExpression, evalExpressionWithConditionBuilder, evalJS, filter, registerTplEnginer, setCustomEvalExpression, setCustomEvalJs } from './utils/tpl.js';
export { uncontrollable } from './utils/uncontrollable.js';
export { addRule, str2rules, validate, validateMessages, validateObject, validations } from './utils/validations.js';
export { toNumber } from './utils/toNumber.js';
export { decodeEntity } from './utils/decodeEntity.js';
export { addStyle, findOrCreactStyle, formatStyle, insertCustomStyle, insertStyle, valueMap } from './utils/style-helper.js';
export { getConditionComputers, registerConditionComputer, resolveCondition } from './utils/resolveCondition.js';
export { default as animation } from './utils/Animation.js';
import { RendererStore } from './store/index.js';
export { RegisterStore, RendererStore } from './store/index.js';
import { getEnv } from 'mobx-state-tree';
import { getDefaultLocale, makeTranslator } from './locale.js';
export { extendLocale, getDefaultLocale, localeable, makeTranslator, register as registerLocale, removeLocaleData, setDefaultLocale } from './locale.js';
export { HocScoped as Scoped, ScopedContext } from './Scoped.js';
import { getClassPrefix, getTheme } from './theme.js';
export { classnames, getClassPrefix, getTheme, makeClassnames, setDefaultTheme, theme, themeable } from './theme.js';
import './actions/LoopAction.js';
import './actions/BreakAction.js';
import './actions/ContinueAction.js';
import './actions/SwitchAction.js';
import './actions/ParallelAction.js';
import './actions/CustomAction.js';
import './actions/BroadcastAction.js';
import './actions/CmptAction.js';
import './actions/AjaxAction.js';
import './actions/CopyAction.js';
import './actions/DialogAction.js';
import './actions/DrawerAction.js';
import './actions/EmailAction.js';
import './actions/LinkAction.js';
import './actions/ToastAction.js';
import './actions/PageAction.js';
export { LoopStatus, getActionByType, registerAction, runAction, runActions } from './actions/Action.js';
export { FormItem, FormItemWrap, getFormItemByName, registerFormItem } from './renderers/Item.js';
export { OptionsControl, registerOptionsControl } from './renderers/Options.js';
import ScopedRootRenderer from './Root.js';
export { addRootWrapper } from './Root.js';
import { envOverwrite } from './envOverwrite.js';
import { EnvContext } from './env.js';
export { EnvContext } from './env.js';
import React from 'react';
export { AsyncEvaluator, Evaluator, evaluate, evaluateForAsync, extendsFilters, filters, getFilters, lexer, parse, registerFilter, registerFunction } from 'amis-formula';
export { default as LazyComponent } from './components/LazyComponent.js';
export { default as Overlay } from './components/Overlay.js';
export { default as PopOver } from './components/PopOver.js';
export { FormRenderer } from './renderers/Form.js';
export { stripNumber } from './utils/stripNumber.js';
export { iRendererStore } from './store/iRenderer.js';
export { ServiceStore } from './store/service.js';
export { FormStore } from './store/form.js';
export { ComboStore } from './store/combo.js';
export { CRUDStore } from './store/crud.js';
export { TableStore } from './store/table.js';
export { TableStore2 } from './store/table2.js';
export { ListStore } from './store/list.js';
export { ModalStore } from './store/modal.js';
export { FormItemStore } from './store/formItem.js';
export { PaginationStore } from './store/pagination.js';
export { AppStore } from './store/app.js';
export { StoreNode } from './store/node.js';

/** @license amis v@version
 *
 * Copyright Baidu
 *
 * This source code is licensed under the Apache license found in the
 * LICENSE file in the root directory of this source tree.
 */
var classPrefix = getClassPrefix();
// @ts-ignore
var version = '2.9.0';
function render(schema, props, options, pathPrefix) {
    if (props === void 0) { props = {}; }
    if (options === void 0) { options = {}; }
    if (pathPrefix === void 0) { pathPrefix = ''; }
    var locale = props.locale || getDefaultLocale();
    // 兼容 locale 的不同写法
    locale = locale.replace('_', '-');
    locale = locale === 'en' ? 'en-US' : locale;
    locale = locale === 'zh' ? 'zh-CN' : locale;
    locale = locale === 'cn' ? 'zh-CN' : locale;
    var translate = props.translate || makeTranslator(locale);
    var store = stores[options.session || 'global'];
    // 根据环境覆盖 schema，这个要在最前面做，不然就无法覆盖 validations
    envOverwrite(schema, locale);
    if (!store) {
        options = __assign(__assign(__assign({}, defaultOptions), options), { fetcher: options.fetcher
                ? wrapFetcher(options.fetcher, options.tracker)
                : defaultOptions.fetcher, confirm: promisify(options.confirm || defaultOptions.confirm || window.confirm), locale: locale, translate: translate });
        if (options.enableAMISDebug) {
            // 因为里面还有 render
            setTimeout(function () {
                enableDebug();
            }, 10);
        }
        store = RendererStore.create({}, options);
        stores[options.session || 'global'] = store;
    }
    window.amisStore = store; // 为了方便 debug.
    var env = getEnv(store);
    var theme = props.theme || options.theme || 'cxd';
    if (theme === 'default') {
        theme = 'cxd';
    }
    env.theme = getTheme(theme);
    if (props.locale !== undefined) {
        env.translate = translate;
        env.locale = locale;
    }
    // 默认将开启移动端原生 UI
    if (options.useMobileUI !== false) {
        props.useMobileUI = true;
    }
    schema = replaceText(schema, options.replaceText, env.replaceTextIgnoreKeys);
    return (React.createElement(EnvContext.Provider, { value: env },
        React.createElement(ScopedRootRenderer, __assign({}, props, { schema: schema, pathPrefix: pathPrefix, rootStore: store, env: env, theme: theme, locale: locale, translate: translate }))));
}

export { classPrefix, render, version };
