// office-viewer/lib/openxml/Types.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

exports.ST_CalendarType = void 0;
(function (ST_CalendarType) {
    ST_CalendarType["gregorian"] = "gregorian";
    ST_CalendarType["gregorianUs"] = "gregorianUs";
    ST_CalendarType["gregorianMeFrench"] = "gregorianMeFrench";
    ST_CalendarType["gregorianArabic"] = "gregorianArabic";
    ST_CalendarType["hijri"] = "hijri";
    ST_CalendarType["hebrew"] = "hebrew";
    ST_CalendarType["taiwan"] = "taiwan";
    ST_CalendarType["japan"] = "japan";
    ST_CalendarType["thai"] = "thai";
    ST_CalendarType["korea"] = "korea";
    ST_CalendarType["saka"] = "saka";
    ST_CalendarType["gregorianXlitEnglish"] = "gregorianXlitEnglish";
    ST_CalendarType["gregorianXlitFrench"] = "gregorianXlitFrench";
    ST_CalendarType["none"] = "none";
})(exports.ST_CalendarType || (exports.ST_CalendarType = {}));
exports.ST_VerticalAlignRun = void 0;
(function (ST_VerticalAlignRun) {
    ST_VerticalAlignRun["baseline"] = "baseline";
    ST_VerticalAlignRun["superscript"] = "superscript";
    ST_VerticalAlignRun["subscript"] = "subscript";
})(exports.ST_VerticalAlignRun || (exports.ST_VerticalAlignRun = {}));
exports.ST_XAlign = void 0;
(function (ST_XAlign) {
    ST_XAlign["left"] = "left";
    ST_XAlign["center"] = "center";
    ST_XAlign["right"] = "right";
    ST_XAlign["inside"] = "inside";
    ST_XAlign["outside"] = "outside";
})(exports.ST_XAlign || (exports.ST_XAlign = {}));
exports.ST_YAlign = void 0;
(function (ST_YAlign) {
    ST_YAlign["inline"] = "inline";
    ST_YAlign["top"] = "top";
    ST_YAlign["center"] = "center";
    ST_YAlign["bottom"] = "bottom";
    ST_YAlign["inside"] = "inside";
    ST_YAlign["outside"] = "outside";
})(exports.ST_YAlign || (exports.ST_YAlign = {}));
exports.ST_ConformanceClass = void 0;
(function (ST_ConformanceClass) {
    ST_ConformanceClass["strict"] = "strict";
    ST_ConformanceClass["transitional"] = "transitional";
})(exports.ST_ConformanceClass || (exports.ST_ConformanceClass = {}));
exports.ST_HighlightColor = void 0;
(function (ST_HighlightColor) {
    ST_HighlightColor["black"] = "black";
    ST_HighlightColor["blue"] = "blue";
    ST_HighlightColor["cyan"] = "cyan";
    ST_HighlightColor["green"] = "green";
    ST_HighlightColor["magenta"] = "magenta";
    ST_HighlightColor["red"] = "red";
    ST_HighlightColor["yellow"] = "yellow";
    ST_HighlightColor["white"] = "white";
    ST_HighlightColor["darkBlue"] = "darkBlue";
    ST_HighlightColor["darkCyan"] = "darkCyan";
    ST_HighlightColor["darkGreen"] = "darkGreen";
    ST_HighlightColor["darkMagenta"] = "darkMagenta";
    ST_HighlightColor["darkRed"] = "darkRed";
    ST_HighlightColor["darkYellow"] = "darkYellow";
    ST_HighlightColor["darkGray"] = "darkGray";
    ST_HighlightColor["lightGray"] = "lightGray";
    ST_HighlightColor["none"] = "none";
})(exports.ST_HighlightColor || (exports.ST_HighlightColor = {}));
exports.ST_HexColorAuto = void 0;
(function (ST_HexColorAuto) {
    ST_HexColorAuto["auto"] = "auto";
})(exports.ST_HexColorAuto || (exports.ST_HexColorAuto = {}));
exports.ST_Underline = void 0;
(function (ST_Underline) {
    ST_Underline["single"] = "single";
    ST_Underline["words"] = "words";
    ST_Underline["double"] = "double";
    ST_Underline["thick"] = "thick";
    ST_Underline["dotted"] = "dotted";
    ST_Underline["dottedHeavy"] = "dottedHeavy";
    ST_Underline["dash"] = "dash";
    ST_Underline["dashedHeavy"] = "dashedHeavy";
    ST_Underline["dashLong"] = "dashLong";
    ST_Underline["dashLongHeavy"] = "dashLongHeavy";
    ST_Underline["dotDash"] = "dotDash";
    ST_Underline["dashDotHeavy"] = "dashDotHeavy";
    ST_Underline["dotDotDash"] = "dotDotDash";
    ST_Underline["dashDotDotHeavy"] = "dashDotDotHeavy";
    ST_Underline["wave"] = "wave";
    ST_Underline["wavyHeavy"] = "wavyHeavy";
    ST_Underline["wavyDouble"] = "wavyDouble";
    ST_Underline["none"] = "none";
})(exports.ST_Underline || (exports.ST_Underline = {}));
exports.ST_TextEffect = void 0;
(function (ST_TextEffect) {
    ST_TextEffect["blinkBackground"] = "blinkBackground";
    ST_TextEffect["lights"] = "lights";
    ST_TextEffect["antsBlack"] = "antsBlack";
    ST_TextEffect["antsRed"] = "antsRed";
    ST_TextEffect["shimmer"] = "shimmer";
    ST_TextEffect["sparkle"] = "sparkle";
    ST_TextEffect["none"] = "none";
})(exports.ST_TextEffect || (exports.ST_TextEffect = {}));
exports.ST_Border = void 0;
(function (ST_Border) {
    ST_Border["nil"] = "nil";
    ST_Border["none"] = "none";
    ST_Border["single"] = "single";
    ST_Border["thick"] = "thick";
    ST_Border["double"] = "double";
    ST_Border["dotted"] = "dotted";
    ST_Border["dashed"] = "dashed";
    ST_Border["dotDash"] = "dotDash";
    ST_Border["dotDotDash"] = "dotDotDash";
    ST_Border["triple"] = "triple";
    ST_Border["thinThickSmallGap"] = "thinThickSmallGap";
    ST_Border["thickThinSmallGap"] = "thickThinSmallGap";
    ST_Border["thinThickThinSmallGap"] = "thinThickThinSmallGap";
    ST_Border["thinThickMediumGap"] = "thinThickMediumGap";
    ST_Border["thickThinMediumGap"] = "thickThinMediumGap";
    ST_Border["thinThickThinMediumGap"] = "thinThickThinMediumGap";
    ST_Border["thinThickLargeGap"] = "thinThickLargeGap";
    ST_Border["thickThinLargeGap"] = "thickThinLargeGap";
    ST_Border["thinThickThinLargeGap"] = "thinThickThinLargeGap";
    ST_Border["wave"] = "wave";
    ST_Border["doubleWave"] = "doubleWave";
    ST_Border["dashSmallGap"] = "dashSmallGap";
    ST_Border["dashDotStroked"] = "dashDotStroked";
    ST_Border["threeDEmboss"] = "threeDEmboss";
    ST_Border["threeDEngrave"] = "threeDEngrave";
    ST_Border["outset"] = "outset";
    ST_Border["inset"] = "inset";
    ST_Border["apples"] = "apples";
    ST_Border["archedScallops"] = "archedScallops";
    ST_Border["babyPacifier"] = "babyPacifier";
    ST_Border["babyRattle"] = "babyRattle";
    ST_Border["balloons3Colors"] = "balloons3Colors";
    ST_Border["balloonsHotAir"] = "balloonsHotAir";
    ST_Border["basicBlackDashes"] = "basicBlackDashes";
    ST_Border["basicBlackDots"] = "basicBlackDots";
    ST_Border["basicBlackSquares"] = "basicBlackSquares";
    ST_Border["basicThinLines"] = "basicThinLines";
    ST_Border["basicWhiteDashes"] = "basicWhiteDashes";
    ST_Border["basicWhiteDots"] = "basicWhiteDots";
    ST_Border["basicWhiteSquares"] = "basicWhiteSquares";
    ST_Border["basicWideInline"] = "basicWideInline";
    ST_Border["basicWideMidline"] = "basicWideMidline";
    ST_Border["basicWideOutline"] = "basicWideOutline";
    ST_Border["bats"] = "bats";
    ST_Border["birds"] = "birds";
    ST_Border["birdsFlight"] = "birdsFlight";
    ST_Border["cabins"] = "cabins";
    ST_Border["cakeSlice"] = "cakeSlice";
    ST_Border["candyCorn"] = "candyCorn";
    ST_Border["celticKnotwork"] = "celticKnotwork";
    ST_Border["certificateBanner"] = "certificateBanner";
    ST_Border["chainLink"] = "chainLink";
    ST_Border["champagneBottle"] = "champagneBottle";
    ST_Border["checkedBarBlack"] = "checkedBarBlack";
    ST_Border["checkedBarColor"] = "checkedBarColor";
    ST_Border["checkered"] = "checkered";
    ST_Border["christmasTree"] = "christmasTree";
    ST_Border["circlesLines"] = "circlesLines";
    ST_Border["circlesRectangles"] = "circlesRectangles";
    ST_Border["classicalWave"] = "classicalWave";
    ST_Border["clocks"] = "clocks";
    ST_Border["compass"] = "compass";
    ST_Border["confetti"] = "confetti";
    ST_Border["confettiGrays"] = "confettiGrays";
    ST_Border["confettiOutline"] = "confettiOutline";
    ST_Border["confettiStreamers"] = "confettiStreamers";
    ST_Border["confettiWhite"] = "confettiWhite";
    ST_Border["cornerTriangles"] = "cornerTriangles";
    ST_Border["couponCutoutDashes"] = "couponCutoutDashes";
    ST_Border["couponCutoutDots"] = "couponCutoutDots";
    ST_Border["crazyMaze"] = "crazyMaze";
    ST_Border["creaturesButterfly"] = "creaturesButterfly";
    ST_Border["creaturesFish"] = "creaturesFish";
    ST_Border["creaturesInsects"] = "creaturesInsects";
    ST_Border["creaturesLadyBug"] = "creaturesLadyBug";
    ST_Border["crossStitch"] = "crossStitch";
    ST_Border["cup"] = "cup";
    ST_Border["decoArch"] = "decoArch";
    ST_Border["decoArchColor"] = "decoArchColor";
    ST_Border["decoBlocks"] = "decoBlocks";
    ST_Border["diamondsGray"] = "diamondsGray";
    ST_Border["doubleD"] = "doubleD";
    ST_Border["doubleDiamonds"] = "doubleDiamonds";
    ST_Border["earth1"] = "earth1";
    ST_Border["earth2"] = "earth2";
    ST_Border["earth3"] = "earth3";
    ST_Border["eclipsingSquares1"] = "eclipsingSquares1";
    ST_Border["eclipsingSquares2"] = "eclipsingSquares2";
    ST_Border["eggsBlack"] = "eggsBlack";
    ST_Border["fans"] = "fans";
    ST_Border["film"] = "film";
    ST_Border["firecrackers"] = "firecrackers";
    ST_Border["flowersBlockPrint"] = "flowersBlockPrint";
    ST_Border["flowersDaisies"] = "flowersDaisies";
    ST_Border["flowersModern1"] = "flowersModern1";
    ST_Border["flowersModern2"] = "flowersModern2";
    ST_Border["flowersPansy"] = "flowersPansy";
    ST_Border["flowersRedRose"] = "flowersRedRose";
    ST_Border["flowersRoses"] = "flowersRoses";
    ST_Border["flowersTeacup"] = "flowersTeacup";
    ST_Border["flowersTiny"] = "flowersTiny";
    ST_Border["gems"] = "gems";
    ST_Border["gingerbreadMan"] = "gingerbreadMan";
    ST_Border["gradient"] = "gradient";
    ST_Border["handmade1"] = "handmade1";
    ST_Border["handmade2"] = "handmade2";
    ST_Border["heartBalloon"] = "heartBalloon";
    ST_Border["heartGray"] = "heartGray";
    ST_Border["hearts"] = "hearts";
    ST_Border["heebieJeebies"] = "heebieJeebies";
    ST_Border["holly"] = "holly";
    ST_Border["houseFunky"] = "houseFunky";
    ST_Border["hypnotic"] = "hypnotic";
    ST_Border["iceCreamCones"] = "iceCreamCones";
    ST_Border["lightBulb"] = "lightBulb";
    ST_Border["lightning1"] = "lightning1";
    ST_Border["lightning2"] = "lightning2";
    ST_Border["mapPins"] = "mapPins";
    ST_Border["mapleLeaf"] = "mapleLeaf";
    ST_Border["mapleMuffins"] = "mapleMuffins";
    ST_Border["marquee"] = "marquee";
    ST_Border["marqueeToothed"] = "marqueeToothed";
    ST_Border["moons"] = "moons";
    ST_Border["mosaic"] = "mosaic";
    ST_Border["musicNotes"] = "musicNotes";
    ST_Border["northwest"] = "northwest";
    ST_Border["ovals"] = "ovals";
    ST_Border["packages"] = "packages";
    ST_Border["palmsBlack"] = "palmsBlack";
    ST_Border["palmsColor"] = "palmsColor";
    ST_Border["paperClips"] = "paperClips";
    ST_Border["papyrus"] = "papyrus";
    ST_Border["partyFavor"] = "partyFavor";
    ST_Border["partyGlass"] = "partyGlass";
    ST_Border["pencils"] = "pencils";
    ST_Border["people"] = "people";
    ST_Border["peopleWaving"] = "peopleWaving";
    ST_Border["peopleHats"] = "peopleHats";
    ST_Border["poinsettias"] = "poinsettias";
    ST_Border["postageStamp"] = "postageStamp";
    ST_Border["pumpkin1"] = "pumpkin1";
    ST_Border["pushPinNote2"] = "pushPinNote2";
    ST_Border["pushPinNote1"] = "pushPinNote1";
    ST_Border["pyramids"] = "pyramids";
    ST_Border["pyramidsAbove"] = "pyramidsAbove";
    ST_Border["quadrants"] = "quadrants";
    ST_Border["rings"] = "rings";
    ST_Border["safari"] = "safari";
    ST_Border["sawtooth"] = "sawtooth";
    ST_Border["sawtoothGray"] = "sawtoothGray";
    ST_Border["scaredCat"] = "scaredCat";
    ST_Border["seattle"] = "seattle";
    ST_Border["shadowedSquares"] = "shadowedSquares";
    ST_Border["sharksTeeth"] = "sharksTeeth";
    ST_Border["shorebirdTracks"] = "shorebirdTracks";
    ST_Border["skyrocket"] = "skyrocket";
    ST_Border["snowflakeFancy"] = "snowflakeFancy";
    ST_Border["snowflakes"] = "snowflakes";
    ST_Border["sombrero"] = "sombrero";
    ST_Border["southwest"] = "southwest";
    ST_Border["stars"] = "stars";
    ST_Border["starsTop"] = "starsTop";
    ST_Border["stars3d"] = "stars3d";
    ST_Border["starsBlack"] = "starsBlack";
    ST_Border["starsShadowed"] = "starsShadowed";
    ST_Border["sun"] = "sun";
    ST_Border["swirligig"] = "swirligig";
    ST_Border["tornPaper"] = "tornPaper";
    ST_Border["tornPaperBlack"] = "tornPaperBlack";
    ST_Border["trees"] = "trees";
    ST_Border["triangleParty"] = "triangleParty";
    ST_Border["triangles"] = "triangles";
    ST_Border["triangle1"] = "triangle1";
    ST_Border["triangle2"] = "triangle2";
    ST_Border["triangleCircle1"] = "triangleCircle1";
    ST_Border["triangleCircle2"] = "triangleCircle2";
    ST_Border["shapes1"] = "shapes1";
    ST_Border["shapes2"] = "shapes2";
    ST_Border["twistedLines1"] = "twistedLines1";
    ST_Border["twistedLines2"] = "twistedLines2";
    ST_Border["vine"] = "vine";
    ST_Border["waveline"] = "waveline";
    ST_Border["weavingAngles"] = "weavingAngles";
    ST_Border["weavingBraid"] = "weavingBraid";
    ST_Border["weavingRibbon"] = "weavingRibbon";
    ST_Border["weavingStrips"] = "weavingStrips";
    ST_Border["whiteFlowers"] = "whiteFlowers";
    ST_Border["woodwork"] = "woodwork";
    ST_Border["xIllusions"] = "xIllusions";
    ST_Border["zanyTriangles"] = "zanyTriangles";
    ST_Border["zigZag"] = "zigZag";
    ST_Border["zigZagStitch"] = "zigZagStitch";
    ST_Border["custom"] = "custom";
})(exports.ST_Border || (exports.ST_Border = {}));
exports.ST_Shd = void 0;
(function (ST_Shd) {
    ST_Shd["nil"] = "nil";
    ST_Shd["clear"] = "clear";
    ST_Shd["solid"] = "solid";
    ST_Shd["horzStripe"] = "horzStripe";
    ST_Shd["vertStripe"] = "vertStripe";
    ST_Shd["reverseDiagStripe"] = "reverseDiagStripe";
    ST_Shd["diagStripe"] = "diagStripe";
    ST_Shd["horzCross"] = "horzCross";
    ST_Shd["diagCross"] = "diagCross";
    ST_Shd["thinHorzStripe"] = "thinHorzStripe";
    ST_Shd["thinVertStripe"] = "thinVertStripe";
    ST_Shd["thinReverseDiagStripe"] = "thinReverseDiagStripe";
    ST_Shd["thinDiagStripe"] = "thinDiagStripe";
    ST_Shd["thinHorzCross"] = "thinHorzCross";
    ST_Shd["thinDiagCross"] = "thinDiagCross";
    ST_Shd["pct5"] = "pct5";
    ST_Shd["pct10"] = "pct10";
    ST_Shd["pct12"] = "pct12";
    ST_Shd["pct15"] = "pct15";
    ST_Shd["pct20"] = "pct20";
    ST_Shd["pct25"] = "pct25";
    ST_Shd["pct30"] = "pct30";
    ST_Shd["pct35"] = "pct35";
    ST_Shd["pct37"] = "pct37";
    ST_Shd["pct40"] = "pct40";
    ST_Shd["pct45"] = "pct45";
    ST_Shd["pct50"] = "pct50";
    ST_Shd["pct55"] = "pct55";
    ST_Shd["pct60"] = "pct60";
    ST_Shd["pct62"] = "pct62";
    ST_Shd["pct65"] = "pct65";
    ST_Shd["pct70"] = "pct70";
    ST_Shd["pct75"] = "pct75";
    ST_Shd["pct80"] = "pct80";
    ST_Shd["pct85"] = "pct85";
    ST_Shd["pct87"] = "pct87";
    ST_Shd["pct90"] = "pct90";
    ST_Shd["pct95"] = "pct95";
})(exports.ST_Shd || (exports.ST_Shd = {}));
exports.ST_Em = void 0;
(function (ST_Em) {
    ST_Em["none"] = "none";
    ST_Em["dot"] = "dot";
    ST_Em["comma"] = "comma";
    ST_Em["circle"] = "circle";
    ST_Em["underDot"] = "underDot";
})(exports.ST_Em || (exports.ST_Em = {}));
exports.ST_CombineBrackets = void 0;
(function (ST_CombineBrackets) {
    ST_CombineBrackets["none"] = "none";
    ST_CombineBrackets["round"] = "round";
    ST_CombineBrackets["square"] = "square";
    ST_CombineBrackets["angle"] = "angle";
    ST_CombineBrackets["curly"] = "curly";
})(exports.ST_CombineBrackets || (exports.ST_CombineBrackets = {}));
exports.ST_HeightRule = void 0;
(function (ST_HeightRule) {
    ST_HeightRule["auto"] = "auto";
    ST_HeightRule["exact"] = "exact";
    ST_HeightRule["atLeast"] = "atLeast";
})(exports.ST_HeightRule || (exports.ST_HeightRule = {}));
exports.ST_Wrap = void 0;
(function (ST_Wrap) {
    ST_Wrap["auto"] = "auto";
    ST_Wrap["notBeside"] = "notBeside";
    ST_Wrap["around"] = "around";
    ST_Wrap["tight"] = "tight";
    ST_Wrap["through"] = "through";
    ST_Wrap["none"] = "none";
})(exports.ST_Wrap || (exports.ST_Wrap = {}));
exports.ST_VAnchor = void 0;
(function (ST_VAnchor) {
    ST_VAnchor["text"] = "text";
    ST_VAnchor["margin"] = "margin";
    ST_VAnchor["page"] = "page";
})(exports.ST_VAnchor || (exports.ST_VAnchor = {}));
exports.ST_HAnchor = void 0;
(function (ST_HAnchor) {
    ST_HAnchor["text"] = "text";
    ST_HAnchor["margin"] = "margin";
    ST_HAnchor["page"] = "page";
})(exports.ST_HAnchor || (exports.ST_HAnchor = {}));
exports.ST_DropCap = void 0;
(function (ST_DropCap) {
    ST_DropCap["none"] = "none";
    ST_DropCap["drop"] = "drop";
    ST_DropCap["margin"] = "margin";
})(exports.ST_DropCap || (exports.ST_DropCap = {}));
exports.ST_TabJc = void 0;
(function (ST_TabJc) {
    ST_TabJc["clear"] = "clear";
    ST_TabJc["start"] = "start";
    ST_TabJc["center"] = "center";
    ST_TabJc["end"] = "end";
    ST_TabJc["decimal"] = "decimal";
    ST_TabJc["bar"] = "bar";
    ST_TabJc["num"] = "num";
})(exports.ST_TabJc || (exports.ST_TabJc = {}));
exports.ST_TabTlc = void 0;
(function (ST_TabTlc) {
    ST_TabTlc["none"] = "none";
    ST_TabTlc["dot"] = "dot";
    ST_TabTlc["hyphen"] = "hyphen";
    ST_TabTlc["underscore"] = "underscore";
    ST_TabTlc["heavy"] = "heavy";
    ST_TabTlc["middleDot"] = "middleDot";
})(exports.ST_TabTlc || (exports.ST_TabTlc = {}));
exports.ST_LineSpacingRule = void 0;
(function (ST_LineSpacingRule) {
    ST_LineSpacingRule["auto"] = "auto";
    ST_LineSpacingRule["exact"] = "exact";
    ST_LineSpacingRule["atLeast"] = "atLeast";
})(exports.ST_LineSpacingRule || (exports.ST_LineSpacingRule = {}));
exports.ST_Jc = void 0;
(function (ST_Jc) {
    ST_Jc["start"] = "start";
    ST_Jc["center"] = "center";
    ST_Jc["end"] = "end";
    ST_Jc["both"] = "both";
    ST_Jc["mediumKashida"] = "mediumKashida";
    ST_Jc["distribute"] = "distribute";
    ST_Jc["numTab"] = "numTab";
    ST_Jc["highKashida"] = "highKashida";
    ST_Jc["lowKashida"] = "lowKashida";
    ST_Jc["thaiDistribute"] = "thaiDistribute";
})(exports.ST_Jc || (exports.ST_Jc = {}));
exports.ST_JcTable = void 0;
(function (ST_JcTable) {
    ST_JcTable["center"] = "center";
    ST_JcTable["end"] = "end";
    ST_JcTable["start"] = "start";
})(exports.ST_JcTable || (exports.ST_JcTable = {}));
exports.ST_View = void 0;
(function (ST_View) {
    ST_View["none"] = "none";
    ST_View["print"] = "print";
    ST_View["outline"] = "outline";
    ST_View["masterPages"] = "masterPages";
    ST_View["normal"] = "normal";
    ST_View["web"] = "web";
})(exports.ST_View || (exports.ST_View = {}));
exports.ST_Zoom = void 0;
(function (ST_Zoom) {
    ST_Zoom["none"] = "none";
    ST_Zoom["fullPage"] = "fullPage";
    ST_Zoom["bestFit"] = "bestFit";
    ST_Zoom["textFit"] = "textFit";
})(exports.ST_Zoom || (exports.ST_Zoom = {}));
exports.ST_Proof = void 0;
(function (ST_Proof) {
    ST_Proof["clean"] = "clean";
    ST_Proof["dirty"] = "dirty";
})(exports.ST_Proof || (exports.ST_Proof = {}));
exports.ST_DocProtect = void 0;
(function (ST_DocProtect) {
    ST_DocProtect["none"] = "none";
    ST_DocProtect["readOnly"] = "readOnly";
    ST_DocProtect["comments"] = "comments";
    ST_DocProtect["trackedChanges"] = "trackedChanges";
    ST_DocProtect["forms"] = "forms";
})(exports.ST_DocProtect || (exports.ST_DocProtect = {}));
exports.ST_MailMergeDocType = void 0;
(function (ST_MailMergeDocType) {
    ST_MailMergeDocType["catalog"] = "catalog";
    ST_MailMergeDocType["envelopes"] = "envelopes";
    ST_MailMergeDocType["mailingLabels"] = "mailingLabels";
    ST_MailMergeDocType["formLetters"] = "formLetters";
    ST_MailMergeDocType["email"] = "email";
    ST_MailMergeDocType["fax"] = "fax";
})(exports.ST_MailMergeDocType || (exports.ST_MailMergeDocType = {}));
exports.ST_MailMergeDest = void 0;
(function (ST_MailMergeDest) {
    ST_MailMergeDest["newDocument"] = "newDocument";
    ST_MailMergeDest["printer"] = "printer";
    ST_MailMergeDest["email"] = "email";
    ST_MailMergeDest["fax"] = "fax";
})(exports.ST_MailMergeDest || (exports.ST_MailMergeDest = {}));
exports.ST_MailMergeOdsoFMDFieldType = void 0;
(function (ST_MailMergeOdsoFMDFieldType) {
    ST_MailMergeOdsoFMDFieldType["null"] = "null";
    ST_MailMergeOdsoFMDFieldType["dbColumn"] = "dbColumn";
})(exports.ST_MailMergeOdsoFMDFieldType || (exports.ST_MailMergeOdsoFMDFieldType = {}));
exports.ST_TextDirection = void 0;
(function (ST_TextDirection) {
    ST_TextDirection["tb"] = "tb";
    ST_TextDirection["rl"] = "rl";
    ST_TextDirection["lr"] = "lr";
    ST_TextDirection["tbV"] = "tbV";
    ST_TextDirection["rlV"] = "rlV";
    ST_TextDirection["lrV"] = "lrV";
})(exports.ST_TextDirection || (exports.ST_TextDirection = {}));
exports.ST_TextAlignment = void 0;
(function (ST_TextAlignment) {
    ST_TextAlignment["top"] = "top";
    ST_TextAlignment["center"] = "center";
    ST_TextAlignment["baseline"] = "baseline";
    ST_TextAlignment["bottom"] = "bottom";
    ST_TextAlignment["auto"] = "auto";
})(exports.ST_TextAlignment || (exports.ST_TextAlignment = {}));
exports.ST_DisplacedByCustomXml = void 0;
(function (ST_DisplacedByCustomXml) {
    ST_DisplacedByCustomXml["next"] = "next";
    ST_DisplacedByCustomXml["prev"] = "prev";
})(exports.ST_DisplacedByCustomXml || (exports.ST_DisplacedByCustomXml = {}));
exports.ST_AnnotationVMerge = void 0;
(function (ST_AnnotationVMerge) {
    ST_AnnotationVMerge["cont"] = "cont";
    ST_AnnotationVMerge["rest"] = "rest";
})(exports.ST_AnnotationVMerge || (exports.ST_AnnotationVMerge = {}));
exports.ST_TextboxTightWrap = void 0;
(function (ST_TextboxTightWrap) {
    ST_TextboxTightWrap["none"] = "none";
    ST_TextboxTightWrap["allLines"] = "allLines";
    ST_TextboxTightWrap["firstAndLastLine"] = "firstAndLastLine";
    ST_TextboxTightWrap["firstLineOnly"] = "firstLineOnly";
    ST_TextboxTightWrap["lastLineOnly"] = "lastLineOnly";
})(exports.ST_TextboxTightWrap || (exports.ST_TextboxTightWrap = {}));
exports.ST_ObjectDrawAspect = void 0;
(function (ST_ObjectDrawAspect) {
    ST_ObjectDrawAspect["content"] = "content";
    ST_ObjectDrawAspect["icon"] = "icon";
})(exports.ST_ObjectDrawAspect || (exports.ST_ObjectDrawAspect = {}));
exports.ST_ObjectUpdateMode = void 0;
(function (ST_ObjectUpdateMode) {
    ST_ObjectUpdateMode["always"] = "always";
    ST_ObjectUpdateMode["onCall"] = "onCall";
})(exports.ST_ObjectUpdateMode || (exports.ST_ObjectUpdateMode = {}));
exports.ST_FldCharType = void 0;
(function (ST_FldCharType) {
    ST_FldCharType["begin"] = "begin";
    ST_FldCharType["separate"] = "separate";
    ST_FldCharType["end"] = "end";
})(exports.ST_FldCharType || (exports.ST_FldCharType = {}));
exports.ST_InfoTextType = void 0;
(function (ST_InfoTextType) {
    ST_InfoTextType["text"] = "text";
    ST_InfoTextType["autoText"] = "autoText";
})(exports.ST_InfoTextType || (exports.ST_InfoTextType = {}));
exports.ST_FFTextType = void 0;
(function (ST_FFTextType) {
    ST_FFTextType["regular"] = "regular";
    ST_FFTextType["number"] = "number";
    ST_FFTextType["date"] = "date";
    ST_FFTextType["currentTime"] = "currentTime";
    ST_FFTextType["currentDate"] = "currentDate";
    ST_FFTextType["calculated"] = "calculated";
})(exports.ST_FFTextType || (exports.ST_FFTextType = {}));
exports.ST_SectionMark = void 0;
(function (ST_SectionMark) {
    ST_SectionMark["nextPage"] = "nextPage";
    ST_SectionMark["nextColumn"] = "nextColumn";
    ST_SectionMark["continuous"] = "continuous";
    ST_SectionMark["evenPage"] = "evenPage";
    ST_SectionMark["oddPage"] = "oddPage";
})(exports.ST_SectionMark || (exports.ST_SectionMark = {}));
exports.ST_NumberFormat = void 0;
(function (ST_NumberFormat) {
    ST_NumberFormat["decimal"] = "decimal";
    ST_NumberFormat["upperRoman"] = "upperRoman";
    ST_NumberFormat["lowerRoman"] = "lowerRoman";
    ST_NumberFormat["upperLetter"] = "upperLetter";
    ST_NumberFormat["lowerLetter"] = "lowerLetter";
    ST_NumberFormat["ordinal"] = "ordinal";
    ST_NumberFormat["cardinalText"] = "cardinalText";
    ST_NumberFormat["ordinalText"] = "ordinalText";
    ST_NumberFormat["hex"] = "hex";
    ST_NumberFormat["chicago"] = "chicago";
    ST_NumberFormat["ideographDigital"] = "ideographDigital";
    ST_NumberFormat["japaneseCounting"] = "japaneseCounting";
    ST_NumberFormat["aiueo"] = "aiueo";
    ST_NumberFormat["iroha"] = "iroha";
    ST_NumberFormat["decimalFullWidth"] = "decimalFullWidth";
    ST_NumberFormat["decimalHalfWidth"] = "decimalHalfWidth";
    ST_NumberFormat["japaneseLegal"] = "japaneseLegal";
    ST_NumberFormat["japaneseDigitalTenThousand"] = "japaneseDigitalTenThousand";
    ST_NumberFormat["decimalEnclosedCircle"] = "decimalEnclosedCircle";
    ST_NumberFormat["decimalFullWidth2"] = "decimalFullWidth2";
    ST_NumberFormat["aiueoFullWidth"] = "aiueoFullWidth";
    ST_NumberFormat["irohaFullWidth"] = "irohaFullWidth";
    ST_NumberFormat["decimalZero"] = "decimalZero";
    ST_NumberFormat["bullet"] = "bullet";
    ST_NumberFormat["ganada"] = "ganada";
    ST_NumberFormat["chosung"] = "chosung";
    ST_NumberFormat["decimalEnclosedFullstop"] = "decimalEnclosedFullstop";
    ST_NumberFormat["decimalEnclosedParen"] = "decimalEnclosedParen";
    ST_NumberFormat["decimalEnclosedCircleChinese"] = "decimalEnclosedCircleChinese";
    ST_NumberFormat["ideographEnclosedCircle"] = "ideographEnclosedCircle";
    ST_NumberFormat["ideographTraditional"] = "ideographTraditional";
    ST_NumberFormat["ideographZodiac"] = "ideographZodiac";
    ST_NumberFormat["ideographZodiacTraditional"] = "ideographZodiacTraditional";
    ST_NumberFormat["taiwaneseCounting"] = "taiwaneseCounting";
    ST_NumberFormat["ideographLegalTraditional"] = "ideographLegalTraditional";
    ST_NumberFormat["taiwaneseCountingThousand"] = "taiwaneseCountingThousand";
    ST_NumberFormat["taiwaneseDigital"] = "taiwaneseDigital";
    ST_NumberFormat["chineseCounting"] = "chineseCounting";
    ST_NumberFormat["chineseLegalSimplified"] = "chineseLegalSimplified";
    ST_NumberFormat["chineseCountingThousand"] = "chineseCountingThousand";
    ST_NumberFormat["koreanDigital"] = "koreanDigital";
    ST_NumberFormat["koreanCounting"] = "koreanCounting";
    ST_NumberFormat["koreanLegal"] = "koreanLegal";
    ST_NumberFormat["koreanDigital2"] = "koreanDigital2";
    ST_NumberFormat["vietnameseCounting"] = "vietnameseCounting";
    ST_NumberFormat["russianLower"] = "russianLower";
    ST_NumberFormat["russianUpper"] = "russianUpper";
    ST_NumberFormat["none"] = "none";
    ST_NumberFormat["numberInDash"] = "numberInDash";
    ST_NumberFormat["hebrew1"] = "hebrew1";
    ST_NumberFormat["hebrew2"] = "hebrew2";
    ST_NumberFormat["arabicAlpha"] = "arabicAlpha";
    ST_NumberFormat["arabicAbjad"] = "arabicAbjad";
    ST_NumberFormat["hindiVowels"] = "hindiVowels";
    ST_NumberFormat["hindiConsonants"] = "hindiConsonants";
    ST_NumberFormat["hindiNumbers"] = "hindiNumbers";
    ST_NumberFormat["hindiCounting"] = "hindiCounting";
    ST_NumberFormat["thaiLetters"] = "thaiLetters";
    ST_NumberFormat["thaiNumbers"] = "thaiNumbers";
    ST_NumberFormat["thaiCounting"] = "thaiCounting";
    ST_NumberFormat["bahtText"] = "bahtText";
    ST_NumberFormat["dollarText"] = "dollarText";
    ST_NumberFormat["custom"] = "custom";
})(exports.ST_NumberFormat || (exports.ST_NumberFormat = {}));
exports.ST_PageOrientation = void 0;
(function (ST_PageOrientation) {
    ST_PageOrientation["portrait"] = "portrait";
    ST_PageOrientation["landscape"] = "landscape";
})(exports.ST_PageOrientation || (exports.ST_PageOrientation = {}));
exports.ST_PageBorderZOrder = void 0;
(function (ST_PageBorderZOrder) {
    ST_PageBorderZOrder["front"] = "front";
    ST_PageBorderZOrder["back"] = "back";
})(exports.ST_PageBorderZOrder || (exports.ST_PageBorderZOrder = {}));
exports.ST_PageBorderDisplay = void 0;
(function (ST_PageBorderDisplay) {
    ST_PageBorderDisplay["allPages"] = "allPages";
    ST_PageBorderDisplay["firstPage"] = "firstPage";
    ST_PageBorderDisplay["notFirstPage"] = "notFirstPage";
})(exports.ST_PageBorderDisplay || (exports.ST_PageBorderDisplay = {}));
exports.ST_PageBorderOffset = void 0;
(function (ST_PageBorderOffset) {
    ST_PageBorderOffset["page"] = "page";
    ST_PageBorderOffset["text"] = "text";
})(exports.ST_PageBorderOffset || (exports.ST_PageBorderOffset = {}));
exports.ST_ChapterSep = void 0;
(function (ST_ChapterSep) {
    ST_ChapterSep["hyphen"] = "hyphen";
    ST_ChapterSep["period"] = "period";
    ST_ChapterSep["colon"] = "colon";
    ST_ChapterSep["emDash"] = "emDash";
    ST_ChapterSep["enDash"] = "enDash";
})(exports.ST_ChapterSep || (exports.ST_ChapterSep = {}));
exports.ST_LineNumberRestart = void 0;
(function (ST_LineNumberRestart) {
    ST_LineNumberRestart["newPage"] = "newPage";
    ST_LineNumberRestart["newSection"] = "newSection";
    ST_LineNumberRestart["continuous"] = "continuous";
})(exports.ST_LineNumberRestart || (exports.ST_LineNumberRestart = {}));
exports.ST_VerticalJc = void 0;
(function (ST_VerticalJc) {
    ST_VerticalJc["top"] = "top";
    ST_VerticalJc["center"] = "center";
    ST_VerticalJc["both"] = "both";
    ST_VerticalJc["bottom"] = "bottom";
})(exports.ST_VerticalJc || (exports.ST_VerticalJc = {}));
exports.ST_DocGrid = void 0;
(function (ST_DocGrid) {
    ST_DocGrid["default"] = "default";
    ST_DocGrid["lines"] = "lines";
    ST_DocGrid["linesAndChars"] = "linesAndChars";
    ST_DocGrid["snapToChars"] = "snapToChars";
})(exports.ST_DocGrid || (exports.ST_DocGrid = {}));
exports.ST_HdrFtr = void 0;
(function (ST_HdrFtr) {
    ST_HdrFtr["even"] = "even";
    ST_HdrFtr["default"] = "default";
    ST_HdrFtr["first"] = "first";
})(exports.ST_HdrFtr || (exports.ST_HdrFtr = {}));
exports.ST_FtnEdn = void 0;
(function (ST_FtnEdn) {
    ST_FtnEdn["normal"] = "normal";
    ST_FtnEdn["separator"] = "separator";
    ST_FtnEdn["continuationSeparator"] = "continuationSeparator";
    ST_FtnEdn["continuationNotice"] = "continuationNotice";
})(exports.ST_FtnEdn || (exports.ST_FtnEdn = {}));
exports.ST_BrType = void 0;
(function (ST_BrType) {
    ST_BrType["page"] = "page";
    ST_BrType["column"] = "column";
    ST_BrType["textWrapping"] = "textWrapping";
})(exports.ST_BrType || (exports.ST_BrType = {}));
exports.ST_BrClear = void 0;
(function (ST_BrClear) {
    ST_BrClear["none"] = "none";
    ST_BrClear["left"] = "left";
    ST_BrClear["right"] = "right";
    ST_BrClear["all"] = "all";
})(exports.ST_BrClear || (exports.ST_BrClear = {}));
exports.ST_PTabAlignment = void 0;
(function (ST_PTabAlignment) {
    ST_PTabAlignment["left"] = "left";
    ST_PTabAlignment["center"] = "center";
    ST_PTabAlignment["right"] = "right";
})(exports.ST_PTabAlignment || (exports.ST_PTabAlignment = {}));
exports.ST_PTabRelativeTo = void 0;
(function (ST_PTabRelativeTo) {
    ST_PTabRelativeTo["margin"] = "margin";
    ST_PTabRelativeTo["indent"] = "indent";
})(exports.ST_PTabRelativeTo || (exports.ST_PTabRelativeTo = {}));
exports.ST_PTabLeader = void 0;
(function (ST_PTabLeader) {
    ST_PTabLeader["none"] = "none";
    ST_PTabLeader["dot"] = "dot";
    ST_PTabLeader["hyphen"] = "hyphen";
    ST_PTabLeader["underscore"] = "underscore";
    ST_PTabLeader["middleDot"] = "middleDot";
})(exports.ST_PTabLeader || (exports.ST_PTabLeader = {}));
exports.ST_ProofErr = void 0;
(function (ST_ProofErr) {
    ST_ProofErr["spellStart"] = "spellStart";
    ST_ProofErr["spellEnd"] = "spellEnd";
    ST_ProofErr["gramStart"] = "gramStart";
    ST_ProofErr["gramEnd"] = "gramEnd";
})(exports.ST_ProofErr || (exports.ST_ProofErr = {}));
exports.ST_EdGrp = void 0;
(function (ST_EdGrp) {
    ST_EdGrp["none"] = "none";
    ST_EdGrp["everyone"] = "everyone";
    ST_EdGrp["administrators"] = "administrators";
    ST_EdGrp["contributors"] = "contributors";
    ST_EdGrp["editors"] = "editors";
    ST_EdGrp["owners"] = "owners";
    ST_EdGrp["current"] = "current";
})(exports.ST_EdGrp || (exports.ST_EdGrp = {}));
exports.ST_Hint = void 0;
(function (ST_Hint) {
    ST_Hint["default"] = "default";
    ST_Hint["eastAsia"] = "eastAsia";
})(exports.ST_Hint || (exports.ST_Hint = {}));
exports.ST_Theme = void 0;
(function (ST_Theme) {
    ST_Theme["majorEastAsia"] = "majorEastAsia";
    ST_Theme["majorBidi"] = "majorBidi";
    ST_Theme["majorAscii"] = "majorAscii";
    ST_Theme["majorHAnsi"] = "majorHAnsi";
    ST_Theme["minorEastAsia"] = "minorEastAsia";
    ST_Theme["minorBidi"] = "minorBidi";
    ST_Theme["minorAscii"] = "minorAscii";
    ST_Theme["minorHAnsi"] = "minorHAnsi";
})(exports.ST_Theme || (exports.ST_Theme = {}));
exports.ST_RubyAlign = void 0;
(function (ST_RubyAlign) {
    ST_RubyAlign["center"] = "center";
    ST_RubyAlign["distributeLetter"] = "distributeLetter";
    ST_RubyAlign["distributeSpace"] = "distributeSpace";
    ST_RubyAlign["left"] = "left";
    ST_RubyAlign["right"] = "right";
    ST_RubyAlign["rightVertical"] = "rightVertical";
})(exports.ST_RubyAlign || (exports.ST_RubyAlign = {}));
exports.ST_Lock = void 0;
(function (ST_Lock) {
    ST_Lock["sdtLocked"] = "sdtLocked";
    ST_Lock["contentLocked"] = "contentLocked";
    ST_Lock["unlocked"] = "unlocked";
    ST_Lock["sdtContentLocked"] = "sdtContentLocked";
})(exports.ST_Lock || (exports.ST_Lock = {}));
exports.ST_SdtDateMappingType = void 0;
(function (ST_SdtDateMappingType) {
    ST_SdtDateMappingType["text"] = "text";
    ST_SdtDateMappingType["date"] = "date";
    ST_SdtDateMappingType["dateTime"] = "dateTime";
})(exports.ST_SdtDateMappingType || (exports.ST_SdtDateMappingType = {}));
exports.ST_Direction = void 0;
(function (ST_Direction) {
    ST_Direction["ltr"] = "ltr";
    ST_Direction["rtl"] = "rtl";
})(exports.ST_Direction || (exports.ST_Direction = {}));
exports.ST_TblWidth = void 0;
(function (ST_TblWidth) {
    ST_TblWidth["nil"] = "nil";
    ST_TblWidth["pct"] = "pct";
    ST_TblWidth["dxa"] = "dxa";
    ST_TblWidth["auto"] = "auto";
})(exports.ST_TblWidth || (exports.ST_TblWidth = {}));
exports.ST_Merge = void 0;
(function (ST_Merge) {
    ST_Merge["continue"] = "continue";
    ST_Merge["restart"] = "restart";
})(exports.ST_Merge || (exports.ST_Merge = {}));
exports.ST_TblLayoutType = void 0;
(function (ST_TblLayoutType) {
    ST_TblLayoutType["fixed"] = "fixed";
    ST_TblLayoutType["autofit"] = "autofit";
})(exports.ST_TblLayoutType || (exports.ST_TblLayoutType = {}));
exports.ST_TblOverlap = void 0;
(function (ST_TblOverlap) {
    ST_TblOverlap["never"] = "never";
    ST_TblOverlap["overlap"] = "overlap";
})(exports.ST_TblOverlap || (exports.ST_TblOverlap = {}));
exports.ST_FtnPos = void 0;
(function (ST_FtnPos) {
    ST_FtnPos["pageBottom"] = "pageBottom";
    ST_FtnPos["beneathText"] = "beneathText";
    ST_FtnPos["sectEnd"] = "sectEnd";
    ST_FtnPos["docEnd"] = "docEnd";
})(exports.ST_FtnPos || (exports.ST_FtnPos = {}));
exports.ST_EdnPos = void 0;
(function (ST_EdnPos) {
    ST_EdnPos["sectEnd"] = "sectEnd";
    ST_EdnPos["docEnd"] = "docEnd";
})(exports.ST_EdnPos || (exports.ST_EdnPos = {}));
exports.ST_RestartNumber = void 0;
(function (ST_RestartNumber) {
    ST_RestartNumber["continuous"] = "continuous";
    ST_RestartNumber["eachSect"] = "eachSect";
    ST_RestartNumber["eachPage"] = "eachPage";
})(exports.ST_RestartNumber || (exports.ST_RestartNumber = {}));
exports.ST_MailMergeSourceType = void 0;
(function (ST_MailMergeSourceType) {
    ST_MailMergeSourceType["database"] = "database";
    ST_MailMergeSourceType["addressBook"] = "addressBook";
    ST_MailMergeSourceType["document1"] = "document1";
    ST_MailMergeSourceType["document2"] = "document2";
    ST_MailMergeSourceType["text"] = "text";
    ST_MailMergeSourceType["email"] = "email";
    ST_MailMergeSourceType["native"] = "native";
    ST_MailMergeSourceType["legacy"] = "legacy";
    ST_MailMergeSourceType["master"] = "master";
})(exports.ST_MailMergeSourceType || (exports.ST_MailMergeSourceType = {}));
exports.ST_CharacterSpacing = void 0;
(function (ST_CharacterSpacing) {
    ST_CharacterSpacing["doNotCompress"] = "doNotCompress";
    ST_CharacterSpacing["compressPunctuation"] = "compressPunctuation";
    ST_CharacterSpacing["compressPunctuationAndJapaneseKana"] = "compressPunctuationAndJapaneseKana";
})(exports.ST_CharacterSpacing || (exports.ST_CharacterSpacing = {}));
exports.ST_WmlColorSchemeIndex = void 0;
(function (ST_WmlColorSchemeIndex) {
    ST_WmlColorSchemeIndex["dark1"] = "dark1";
    ST_WmlColorSchemeIndex["light1"] = "light1";
    ST_WmlColorSchemeIndex["dark2"] = "dark2";
    ST_WmlColorSchemeIndex["light2"] = "light2";
    ST_WmlColorSchemeIndex["accent1"] = "accent1";
    ST_WmlColorSchemeIndex["accent2"] = "accent2";
    ST_WmlColorSchemeIndex["accent3"] = "accent3";
    ST_WmlColorSchemeIndex["accent4"] = "accent4";
    ST_WmlColorSchemeIndex["accent5"] = "accent5";
    ST_WmlColorSchemeIndex["accent6"] = "accent6";
    ST_WmlColorSchemeIndex["hyperlink"] = "hyperlink";
    ST_WmlColorSchemeIndex["followedHyperlink"] = "followedHyperlink";
})(exports.ST_WmlColorSchemeIndex || (exports.ST_WmlColorSchemeIndex = {}));
exports.ST_StyleSort = void 0;
(function (ST_StyleSort) {
    ST_StyleSort["name"] = "name";
    ST_StyleSort["priority"] = "priority";
    ST_StyleSort["default"] = "default";
    ST_StyleSort["font"] = "font";
    ST_StyleSort["basedOn"] = "basedOn";
    ST_StyleSort["type"] = "type";
})(exports.ST_StyleSort || (exports.ST_StyleSort = {}));
exports.ST_FrameScrollbar = void 0;
(function (ST_FrameScrollbar) {
    ST_FrameScrollbar["on"] = "on";
    ST_FrameScrollbar["off"] = "off";
    ST_FrameScrollbar["auto"] = "auto";
})(exports.ST_FrameScrollbar || (exports.ST_FrameScrollbar = {}));
exports.ST_FrameLayout = void 0;
(function (ST_FrameLayout) {
    ST_FrameLayout["rows"] = "rows";
    ST_FrameLayout["cols"] = "cols";
    ST_FrameLayout["none"] = "none";
})(exports.ST_FrameLayout || (exports.ST_FrameLayout = {}));
exports.ST_LevelSuffix = void 0;
(function (ST_LevelSuffix) {
    ST_LevelSuffix["tab"] = "tab";
    ST_LevelSuffix["space"] = "space";
    ST_LevelSuffix["nothing"] = "nothing";
})(exports.ST_LevelSuffix || (exports.ST_LevelSuffix = {}));
exports.ST_MultiLevelType = void 0;
(function (ST_MultiLevelType) {
    ST_MultiLevelType["singleLevel"] = "singleLevel";
    ST_MultiLevelType["multilevel"] = "multilevel";
    ST_MultiLevelType["hybridMultilevel"] = "hybridMultilevel";
})(exports.ST_MultiLevelType || (exports.ST_MultiLevelType = {}));
exports.ST_TblStyleOverrideType = void 0;
(function (ST_TblStyleOverrideType) {
    ST_TblStyleOverrideType["wholeTable"] = "wholeTable";
    ST_TblStyleOverrideType["firstRow"] = "firstRow";
    ST_TblStyleOverrideType["lastRow"] = "lastRow";
    ST_TblStyleOverrideType["firstCol"] = "firstCol";
    ST_TblStyleOverrideType["lastCol"] = "lastCol";
    ST_TblStyleOverrideType["band1Vert"] = "band1Vert";
    ST_TblStyleOverrideType["band2Vert"] = "band2Vert";
    ST_TblStyleOverrideType["band1Horz"] = "band1Horz";
    ST_TblStyleOverrideType["band2Horz"] = "band2Horz";
    ST_TblStyleOverrideType["neCell"] = "neCell";
    ST_TblStyleOverrideType["nwCell"] = "nwCell";
    ST_TblStyleOverrideType["seCell"] = "seCell";
    ST_TblStyleOverrideType["swCell"] = "swCell";
})(exports.ST_TblStyleOverrideType || (exports.ST_TblStyleOverrideType = {}));
exports.ST_StyleType = void 0;
(function (ST_StyleType) {
    ST_StyleType["paragraph"] = "paragraph";
    ST_StyleType["character"] = "character";
    ST_StyleType["table"] = "table";
    ST_StyleType["numbering"] = "numbering";
})(exports.ST_StyleType || (exports.ST_StyleType = {}));
exports.ST_FontFamily = void 0;
(function (ST_FontFamily) {
    ST_FontFamily["decorative"] = "decorative";
    ST_FontFamily["modern"] = "modern";
    ST_FontFamily["roman"] = "roman";
    ST_FontFamily["script"] = "script";
    ST_FontFamily["swiss"] = "swiss";
    ST_FontFamily["auto"] = "auto";
})(exports.ST_FontFamily || (exports.ST_FontFamily = {}));
exports.ST_Pitch = void 0;
(function (ST_Pitch) {
    ST_Pitch["fixed"] = "fixed";
    ST_Pitch["variable"] = "variable";
    ST_Pitch["default"] = "default";
})(exports.ST_Pitch || (exports.ST_Pitch = {}));
exports.ST_ThemeColor = void 0;
(function (ST_ThemeColor) {
    ST_ThemeColor["dark1"] = "dark1";
    ST_ThemeColor["light1"] = "light1";
    ST_ThemeColor["dark2"] = "dark2";
    ST_ThemeColor["light2"] = "light2";
    ST_ThemeColor["accent1"] = "accent1";
    ST_ThemeColor["accent2"] = "accent2";
    ST_ThemeColor["accent3"] = "accent3";
    ST_ThemeColor["accent4"] = "accent4";
    ST_ThemeColor["accent5"] = "accent5";
    ST_ThemeColor["accent6"] = "accent6";
    ST_ThemeColor["hyperlink"] = "hyperlink";
    ST_ThemeColor["followedHyperlink"] = "followedHyperlink";
    ST_ThemeColor["none"] = "none";
    ST_ThemeColor["background1"] = "background1";
    ST_ThemeColor["text1"] = "text1";
    ST_ThemeColor["background2"] = "background2";
    ST_ThemeColor["text2"] = "text2";
})(exports.ST_ThemeColor || (exports.ST_ThemeColor = {}));
exports.ST_DocPartBehavior = void 0;
(function (ST_DocPartBehavior) {
    ST_DocPartBehavior["content"] = "content";
    ST_DocPartBehavior["p"] = "p";
    ST_DocPartBehavior["pg"] = "pg";
})(exports.ST_DocPartBehavior || (exports.ST_DocPartBehavior = {}));
exports.ST_DocPartType = void 0;
(function (ST_DocPartType) {
    ST_DocPartType["none"] = "none";
    ST_DocPartType["normal"] = "normal";
    ST_DocPartType["autoExp"] = "autoExp";
    ST_DocPartType["toolbar"] = "toolbar";
    ST_DocPartType["speller"] = "speller";
    ST_DocPartType["formFld"] = "formFld";
    ST_DocPartType["bbPlcHdr"] = "bbPlcHdr";
})(exports.ST_DocPartType || (exports.ST_DocPartType = {}));
exports.ST_DocPartGallery = void 0;
(function (ST_DocPartGallery) {
    ST_DocPartGallery["placeholder"] = "placeholder";
    ST_DocPartGallery["any"] = "any";
    ST_DocPartGallery["default"] = "default";
    ST_DocPartGallery["docParts"] = "docParts";
    ST_DocPartGallery["coverPg"] = "coverPg";
    ST_DocPartGallery["eq"] = "eq";
    ST_DocPartGallery["ftrs"] = "ftrs";
    ST_DocPartGallery["hdrs"] = "hdrs";
    ST_DocPartGallery["pgNum"] = "pgNum";
    ST_DocPartGallery["tbls"] = "tbls";
    ST_DocPartGallery["watermarks"] = "watermarks";
    ST_DocPartGallery["autoTxt"] = "autoTxt";
    ST_DocPartGallery["txtBox"] = "txtBox";
    ST_DocPartGallery["pgNumT"] = "pgNumT";
    ST_DocPartGallery["pgNumB"] = "pgNumB";
    ST_DocPartGallery["pgNumMargins"] = "pgNumMargins";
    ST_DocPartGallery["tblOfContents"] = "tblOfContents";
    ST_DocPartGallery["bib"] = "bib";
    ST_DocPartGallery["custQuickParts"] = "custQuickParts";
    ST_DocPartGallery["custCoverPg"] = "custCoverPg";
    ST_DocPartGallery["custEq"] = "custEq";
    ST_DocPartGallery["custFtrs"] = "custFtrs";
    ST_DocPartGallery["custHdrs"] = "custHdrs";
    ST_DocPartGallery["custPgNum"] = "custPgNum";
    ST_DocPartGallery["custTbls"] = "custTbls";
    ST_DocPartGallery["custWatermarks"] = "custWatermarks";
    ST_DocPartGallery["custAutoTxt"] = "custAutoTxt";
    ST_DocPartGallery["custTxtBox"] = "custTxtBox";
    ST_DocPartGallery["custPgNumT"] = "custPgNumT";
    ST_DocPartGallery["custPgNumB"] = "custPgNumB";
    ST_DocPartGallery["custPgNumMargins"] = "custPgNumMargins";
    ST_DocPartGallery["custTblOfContents"] = "custTblOfContents";
    ST_DocPartGallery["custBib"] = "custBib";
    ST_DocPartGallery["custom1"] = "custom1";
    ST_DocPartGallery["custom2"] = "custom2";
    ST_DocPartGallery["custom3"] = "custom3";
    ST_DocPartGallery["custom4"] = "custom4";
    ST_DocPartGallery["custom5"] = "custom5";
})(exports.ST_DocPartGallery || (exports.ST_DocPartGallery = {}));
exports.ST_CaptionPos = void 0;
(function (ST_CaptionPos) {
    ST_CaptionPos["above"] = "above";
    ST_CaptionPos["below"] = "below";
    ST_CaptionPos["left"] = "left";
    ST_CaptionPos["right"] = "right";
})(exports.ST_CaptionPos || (exports.ST_CaptionPos = {}));
