// amis/esm/renderers/Divider.js
/**
 * amis v2.9.0
 * Copyright 2018-2023 baidu
 */

import { __extends, __decorate } from 'tslib';
import React from 'react';
import { Renderer } from 'amis-core';

var Divider = /** @class */ (function (_super) {
    __extends(Divider, _super);
    function Divider() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Divider.prototype.render = function () {
        var _a = this.props, cx = _a.classnames, className = _a.className, style = _a.style, lineStyle = _a.lineStyle;
        return (React.createElement("div", { className: cx('Divider', lineStyle ? "Divider--".concat(lineStyle) : '', className), style: style }));
    };
    Divider.defaultProps = {
        className: '',
        lineStyle: 'dashed'
    };
    return Divider;
}(React.Component));
/** @class */ ((function (_super) {
    __extends(DividerRenderer, _super);
    function DividerRenderer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DividerRenderer = __decorate([
        Renderer({
            type: 'divider'
        })
    ], DividerRenderer);
    return DividerRenderer;
})(Divider));

export { Divider as default };
