// amis-ui/lib/components/formula/Editor.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var React = require('react');
var amisCore = require('amis-core');
var doc = require('amis-formula/lib/doc');
var plugin = require('./plugin.js');
var FuncList = require('./FuncList.js');
var VariableList = require('./VariableList.js');
var CodeMirror = require('../CodeMirror.js');
var Toast = require('../Toast.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var FormulaEditor = /** @class */ (function (_super) {
    tslib.__extends(FormulaEditor, _super);
    function FormulaEditor() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            focused: false
        };
        return _this;
    }
    FormulaEditor.buildDefaultFunctions = function (doc) {
        var funcs = [];
        doc.forEach(function (item) {
            var namespace = item.namespace || 'Others';
            var exists = funcs.find(function (item) { return item.groupName === namespace; });
            if (!exists) {
                exists = {
                    groupName: namespace,
                    items: []
                };
                funcs.push(exists);
            }
            exists.items.push(item);
        });
        return funcs;
    };
    FormulaEditor.replaceStrByIndex = function (str, idx, key, replaceKey) {
        var from = str.slice(0, idx);
        var left = str.slice(idx);
        return from + left.replace(key, replaceKey);
    };
    FormulaEditor.getRegExpByMode = function (evalMode, key) {
        var reg = evalMode
            ? "\\b".concat(key, "\\b")
            : "\\$\\{[^\\{\\}]*\\b".concat(key, "\\b[^\\{\\}]*\\}");
        return new RegExp(reg);
    };
    FormulaEditor.highlightValue = function (value, variables, evalMode) {
        if (evalMode === void 0) { evalMode = true; }
        if (!Array.isArray(variables) || !variables.length || !value) {
            return;
        }
        var varMap = {};
        amisCore.eachTree(variables, function (item) {
            if (item.value) {
                var key = item.value;
                varMap[key] = item.label;
            }
        });
        var vars = Object.keys(varMap)
            .filter(function (item) { return item; })
            .sort(function (a, b) { return b.length - a.length; });
        var content = value || '';
        var html = '';
        // 标记方法调用
        html = content.replace(/([A-Z]+)\s*\(/g, function (_, func, pos) {
            return _ === null || _ === void 0 ? void 0 : _.replace(func, "<span class=\"c-func\">".concat(func, "</span>"));
        });
        var REPLACE_KEY = 'AMIS_FORMULA_REPLACE_KEY';
        vars.forEach(function (v) {
            var from = 0;
            var idx = -1;
            while (~(idx = content.indexOf(v, from))) {
                var encodeHtml = FormulaEditor.replaceStrByIndex(html, idx, v, REPLACE_KEY);
                var reg = FormulaEditor.getRegExpByMode(evalMode, REPLACE_KEY);
                // 如果匹配到则高亮，没有匹配到替换成原值
                if (reg.test(encodeHtml)) {
                    html = encodeHtml.replace(REPLACE_KEY, "<span class=\"c-field\">".concat(varMap[v], "</span>"));
                }
                else {
                    html = encodeHtml.replace(REPLACE_KEY, v);
                }
                from = idx + v.length;
            }
        });
        return { html: html };
    };
    FormulaEditor.prototype.componentWillUnmount = function () {
        var _a;
        (_a = this.editorPlugin) === null || _a === void 0 ? void 0 : _a.dispose();
    };
    FormulaEditor.prototype.handleFocus = function () {
        this.setState({
            focused: true
        });
    };
    FormulaEditor.prototype.handleBlur = function () {
        this.setState({
            focused: false
        });
    };
    FormulaEditor.prototype.insertValue = function (value, type) {
        var _a;
        (_a = this.editorPlugin) === null || _a === void 0 ? void 0 : _a.insertContent(value, type);
    };
    FormulaEditor.prototype.handleEditorMounted = function (cm, editor) {
        var _this = this;
        this.editorPlugin = new plugin.FormulaPlugin(editor, cm, function () { return _this.props; });
    };
    FormulaEditor.prototype.validate = function () {
        var value = this.props.value;
        try {
            value
                ? amisCore.parse(value, {
                    evalMode: this.props.evalMode
                })
                : null;
        }
        catch (e) {
            return e.message;
        }
        return;
    };
    FormulaEditor.prototype.handleFunctionSelect = function (item) {
        var _a;
        (_a = this.editorPlugin) === null || _a === void 0 ? void 0 : _a.insertContent("".concat(item.name), 'func');
    };
    FormulaEditor.prototype.handleVariableSelect = function (item) {
        var _a;
        var _b = this.props; _b.evalMode; var selfVariableName = _b.selfVariableName;
        if (item &&
            item.value &&
            selfVariableName &&
            selfVariableName === item.value) {
            Toast.toast.warning('不能使用当前变量[self]，避免循环引用。');
            return;
        }
        if (!item.value) {
            return;
        }
        (_a = this.editorPlugin) === null || _a === void 0 ? void 0 : _a.insertContent({
            key: item.value,
            name: item.label
        }, 'variable');
    };
    FormulaEditor.prototype.handleOnChange = function (value) {
        var onChange = this.props.onChange;
        onChange === null || onChange === void 0 ? void 0 : onChange(value);
    };
    FormulaEditor.prototype.editorFactory = function (dom, cm) {
        return plugin.editorFactory(dom, cm, this.props);
    };
    FormulaEditor.prototype.render = function () {
        var _a = this.props, variables = _a.variables, header = _a.header, value = _a.value, functions = _a.functions, variableMode = _a.variableMode, __ = _a.translate, cx = _a.classnames, variableClassName = _a.variableClassName, functionClassName = _a.functionClassName, classPrefix = _a.classPrefix, selfVariableName = _a.selfVariableName;
        var focused = this.state.focused;
        var customFunctions = Array.isArray(functions) ? functions : [];
        var functionList = tslib.__spreadArray(tslib.__spreadArray([], tslib.__read(FormulaEditor.buildDefaultFunctions(doc.doc)), false), tslib.__read(customFunctions), false);
        return (React__default["default"].createElement("div", { className: cx("FormulaEditor", {
                'is-focused': focused
            }) },
            React__default["default"].createElement("section", { className: cx("FormulaEditor-content") },
                React__default["default"].createElement("header", { className: cx("FormulaEditor-header") }, __(header || 'FormulaEditor.title')),
                React__default["default"].createElement(CodeMirror.CodeMirrorEditor, { className: cx('FormulaEditor-editor'), value: value, onChange: this.handleOnChange, editorFactory: this.editorFactory, editorDidMount: this.handleEditorMounted, onFocus: this.handleFocus, onBlur: this.handleBlur })),
            React__default["default"].createElement("section", { className: cx('FormulaEditor-settings') },
                React__default["default"].createElement("div", { className: cx('FormulaEditor-panel') },
                    variableMode !== 'tabs' ? (React__default["default"].createElement("div", { className: cx('FormulaEditor-panel-header') }, __('FormulaEditor.variable'))) : null,
                    React__default["default"].createElement("div", { className: cx('FormulaEditor-panel-body', variableMode && "FormulaEditor-panel-body--".concat(variableMode)) },
                        React__default["default"].createElement(VariableList["default"], { classPrefix: classPrefix, className: cx('FormulaEditor-VariableList', 'FormulaEditor-VariableList-root', variableClassName), selectMode: variableMode, data: variables, onSelect: this.handleVariableSelect, selfVariableName: selfVariableName }))),
                React__default["default"].createElement(FuncList["default"], { className: functionClassName, title: __('FormulaEditor.function'), data: functionList, onSelect: this.handleFunctionSelect }))));
    };
    FormulaEditor.defaultProps = {
        variables: [],
        evalMode: true
    };
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", []),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaEditor.prototype, "handleFocus", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", []),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaEditor.prototype, "handleBlur", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [Object, String]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaEditor.prototype, "insertValue", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [Object, Object]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaEditor.prototype, "handleEditorMounted", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", []),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaEditor.prototype, "validate", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [Object]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaEditor.prototype, "handleFunctionSelect", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [Object]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaEditor.prototype, "handleVariableSelect", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [Object]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaEditor.prototype, "handleOnChange", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [HTMLElement, Object]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaEditor.prototype, "editorFactory", null);
    return FormulaEditor;
}(React__default["default"].Component));
var Editor = amisCore.uncontrollable(amisCore.themeable(amisCore.localeable(FormulaEditor)), {
    value: 'onChange'
}, ['validate']);

exports.FormulaEditor = FormulaEditor;
exports["default"] = Editor;
