// amis-ui/lib/components/formula/Picker.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var amisCore = require('amis-core');
var React = require('react');
var Editor = require('./Editor.js');
var ResultBox = require('../ResultBox.js');
var Button = require('../Button.js');
var icons = require('../icons.js');
var Modal = require('../Modal.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var FormulaPicker = /** @class */ (function (_super) {
    tslib.__extends(FormulaPicker, _super);
    function FormulaPicker(props) {
        var _this = _super.call(this, props) || this;
        _this.props.onRef && _this.props.onRef(_this);
        _this.state = {
            isOpened: false,
            value: _this.props.value,
            editorValue: _this.value2EditorValue(_this.props),
            isError: false
        };
        return _this;
    }
    FormulaPicker.prototype.componentDidUpdate = function (prevProps) {
        var value = this.props.value;
        if (value !== prevProps.value) {
            this.setState({
                value: typeof value === 'string' ? value : '',
                editorValue: this.value2EditorValue(this.props)
            });
        }
    };
    FormulaPicker.prototype.value2EditorValue = function (props) {
        if (props.mixedMode) {
            if (typeof props.value === 'string' &&
                /^\s*\$\{(.+?)\}\s*$/.test(props.value)) {
                return RegExp.$1;
            }
            else {
                return '';
            }
        }
        return String(props.value || '');
    };
    FormulaPicker.prototype.handleConfirm = function () {
        var _a, _b;
        var value = this.state.value;
        if (this.props.onConfirm) {
            this.props.onConfirm(value);
        }
        else {
            (_b = (_a = this.props).onChange) === null || _b === void 0 ? void 0 : _b.call(_a, value);
        }
    };
    FormulaPicker.prototype.renderFormulaValue = function (item) {
        var _a = this.props, allowInput = _a.allowInput, cx = _a.classnames;
        var html = { __html: item.html };
        if (allowInput) {
            return '';
        }
        return (React__default["default"].createElement("div", { className: cx('FormulaPicker-ResultBox'), dangerouslySetInnerHTML: html }));
    };
    FormulaPicker.prototype.handleInputChange = function (value) {
        var _this = this;
        this.setState({ value: value }, function () { return _this.handleConfirm(); });
    };
    FormulaPicker.prototype.handleInputGroupChange = function (e) {
        var onChange = this.props.onChange;
        onChange && onChange(e.currentTarget.value);
    };
    FormulaPicker.prototype.handleEditorChange = function (value) {
        this.setState({
            editorValue: value,
            isError: false
        });
    };
    FormulaPicker.prototype.handleEditorConfirm = function () {
        this.props.translate;
        var value = this.state.editorValue;
        this.confirm(value);
    };
    FormulaPicker.prototype.confirm = function (value) {
        var _this = this;
        var mixedMode = this.props.mixedMode;
        var validate = this.validate(value);
        if (validate === true) {
            this.setState({ value: mixedMode && value ? "${".concat(value, "}") : value }, function () {
                _this.close(undefined, function () { return _this.handleConfirm(); });
            });
        }
        else {
            this.setState({ isError: validate });
        }
    };
    FormulaPicker.prototype.handleClick = function () {
        var _a, _b;
        return tslib.__awaiter(this, void 0, void 0, function () {
            var state, _c;
            return tslib.__generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _c = [{}];
                        return [4 /*yield*/, ((_b = (_a = this.props).onPickerOpen) === null || _b === void 0 ? void 0 : _b.call(_a, this.props))];
                    case 1:
                        state = tslib.__assign.apply(void 0, [tslib.__assign.apply(void 0, _c.concat([(_d.sent())])), { editorValue: this.value2EditorValue(this.props), isOpened: true }]);
                        this.setState(state);
                        return [2 /*return*/];
                }
            });
        });
    };
    FormulaPicker.prototype.close = function (e, callback) {
        this.setState({
            isOpened: false,
            isError: false
        }, function () {
            if (callback) {
                callback();
                return;
            }
        });
    };
    FormulaPicker.prototype.updateState = function (state) {
        if (state === void 0) { state = {}; }
        state.isOpened; var rest = tslib.__rest(state, ["isOpened"]);
        this.setState(tslib.__assign(tslib.__assign({}, this.state), rest));
    };
    FormulaPicker.prototype.validate = function (value) {
        this.props.translate;
        try {
            value &&
                amisCore.parse(value, {
                    evalMode: this.props.mixedMode ? true : this.props.evalMode,
                    allowFilter: false
                });
            return true;
        }
        catch (e) {
            if (/\s(\d+:\d+)$/.test(e.message)) {
                var _a = tslib.__read(/\s(\d+:\d+)$/.exec(e.message) || [], 2), position = _a[1];
                return position;
            }
            return e.message;
        }
    };
    FormulaPicker.prototype.render = function () {
        var _a, _b, _c;
        var _d, _e, _f, _g, _h;
        var _j = this.props, cx = _j.classnames, __ = _j.translate, disabled = _j.disabled, _k = _j.allowInput, allowInput = _k === void 0 ? true : _k, className = _j.className, style = _j.style; _j.onChange; _j.size; var borderMode = _j.borderMode, placeholder = _j.placeholder, _l = _j.mode, mode = _l === void 0 ? 'input-button' : _l, btnLabel = _j.btnLabel, level = _j.level, btnSize = _j.btnSize, icon = _j.icon, title = _j.title, clearable = _j.clearable, variables = _j.variables, functions = _j.functions, children = _j.children, variableMode = _j.variableMode, mixedMode = _j.mixedMode, evalMode = _j.evalMode, rest = tslib.__rest(_j, ["classnames", "translate", "disabled", "allowInput", "className", "style", "onChange", "size", "borderMode", "placeholder", "mode", "btnLabel", "level", "btnSize", "icon", "title", "clearable", "variables", "functions", "children", "variableMode", "mixedMode", "evalMode"]);
        var _m = this.state, isOpened = _m.isOpened, value = _m.value, editorValue = _m.editorValue, isError = _m.isError;
        var iconElement = amisCore.generateIcon(cx, icon, 'Icon');
        return (React__default["default"].createElement(React__default["default"].Fragment, null,
            children ? (children({
                isOpened: this.state.isOpened,
                onClick: this.handleClick,
                setState: this.updateState
            })) : (React__default["default"].createElement("div", { className: cx('FormulaPicker', className, mode === 'input-group' ? 'is-input-group' : ''), style: style },
                mode === 'button' && (React__default["default"].createElement(Button["default"], { className: cx('FormulaPicker-action', 'w-full'), level: level, size: btnSize, onClick: this.handleClick },
                    iconElement ? (React__default["default"].cloneElement(iconElement, {
                        className: cx((_e = (_d = iconElement === null || iconElement === void 0 ? void 0 : iconElement.props) === null || _d === void 0 ? void 0 : _d.className) !== null && _e !== void 0 ? _e : '', 'FormulaPicker-icon', (_a = {},
                            _a['is-filled'] = !!value,
                            _a))
                    })) : (React__default["default"].createElement(icons.Icon, { icon: "function", className: cx('FormulaPicker-icon', 'icon', (_b = {},
                            _b['is-filled'] = !!value,
                            _b)) })),
                    React__default["default"].createElement("span", { className: cx('FormulaPicker-label') }, __(btnLabel || 'FormulaEditor.btnLabel')))),
                mode === 'input-button' && (React__default["default"].createElement(React__default["default"].Fragment, null,
                    React__default["default"].createElement(ResultBox["default"], { className: cx('FormulaPicker-input', isOpened ? 'is-active' : '', !!isError ? 'is-error' : ''), allowInput: allowInput, clearable: clearable, value: value, result: allowInput
                            ? void 0
                            : Editor.FormulaEditor.highlightValue(value, variables, this.props.evalMode), itemRender: this.renderFormulaValue, onResultChange: amisCore.noop, onChange: this.handleInputChange, disabled: disabled, borderMode: borderMode, placeholder: placeholder }),
                    React__default["default"].createElement(Button["default"], { className: cx('FormulaPicker-action'), onClick: this.handleClick },
                        React__default["default"].createElement(icons.Icon, { icon: "function", className: cx('FormulaPicker-icon', 'icon', (_c = {},
                                _c['is-filled'] = !!value,
                                _c)) })))),
                mode === 'input-group' && (React__default["default"].createElement(React__default["default"].Fragment, null,
                    React__default["default"].createElement(ResultBox["default"], { className: cx('FormulaPicker-input', isOpened ? 'is-active' : '', !!isError ? 'is-error' : ''), allowInput: allowInput, clearable: clearable, value: value, result: allowInput
                            ? void 0
                            : Editor.FormulaEditor.highlightValue(value, variables, this.props.evalMode), itemRender: this.renderFormulaValue, onResultChange: amisCore.noop, onChange: this.handleInputChange, disabled: disabled, borderMode: borderMode, placeholder: placeholder }),
                    React__default["default"].createElement("a", { className: cx("FormulaPicker-toggler"), onClick: this.handleClick },
                        React__default["default"].createElement(icons.Icon, { icon: "function", className: "icon" })))))),
            React__default["default"].createElement(Modal["default"], { size: "md", closeOnEsc: true, show: this.state.isOpened, onHide: this.close },
                React__default["default"].createElement(Modal["default"].Header, { onClose: this.close, className: "font-bold" }, __(title || 'FormulaEditor.title')),
                React__default["default"].createElement(Modal["default"].Body, null,
                    React__default["default"].createElement(Editor["default"], tslib.__assign({}, rest, { evalMode: mixedMode ? true : evalMode, variables: (_f = this.state.variables) !== null && _f !== void 0 ? _f : variables, functions: (_g = this.state.functions) !== null && _g !== void 0 ? _g : functions, variableMode: (_h = this.state.variableMode) !== null && _h !== void 0 ? _h : variableMode, value: editorValue, onChange: this.handleEditorChange, selfVariableName: this.props.selfVariableName }))),
                React__default["default"].createElement(Modal["default"].Footer, null,
                    !!isError ? (React__default["default"].createElement("div", { className: cx('Dialog-info'), key: "info" },
                        React__default["default"].createElement("span", { className: cx('Dialog-error') }, __('FormulaEditor.invalidData', { err: isError })))) : null,
                    React__default["default"].createElement(Button["default"], { onClick: this.close }, __('cancel')),
                    React__default["default"].createElement(Button["default"], { onClick: this.handleEditorConfirm, level: "primary" }, __('confirm'))))));
    };
    FormulaPicker.defaultProps = {
        evalMode: true
    };
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", []),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "handleConfirm", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [Object]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "renderFormulaValue", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [String]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "handleInputChange", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [Object]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "handleInputGroupChange", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [String]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "handleEditorChange", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", []),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "handleEditorConfirm", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", []),
        tslib.__metadata("design:returntype", Promise)
    ], FormulaPicker.prototype, "handleClick", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [Object, Function]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "close", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [Object]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "updateState", null);
    tslib.__decorate([
        amisCore.autobind,
        tslib.__metadata("design:type", Function),
        tslib.__metadata("design:paramtypes", [String]),
        tslib.__metadata("design:returntype", void 0)
    ], FormulaPicker.prototype, "validate", null);
    return FormulaPicker;
}(React__default["default"].Component));
var Picker = amisCore.themeable(amisCore.localeable(amisCore.uncontrollable(FormulaPicker, {
    value: 'onChange'
})));

exports.FormulaPicker = FormulaPicker;
exports["default"] = Picker;
