// amis-core/lib/store/crud.js
/**
 * amis-core v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var fileSaver = require('file-saver');
var mobxStateTree = require('mobx-state-tree');
var service = require('./service.js');
var helper = require('../utils/helper.js');
var pick = require('lodash/pick');
require('amis-formula');
require('moment');
var object = require('../utils/object.js');
var getVariable = require('../utils/getVariable.js');
var resolveVariableAndFilter = require('../utils/resolveVariableAndFilter.js');
require('lodash/isPlainObject');
require('../utils/filter.js');
var api = require('../utils/api.js');
var matchSorter = require('match-sorter');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var pick__default = /*#__PURE__*/_interopDefaultLegacy(pick);

var ServerError = /** @class */ (function (_super) {
    tslib.__extends(ServerError, _super);
    function ServerError() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'ServerError';
        return _this;
    }
    return ServerError;
}(Error));
var CRUDStore = service.ServiceStore.named('CRUDStore')
    .props({
    pristineQuery: mobxStateTree.types.optional(mobxStateTree.types.frozen(), {}),
    query: mobxStateTree.types.optional(mobxStateTree.types.frozen(), {}),
    prevPage: 1,
    page: 1,
    perPage: 10,
    total: 0,
    mode: 'normal',
    hasNext: false,
    selectedAction: mobxStateTree.types.frozen(),
    columns: mobxStateTree.types.frozen(),
    items: mobxStateTree.types.optional(mobxStateTree.types.array(mobxStateTree.types.frozen()), []),
    selectedItems: mobxStateTree.types.optional(mobxStateTree.types.array(mobxStateTree.types.frozen()), []),
    unSelectedItems: mobxStateTree.types.optional(mobxStateTree.types.array(mobxStateTree.types.frozen()), []),
    filterTogggable: false,
    filterVisible: true,
    hasInnerModalOpen: false
})
    .views(function (self) { return ({
    get lastPage() {
        return Math.max(Math.ceil(self.total / (self.perPage < 1 ? 10 : self.perPage)), 1);
    },
    get filterData() {
        return object.createObject(self.data, tslib.__assign({}, self.query));
    },
    get mergedData() {
        return object.extendObject(self.data, tslib.__assign(tslib.__assign(tslib.__assign({}, self.query), self.data), { selectedItems: self.selectedItems, unSelectedItems: self.unSelectedItems }));
    },
    get hasModalOpened() {
        return self.dialogOpen || self.drawerOpen || self.hasInnerModalOpen;
    },
    get selectedItemsAsArray() {
        return self.selectedItems.concat();
    },
    fetchCtxOf: function (data, options) {
        var _a;
        return object.createObject(data, tslib.__assign(tslib.__assign(tslib.__assign({}, self.query), (_a = {}, _a[options.pageField || 'page'] = self.page, _a[options.perPageField || 'perPage'] = self.perPage, _a)), data));
    }
}); })
    .actions(function (self) {
    var fetchCancel = null;
    function setPristineQuery() {
        self.pristineQuery = self.query;
    }
    function updateQuery(values, updater, pageField, perPageField, replace) {
        if (pageField === void 0) { pageField = 'page'; }
        if (perPageField === void 0) { perPageField = 'perPage'; }
        if (replace === void 0) { replace = false; }
        var originQuery = self.query;
        self.query = replace
            ? tslib.__assign({}, values) : tslib.__assign(tslib.__assign({}, self.query), values);
        if (self.query[pageField || 'page']) {
            self.page = parseInt(self.query[pageField || 'page'], 10);
        }
        if (self.query[perPageField || 'perPage']) {
            self.perPage = parseInt(self.query[perPageField || 'perPage'], 10);
        }
        updater &&
            helper.isObjectShallowModified(originQuery, self.query, false) &&
            setTimeout(updater.bind(null, "?".concat(helper.qsstringify(self.query))), 4);
    }
    var fetchInitData = mobxStateTree.flow(function getInitData(api$1, data, options) {
        var items_1, dir, data_1, ctx, json, result, total, count, page, hasNext, oItems, oRows, columns, rest, items, _a, _b, key, rowsData, data_2, filteredItems_1, dir, e_1, env;
        var _c, e_2, _d;
        var _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
        if (options === void 0) { options = {}; }
        return tslib.__generator(this, function (_q) {
            switch (_q.label) {
                case 0:
                    _q.trys.push([0, 2, , 3]);
                    if (!options.forceReload && options.loadDataOnce && self.total) {
                        items_1 = options.source
                            ? resolveVariableAndFilter.resolveVariableAndFilter(options.source, object.createObject(self.mergedData, {
                                items: self.data.itemsRaw,
                                rows: self.data.itemsRaw
                            }), '| raw')
                            : self.items.concat();
                        if (Array.isArray(options.columns)) {
                            options.columns.forEach(function (column) {
                                var value;
                                var key = column.name;
                                if ((column.searchable || column.filterable) && key) {
                                    // value可能为null、undefined、''、0
                                    value = getVariable.getVariable(self.query, key);
                                    if (value != null) {
                                        items_1 = matchSorter.matchSorter(items_1, value, {
                                            keys: [key]
                                        });
                                    }
                                }
                            });
                        }
                        if (self.query.orderBy) {
                            dir = /desc/i.test(self.query.orderDir) ? -1 : 1;
                            items_1 = helper.sortArray(items_1, self.query.orderBy, dir);
                        }
                        data_1 = tslib.__assign(tslib.__assign({}, self.data), { total: items_1.length, items: items_1.slice((self.page - 1) * self.perPage, self.page * self.perPage) });
                        self.total = parseInt((_e = data_1.total) !== null && _e !== void 0 ? _e : data_1.count, 10) || 0;
                        self.reInitData(data_1);
                        return [2 /*return*/];
                    }
                    if (fetchCancel) {
                        fetchCancel();
                        fetchCancel = null;
                        self.fetching = false;
                    }
                    options.silent || self.markFetching(true);
                    ctx = object.createObject(self.data, tslib.__assign(tslib.__assign(tslib.__assign({}, self.query), data), (_c = {}, _c[options.pageField || 'page'] = self.page, _c[options.perPageField || 'perPage'] = self.perPage, _c)));
                    // 一次性加载不要发送 perPage 属性
                    if (options.loadDataOnce) {
                        delete ctx[options.perPageField || 'perPage'];
                    }
                    return [4 /*yield*/, mobxStateTree.getEnv(self).fetcher(api$1, ctx, tslib.__assign(tslib.__assign({}, options), { cancelExecutor: function (executor) { return (fetchCancel = executor); } }))];
                case 1:
                    json = _q.sent();
                    fetchCancel = null;
                    if (!json.ok) {
                        self.updateMessage((_j = (_h = (_g = (_f = api$1 === null || api$1 === void 0 ? void 0 : api$1.messages) === null || _f === void 0 ? void 0 : _f.failed) !== null && _g !== void 0 ? _g : json.msg) !== null && _h !== void 0 ? _h : options.errorMessage) !== null && _j !== void 0 ? _j : self.__('CRUD.fetchFailed'), true);
                        mobxStateTree.getEnv(self).notify('error', json.msg, json.msgTimeout !== undefined
                            ? {
                                closeButton: true,
                                timeout: json.msgTimeout
                            }
                            : undefined);
                    }
                    else {
                        if (!json.data) {
                            throw new Error(self.__('CRUD.invalidData'));
                        }
                        self.updatedAt = Date.now();
                        result = api.normalizeApiResponseData(json.data);
                        total = result.total, count = result.count, page = result.page, hasNext = result.hasNext, oItems = result.items, oRows = result.rows, columns = result.columns, rest = tslib.__rest(result, ["total", "count", "page", "hasNext", "items", "rows", "columns"]);
                        items = void 0;
                        if (options.source) {
                            items = resolveVariableAndFilter.resolveVariableAndFilter(options.source, object.createObject(self.filterData, result), '| raw');
                        }
                        else {
                            items = result.items || result.rows;
                        }
                        // 如果不按照 items 格式返回，就拿第一个数组当成 items
                        if (!Array.isArray(items)) {
                            try {
                                for (_a = tslib.__values(Object.keys(result)), _b = _a.next(); !_b.done; _b = _a.next()) {
                                    key = _b.value;
                                    if (result.hasOwnProperty(key) && Array.isArray(result[key])) {
                                        items = result[key];
                                        break;
                                    }
                                }
                            }
                            catch (e_2_1) { e_2 = { error: e_2_1 }; }
                            finally {
                                try {
                                    if (_b && !_b.done && (_d = _a.return)) _d.call(_a);
                                }
                                finally { if (e_2) throw e_2.error; }
                            }
                        }
                        if (!Array.isArray(items)) {
                            throw new Error(self.__('CRUD.invalidArray'));
                        }
                        else {
                            // 确保成员是对象。
                            items.map(function (item) {
                                return typeof item === 'string' ? { text: item } : item;
                            });
                        }
                        rowsData = [];
                        if (options.loadDataMode && Array.isArray(self.data.items)) {
                            rowsData = self.data.items.concat(items);
                        }
                        else {
                            // 第一次的时候就是直接加载请求的数据
                            rowsData = items;
                        }
                        data_2 = tslib.__assign(tslib.__assign(tslib.__assign({}, (api$1.replaceData ? {} : self.pristine)), { items: rowsData, count: count, total: total }), rest);
                        if (options.loadDataOnce) {
                            // 记录原始集合，后续可能基于原始数据做排序查找。
                            data_2.itemsRaw = oItems || oRows;
                            filteredItems_1 = rowsData.concat();
                            if (Array.isArray(options.columns)) {
                                options.columns.forEach(function (column) {
                                    var value;
                                    var key = column.name;
                                    if (column.searchable &&
                                        key &&
                                        (value = getVariable.getVariable(self.query, key))) {
                                        if (Array.isArray(value)) {
                                            if (value.length > 0) {
                                                var arr_1 = tslib.__spreadArray([], tslib.__read(filteredItems_1), false);
                                                var arrItems_1 = [];
                                                value.forEach(function (item) {
                                                    arrItems_1 = tslib.__spreadArray(tslib.__spreadArray([], tslib.__read(arrItems_1), false), tslib.__read(matchSorter.matchSorter(arr_1, item, {
                                                        keys: [key]
                                                    })), false);
                                                });
                                                filteredItems_1 = filteredItems_1.filter(function (item) { return arrItems_1.find(function (a) { return a === item; }); });
                                            }
                                        }
                                        else {
                                            filteredItems_1 = matchSorter.matchSorter(filteredItems_1, value, {
                                                keys: [key]
                                            });
                                        }
                                    }
                                });
                            }
                            if (self.query.orderBy) {
                                dir = /desc/i.test(self.query.orderDir) ? -1 : 1;
                                filteredItems_1 = helper.sortArray(filteredItems_1, self.query.orderBy, dir);
                            }
                            data_2.items = filteredItems_1.slice((self.page - 1) * self.perPage, self.page * self.perPage);
                            data_2.count = data_2.total = filteredItems_1.length;
                        }
                        if (Array.isArray(columns)) {
                            self.columns = columns.concat();
                        }
                        else if (rest.isTable2) {
                            self.columns = options.columns;
                        }
                        self.items.replace(rowsData);
                        self.reInitData(data_2, !!api$1.replaceData);
                        options.syncResponse2Query !== false &&
                            updateQuery(pick__default["default"](rest, Object.keys(self.query)), undefined, options.pageField || 'page', options.perPageField || 'perPage');
                        self.total = parseInt((_k = data_2.total) !== null && _k !== void 0 ? _k : data_2.count, 10) || 0;
                        typeof page !== 'undefined' && (self.page = parseInt(page, 10));
                        // 分页情况不清楚，只能知道有没有下一页。
                        if (typeof hasNext !== 'undefined') {
                            self.mode = 'simple';
                            self.total = 0;
                            self.hasNext = !!hasNext;
                        }
                        self.updateMessage((_p = (_o = (_m = (_l = api$1.messages) === null || _l === void 0 ? void 0 : _l.success) !== null && _m !== void 0 ? _m : json.msg) !== null && _o !== void 0 ? _o : options.successMessage) !== null && _p !== void 0 ? _p : json.defaultMsg);
                        // 配置了获取成功提示后提示，默认是空不会提示。
                        options &&
                            options.successMessage &&
                            mobxStateTree.getEnv(self).notify('success', self.msg);
                    }
                    self.markFetching(false);
                    return [2 /*return*/, json];
                case 2:
                    e_1 = _q.sent();
                    env = mobxStateTree.getEnv(self);
                    if (!mobxStateTree.isAlive(self) || self.disposed) {
                        return [2 /*return*/];
                    }
                    self.markFetching(false);
                    if (env.isCancel(e_1)) {
                        return [2 /*return*/];
                    }
                    console.error(e_1);
                    env.notify('error', e_1.message);
                    return [2 /*return*/];
                case 3: return [2 /*return*/];
            }
        });
    });
    function changePage(page, perPage) {
        self.page = page;
        perPage && changePerPage(perPage);
    }
    function changePerPage(perPage) {
        self.perPage = parseInt(perPage, 10);
    }
    function selectAction(action) {
        self.selectedAction = action;
    }
    var saveRemote = mobxStateTree.flow(function saveRemote(api$1, data, options) {
        var json, e_3;
        var _a, _b, _c, _d, _e, _f, _g, _h;
        if (options === void 0) { options = {}; }
        return tslib.__generator(this, function (_j) {
            switch (_j.label) {
                case 0:
                    _j.trys.push([0, 2, , 3]);
                    options = tslib.__assign({ method: 'post' }, options);
                    self.markSaving(true);
                    return [4 /*yield*/, mobxStateTree.getEnv(self).fetcher(api$1, data, options)];
                case 1:
                    json = _j.sent();
                    self.markSaving(false);
                    if (!helper.isEmpty(json.data) || json.ok) {
                        self.updateData(api.normalizeApiResponseData(json.data), {
                            __saved: Date.now()
                        }, !!api$1 && api$1.replaceData);
                        self.updatedAt = Date.now();
                    }
                    if (!json.ok) {
                        self.updateMessage((_d = (_c = (_b = (_a = api$1 === null || api$1 === void 0 ? void 0 : api$1.messages) === null || _a === void 0 ? void 0 : _a.failed) !== null && _b !== void 0 ? _b : json.msg) !== null && _c !== void 0 ? _c : options.errorMessage) !== null && _d !== void 0 ? _d : self.__('saveFailed'), true);
                        mobxStateTree.getEnv(self).notify('error', self.msg, json.msgTimeout !== undefined
                            ? {
                                closeButton: true,
                                timeout: json.msgTimeout
                            }
                            : undefined);
                        throw new ServerError(self.msg);
                    }
                    else {
                        self.updateMessage((_h = (_g = (_f = (_e = api$1 === null || api$1 === void 0 ? void 0 : api$1.messages) === null || _e === void 0 ? void 0 : _e.success) !== null && _f !== void 0 ? _f : json.msg) !== null && _g !== void 0 ? _g : options.successMessage) !== null && _h !== void 0 ? _h : json.defaultMsg);
                        self.msg &&
                            mobxStateTree.getEnv(self).notify('success', self.msg, json.msgTimeout !== undefined
                                ? {
                                    closeButton: true,
                                    timeout: json.msgTimeout
                                }
                                : undefined);
                    }
                    return [2 /*return*/, json.data];
                case 2:
                    e_3 = _j.sent();
                    self.markSaving(false);
                    if (!mobxStateTree.isAlive(self) || self.disposed) {
                        return [2 /*return*/];
                    }
                    e_3.type !== 'ServerError' && mobxStateTree.getEnv(self).notify('error', e_3.message);
                    throw e_3;
                case 3: return [2 /*return*/];
            }
        });
    });
    var setFilterTogglable = function (toggable, filterVisible) {
        self.filterTogggable = toggable;
        filterVisible !== void 0 && (self.filterVisible = filterVisible);
    };
    var setFilterVisible = function (visible) {
        self.filterVisible = visible;
    };
    var setSelectedItems = function (items) {
        self.selectedItems.replace(items);
    };
    var setUnSelectedItems = function (items) {
        self.unSelectedItems.replace(items);
    };
    var updateSelectData = function (selected, unSelected) {
        self.selectedItems.replace(selected);
        self.unSelectedItems.replace(unSelected);
        // 同步到data中，使filter等部分也能拿到
        self.reInitData({
            selectedItems: selected,
            unSelectedItems: unSelected
        });
    };
    var setInnerModalOpened = function (value) {
        self.hasInnerModalOpen = value;
    };
    var initFromScope = function (scope, source) {
        var rowsData = resolveVariableAndFilter.resolveVariableAndFilter(source, scope, '| raw');
        if (!Array.isArray(rowsData) && !self.items.length) {
            return;
        }
        rowsData = Array.isArray(rowsData) ? rowsData : [];
        var data = tslib.__assign(tslib.__assign({}, self.pristine), { items: rowsData, count: 0, total: rowsData.length });
        self.items.replace(rowsData);
        self.reInitData(data);
    };
    var exportAsCSV = function (options) {
        if (options === void 0) { options = {}; }
        return tslib.__awaiter(void 0, void 0, void 0, function () {
            var items, env, res;
            return tslib.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        items = options.loadDataOnce ? self.data.itemsRaw : self.data.items;
                        if (!options.api) return [3 /*break*/, 2];
                        env = mobxStateTree.getEnv(self);
                        return [4 /*yield*/, env.fetcher(options.api, options.data)];
                    case 1:
                        res = _a.sent();
                        if (!res.data) {
                            return [2 /*return*/];
                        }
                        if (Array.isArray(res.data)) {
                            items = res.data;
                        }
                        else {
                            items = res.data.rows || res.data.items;
                        }
                        _a.label = 2;
                    case 2:
                        Promise.resolve().then(function() {return new Promise(function(fullfill) {require(['papaparse'], function(mod) {fullfill(tslib.__importStar(mod))})})}).then(function (papaparse) {
                            var e_4, _a;
                            // 将数据里的对象转成 json 字符串，不然输出的 csv 没法显示
                            var csvData = [];
                            try {
                                for (var items_2 = tslib.__values(items), items_2_1 = items_2.next(); !items_2_1.done; items_2_1 = items_2.next()) {
                                    var row = items_2_1.value;
                                    var rowData = {};
                                    for (var key in row) {
                                        var value = row[key];
                                        if (typeof value === 'object') {
                                            rowData[key] = JSON.stringify(value);
                                        }
                                        else {
                                            rowData[key] = value;
                                        }
                                    }
                                    csvData.push(rowData);
                                }
                            }
                            catch (e_4_1) { e_4 = { error: e_4_1 }; }
                            finally {
                                try {
                                    if (items_2_1 && !items_2_1.done && (_a = items_2.return)) _a.call(items_2);
                                }
                                finally { if (e_4) throw e_4.error; }
                            }
                            var csvText = papaparse.unparse(csvData);
                            if (csvText) {
                                var blob = new Blob(
                                // 加上 BOM 这样 Excel 打开的时候就不会乱码
                                [new Uint8Array([0xef, 0xbb, 0xbf]), csvText], {
                                    type: 'text/plain;charset=utf-8'
                                });
                                fileSaver.saveAs(blob, 'data.csv');
                            }
                        });
                        return [2 /*return*/];
                }
            });
        });
    };
    var getData = function (superData) {
        return object.createObject(superData, {
            total: self.total,
            page: self.page,
            items: self.items.concat(),
            selectedItems: self.selectedItems.concat(),
            unSelectedItems: self.unSelectedItems.concat()
        });
    };
    var updateColumns = function (columns) {
        self.columns = columns;
    };
    return {
        getData: getData,
        updateSelectData: updateSelectData,
        setPristineQuery: setPristineQuery,
        updateQuery: updateQuery,
        fetchInitData: fetchInitData,
        changePage: changePage,
        changePerPage: changePerPage,
        selectAction: selectAction,
        saveRemote: saveRemote,
        setFilterTogglable: setFilterTogglable,
        setFilterVisible: setFilterVisible,
        setSelectedItems: setSelectedItems,
        setUnSelectedItems: setUnSelectedItems,
        setInnerModalOpened: setInnerModalOpened,
        initFromScope: initFromScope,
        exportAsCSV: exportAsCSV,
        updateColumns: updateColumns
    };
});

exports.CRUDStore = CRUDStore;
