// amis/esm/renderers/Form/JSONSchema.js
/**
 * amis v2.9.0
 * Copyright 2018-2023 baidu
 */

import { __extends, __rest, __assign, __decorate } from 'tslib';
import React from 'react';
import { FormItem } from 'amis-core';
import { withRemoteConfig, InputJSONSchema } from 'amis-ui';

var EnhancedInputJSONSchema = withRemoteConfig({
    sourceField: 'schema',
    injectedPropsFilter: function (injectedProps, props) {
        return {
            schema: injectedProps.config,
            loading: injectedProps.loading
        };
    }
})(InputJSONSchema);
var JSONSchemaControl = /** @class */ (function (_super) {
    __extends(JSONSchemaControl, _super);
    function JSONSchemaControl() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    JSONSchemaControl.prototype.render = function () {
        var rest = __rest(this.props, []);
        return React.createElement(EnhancedInputJSONSchema, __assign({}, rest));
    };
    return JSONSchemaControl;
}(React.PureComponent));
/** @class */ ((function (_super) {
    __extends(JSONSchemaRenderer, _super);
    function JSONSchemaRenderer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    JSONSchemaRenderer = __decorate([
        FormItem({
            type: 'json-schema',
            strictMode: false
        })
    ], JSONSchemaRenderer);
    return JSONSchemaRenderer;
})(JSONSchemaControl));

export { JSONSchemaControl as default };
