// office-viewer/esm/OpenXML.js
/**
 * openxml 一些辅助函数
 */
/**
 * 获取 w:val 的值
 */
function getVal(element) {
    return element.getAttribute('w:val') || '';
}
/**
 * 获取 w:val 的值，转成 number 类型
 */
function getValNumber(element) {
    return parseInt(getVal(element), 10);
}
/**
 * 有可能是 on 或 off 之类的值，都归一化为 boolean
 * @param value
 * @param defaultValue 默认值
 * @returns
 */
function normalizeBoolean(value, defaultValue) {
    if (defaultValue === void 0) { defaultValue = false; }
    if (typeof value === 'boolean') {
        return value;
    }
    if (typeof value === 'string') {
        switch (value) {
            case '1':
                return true;
            case '0':
                return false;
            case 'on':
                return true;
            case 'off':
                return false;
            case 'true':
                return true;
            case 'false':
                return false;
        }
        if (typeof value === 'number') {
            return value !== 0;
        }
    }
    return defaultValue;
}
function getValBoolean(element, defaultValue) {
    if (defaultValue === void 0) { defaultValue = true; }
    return normalizeBoolean(getVal(element), defaultValue);
}
function getAttrBoolean(element, attr, defaultValue) {
    if (defaultValue === void 0) { defaultValue = true; }
    return normalizeBoolean(element.getAttribute(attr), defaultValue);
}
function getValHex(element) {
    return parseInt(getVal(element) || '0', 16);
}

export { getAttrBoolean, getVal, getValBoolean, getValHex, getValNumber, normalizeBoolean };
