// amis-ui/esm/components/Alert2.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

import { __extends } from 'tslib';
import React__default from 'react';
import { generateIcon, themeable } from 'amis-core';
import { getIcon, Icon } from './icons.js';

/**
 * @file Alert2
 * @author fex
 */
var Alert = /** @class */ (function (_super) {
    __extends(Alert, _super);
    function Alert(props) {
        var _this = _super.call(this, props) || this;
        _this.handleClick = _this.handleClick.bind(_this);
        _this.state = {
            show: true
        };
        return _this;
    }
    Alert.prototype.handleClick = function () {
        this.setState({
            show: false
        }, this.props.onClose);
    };
    Alert.prototype.render = function () {
        var _a;
        var _b = this.props, cx = _b.classnames, className = _b.className, style = _b.style, level = _b.level, children = _b.children, showCloseButton = _b.showCloseButton, title = _b.title, icon = _b.icon, showIcon = _b.showIcon, iconClassName = _b.iconClassName, closeButtonClassName = _b.closeButtonClassName;
        // 优先使用内置svg，其次使用icon库
        var iconNode = icon ? (['string', 'object'].includes(typeof icon) ? (typeof icon === 'object' ? (generateIcon(cx, icon, 'icon')) : (getIcon(icon) && React__default.createElement(Icon, { icon: icon, className: cx("icon") }))) : React__default.isValidElement(icon) ? (React__default.cloneElement(icon, {
            className: cx("Alert-icon", (_a = icon.props) === null || _a === void 0 ? void 0 : _a.className)
        })) : null) : showIcon ? (React__default.createElement(Icon, { icon: "alert-".concat(level), className: cx("icon") })) : null;
        return this.state.show ? (React__default.createElement("div", { className: cx('Alert', level ? "Alert--".concat(level) : '', title ? 'Alert-has-title' : '', className), style: style },
            showIcon && iconNode ? (React__default.createElement("div", { className: cx('Alert-icon', iconClassName) }, iconNode)) : null,
            React__default.createElement("div", { className: cx('Alert-content') },
                title ? React__default.createElement("div", { className: cx('Alert-title') }, title) : null,
                React__default.createElement("div", { className: cx('Alert-desc') }, children)),
            showCloseButton ? (React__default.createElement("button", { className: cx('Alert-close', closeButtonClassName), onClick: this.handleClick, type: "button" },
                React__default.createElement(Icon, { icon: "close", className: "icon" }))) : null)) : null;
    };
    Alert.defaultProps = {
        level: 'info',
        className: '',
        showCloseButton: false
    };
    Alert.propsList = [
        'level',
        'className',
        'showCloseButton',
        'onClose'
    ];
    return Alert;
}(React__default.Component));
var Alert2 = themeable(Alert);

export { Alert, Alert2 as default };
