// amis-ui/esm/components/Cascader.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

import { __extends, __spreadArray, __assign, __read, __decorate, __metadata, __values } from 'tslib';
import React__default from 'react';
import { getTreeAncestors, findTree, flattenTree, getTreeDepth, autobind, themeable } from 'amis-core';
import intersectionBy from 'lodash/intersectionBy';
import compact from 'lodash/compact';
import find from 'lodash/find';
import uniqBy from 'lodash/uniqBy';
import Button from './Button.js';

/**
 * @file Cascader
 * @author fex
 */
var Cascader = /** @class */ (function (_super) {
    __extends(Cascader, _super);
    function Cascader(props) {
        var _this = _super.call(this, props) || this;
        _this.tabsRef = React__default.createRef();
        _this.tabRef = React__default.createRef();
        _this.getParentTree = function (option, arr) {
            var parentOption = _this.getOptionParent(option);
            if (parentOption) {
                arr.push(parentOption);
                return _this.getParentTree(parentOption, arr);
            }
            return arr;
        };
        _this.state = {
            selectedOptions: _this.props.selectedOptions || [],
            activeTab: 0,
            tabs: [
                {
                    options: _this.props.options.slice() || []
                }
            ],
            disableConfirm: false
        };
        return _this;
    }
    Cascader.prototype.componentDidMount = function () {
        var _a = this.props, multiple = _a.multiple, options = _a.options, _b = _a.valueField, valueField = _b === void 0 ? 'value' : _b, cascade = _a.cascade, onlyLeaf = _a.onlyLeaf;
        var selectedOptions = this.props.selectedOptions.slice();
        var parentsCount = 0;
        var parentTree = [];
        selectedOptions.forEach(function (item) {
            var parents = getTreeAncestors(options, item);
            // 获取最长路径
            if (parents && (parents === null || parents === void 0 ? void 0 : parents.length) > parentsCount) {
                parentTree = parents;
                parentsCount = parentTree.length;
            }
        });
        var selectedValues = selectedOptions.map(function (option) { return option[valueField]; });
        var tabs = parentTree.map(function (option) {
            var _a;
            if (multiple && !cascade) {
                if (selectedValues.includes(option[valueField]) &&
                    ((_a = option === null || option === void 0 ? void 0 : option.children) === null || _a === void 0 ? void 0 : _a.length)) {
                    option.children.forEach(function (option) { return (option.disabled = true); });
                }
            }
            return multiple && !onlyLeaf
                ? {
                    options: __spreadArray([
                        __assign(__assign({}, option), { isCheckAll: true })
                    ], __read((option.children ? option.children : [])), false)
                }
                : {
                    options: option.children ? option.children : []
                };
        });
        this.setState({
            selectedOptions: selectedOptions,
            tabs: __spreadArray(__spreadArray([], __read(this.state.tabs), false), __read(tabs), false)
        });
    };
    Cascader.prototype.handleTabSelect = function (index) {
        var tabs = this.state.tabs.slice(0, index + 1);
        this.setState({
            activeTab: index,
            tabs: tabs
        });
    };
    Cascader.prototype.getOptionParent = function (option) {
        var _a = this.props, options = _a.options, _b = _a.valueField, valueField = _b === void 0 ? 'value' : _b;
        var ancestors = [];
        findTree(options, function (item, index, level, paths) {
            if (item[valueField] === option[valueField]) {
                ancestors = paths;
                return true;
            }
            return false;
        });
        return ancestors.length ? ancestors[ancestors.length - 1] : null;
    };
    Cascader.prototype.dealParentSelect = function (option, selectedOptions) {
        var _a;
        var _b = this.props, _c = _b.valueField, valueField = _c === void 0 ? 'value' : _c, onlyLeaf = _b.onlyLeaf;
        var parentOption = this.getOptionParent(option);
        if (parentOption && !onlyLeaf) {
            var parentChildren = parentOption === null || parentOption === void 0 ? void 0 : parentOption.children;
            var equalOption = intersectionBy(selectedOptions, parentChildren, valueField);
            // 包含则选中父节点
            var isParentSelected = find(selectedOptions, (_a = {},
                _a[valueField] = parentOption[valueField],
                _a));
            if (equalOption.length === (parentChildren === null || parentChildren === void 0 ? void 0 : parentChildren.length) && !isParentSelected) {
                selectedOptions.push(parentOption);
            }
            if (equalOption.length !== (parentChildren === null || parentChildren === void 0 ? void 0 : parentChildren.length) && isParentSelected) {
                var index = selectedOptions.findIndex(function (item) { return item[valueField] === parentOption[valueField]; });
                selectedOptions.splice(index, 1);
            }
            return this.dealParentSelect(parentOption, selectedOptions);
        }
        else {
            return selectedOptions;
        }
    };
    Cascader.prototype.flattenTreeWithLeafNodes = function (option) {
        return compact(flattenTree(Array.isArray(option) ? option : [option], function (node) { return node; }));
    };
    Cascader.prototype.adjustOptionSelect = function (option) {
        var _a = this.props.valueField, valueField = _a === void 0 ? 'value' : _a;
        var selectedOptions = this.state.selectedOptions;
        function loop(arr) {
            if (!arr.length) {
                return false;
            }
            return arr.some(function (item) { return item[valueField] === option[valueField]; });
        }
        return loop(selectedOptions);
    };
    Cascader.prototype.getSelectedChildNum = function (option) {
        var _this = this;
        var count = 0;
        var loop = function (arr) {
            var e_1, _a;
            if (!arr || !arr.length) {
                return;
            }
            try {
                for (var arr_1 = __values(arr), arr_1_1 = arr_1.next(); !arr_1_1.done; arr_1_1 = arr_1.next()) {
                    var item = arr_1_1.value;
                    if (item.children) {
                        loop(item.children || []);
                    }
                    else {
                        if (_this.adjustOptionSelect(item)) {
                            count++;
                        }
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (arr_1_1 && !arr_1_1.done && (_a = arr_1.return)) _a.call(arr_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        };
        loop(option.children || []);
        return count;
    };
    Cascader.prototype.dealOptionDisable = function (selectedOptions) {
        var _a = this.props, _b = _a.valueField, valueField = _b === void 0 ? 'value' : _b, options = _a.options, cascade = _a.cascade, multiple = _a.multiple, onlyLeaf = _a.onlyLeaf, onlyChildren = _a.onlyChildren // 子节点可点击
        ;
        if (!multiple || cascade || onlyChildren || onlyLeaf) {
            return;
        }
        var selectedValues = selectedOptions.map(function (option) { return option[valueField]; });
        var loop = function (option) {
            if (!option.children) {
                return;
            }
            option.children &&
                option.children.forEach(function (childOption) {
                    if (!selectedValues.includes(option[valueField]) &&
                        !option.disabled) {
                        childOption.disabled = false;
                    }
                    if (selectedValues.includes(option[valueField]) || option.disabled) {
                        childOption.disabled = true;
                    }
                    loop(childOption);
                });
        };
        options.forEach(function (option) { return loop(option); });
    };
    Cascader.prototype.dealChildrenSelect = function (option, selectedOptions) {
        var _a;
        var _b = this.props, _c = _b.valueField, valueField = _c === void 0 ? 'value' : _c, onlyChildren = _b.onlyChildren;
        var index = selectedOptions.findIndex(function (item) { return item[valueField] === option[valueField]; });
        if (index !== -1) {
            selectedOptions.splice(index, 1);
        }
        else {
            if (!(onlyChildren && ((_a = option.children) === null || _a === void 0 ? void 0 : _a.length))) {
                selectedOptions.push(option);
            }
        }
        function loop(option) {
            if (!option.children) {
                return;
            }
            option.children.forEach(function (item) {
                var _a;
                if (index !== -1) {
                    // 删除选中节点及其子节点
                    selectedOptions = selectedOptions.filter(function (sItem) { return sItem[valueField] !== item[valueField]; });
                }
                else {
                    // 添加节点及其子节点
                    if (!(onlyChildren && ((_a = item.children) === null || _a === void 0 ? void 0 : _a.length))) {
                        selectedOptions.push(item);
                    }
                }
                loop(item);
            });
        }
        loop(option);
        return selectedOptions;
    };
    Cascader.prototype.onSelect = function (option, tabIndex) {
        var _this = this;
        var _a = this.props, multiple = _a.multiple, _b = _a.valueField, valueField = _b === void 0 ? 'value' : _b, cascade = _a.cascade, onlyLeaf = _a.onlyLeaf, onlyChildren = _a.onlyChildren;
        var tabs = this.state.tabs.slice();
        var activeTab = this.state.activeTab;
        var selectedOptions = this.state.selectedOptions;
        var isDisable = option.disabled;
        if (!isDisable) {
            if (multiple) {
                // 父子级分离
                if (cascade) {
                    if (option.isCheckAll ||
                        !option.children ||
                        !option.children.length) {
                        var index = selectedOptions.findIndex(function (item) { return item[valueField] === option[valueField]; });
                        if (index !== -1) {
                            selectedOptions.splice(index, 1);
                        }
                        else {
                            selectedOptions.push(option);
                        }
                    }
                }
                else {
                    if (option.isCheckAll ||
                        !option.children ||
                        !option.children.length) {
                        selectedOptions = this.dealChildrenSelect(option, selectedOptions);
                        if (!onlyChildren) {
                            selectedOptions = this.dealParentSelect(option, selectedOptions);
                        }
                    }
                }
            }
            else {
                // 单选
                selectedOptions = [option];
            }
        }
        this.dealOptionDisable(selectedOptions);
        if (tabs.length > tabIndex + 1) {
            tabs = tabs.slice(0, tabIndex + 1);
        }
        requestAnimationFrame(function () {
            var _a, _b;
            var tabWidth = ((_a = _this.tabRef.current) === null || _a === void 0 ? void 0 : _a.offsetWidth) || 1;
            var parentTree = _this.getParentTree(option, [option]);
            var scrollLeft = (parentTree.length - 2) * tabWidth;
            if (scrollLeft !== 0) {
                (_b = _this.tabsRef.current) === null || _b === void 0 ? void 0 : _b.scrollTo(scrollLeft, 0);
            }
        });
        if ((option === null || option === void 0 ? void 0 : option.children) && !option.isCheckAll) {
            var nextTab = multiple && !onlyLeaf
                ? {
                    options: __spreadArray([
                        __assign(__assign({}, option), { isCheckAll: true })
                    ], __read(option.children), false)
                }
                : {
                    options: option.children
                };
            if (tabs[tabIndex + 1]) {
                tabs[tabIndex + 1] = nextTab;
            }
            else {
                tabs.push(nextTab);
            }
            activeTab += 1;
        }
        var disableConfirm = false;
        if (onlyLeaf && selectedOptions.length && selectedOptions[0].children) {
            disableConfirm = true;
        }
        this.setState({
            tabs: tabs,
            activeTab: activeTab,
            selectedOptions: selectedOptions,
            disableConfirm: disableConfirm
        });
    };
    Cascader.prototype.onNextClick = function (option, tabIndex) {
        var activeTab = this.state.activeTab;
        var tabs = this.state.tabs.slice();
        if (option.c)
            if (option === null || option === void 0 ? void 0 : option.children) {
                var nextTab = {
                    options: option.children
                };
                if (tabs[tabIndex + 1]) {
                    tabs[tabIndex + 1] = nextTab;
                }
                else {
                    tabs.push(nextTab);
                }
                activeTab += 1;
            }
        this.setState({
            tabs: tabs,
            activeTab: activeTab
        });
    };
    Cascader.prototype.getSubmitOptions = function (selectedOptions) {
        var _selectedOptions = [];
        var _a = this.props, multiple = _a.multiple, options = _a.options, _b = _a.valueField, valueField = _b === void 0 ? 'value' : _b, cascade = _a.cascade, onlyChildren = _a.onlyChildren, withChildren = _a.withChildren;
        if (cascade || onlyChildren || withChildren || !multiple) {
            return selectedOptions;
        }
        var selectedValues = selectedOptions.map(function (option) { return option[valueField]; });
        function loop(options) {
            if (!options || !options.length) {
                return;
            }
            options.forEach(function (option) {
                if (selectedValues.includes(option[valueField])) {
                    _selectedOptions.push(option);
                }
                else {
                    loop(option.children ? option.children : []);
                }
            });
        }
        loop(options);
        return _selectedOptions;
    };
    Cascader.prototype.confirm = function () {
        var _a = this.props, onChange = _a.onChange; _a.joinValues; _a.delimiter; _a.extractValue; _a.valueField; var onClose = _a.onClose, onlyLeaf = _a.onlyLeaf;
        var selectedOptions = this.getSelectedOptions();
        if (onlyLeaf && selectedOptions.length && selectedOptions[0].children) {
            return;
        }
        onChange(selectedOptions);
        onClose && onClose();
    };
    Cascader.prototype.getSelectedOptions = function () {
        return uniqBy(this.getSubmitOptions(this.state.selectedOptions), this.props.valueField);
    };
    Cascader.prototype.renderOption = function (option, tabIndex) {
        var _this = this;
        var _a = this.props, activeColor = _a.activeColor, optionRender = _a.optionRender, labelField = _a.labelField, _b = _a.valueField, valueField = _b === void 0 ? 'value' : _b, cx = _a.classnames; _a.cascade; _a.multiple;
        var selectedOptions = this.state.selectedOptions;
        var selectedValueArr = selectedOptions.map(function (item) { return item[valueField]; });
        var selfChecked = selectedValueArr.includes(option[valueField]);
        var color = option.color || (selfChecked ? activeColor : undefined);
        var Text = optionRender ? (optionRender({ option: option, selected: selfChecked })) : (React__default.createElement("span", null, option[labelField]));
        return (React__default.createElement("li", { className: cx('Cascader-option', {
                selected: selfChecked,
                disabled: option.disabled
            }, option.className), style: { color: color }, onClick: function () { return _this.onSelect(option, tabIndex); }, key: tabIndex + '-' + option[valueField] },
            React__default.createElement("span", { className: cx('Cascader-option--text') }, Text)));
    };
    Cascader.prototype.renderOptions = function (options, tabIndex) {
        var _this = this;
        var cx = this.props.classnames;
        return (React__default.createElement("ul", { key: tabIndex, className: cx('Cascader-options') }, options.map(function (option) { return _this.renderOption(option, tabIndex); })));
    };
    Cascader.prototype.renderTabs = function () {
        var _this = this;
        var _a = this.props, cx = _a.classnames, options = _a.options;
        var tabs = this.state.tabs;
        var depth = getTreeDepth(options);
        return (React__default.createElement("div", { className: cx("Cascader-tabs", depth > 3 ? 'scrollable' : ''), ref: this.tabsRef },
            tabs.map(function (tab, tabIndex) {
                var options = tab.options;
                return (React__default.createElement("div", { className: cx("Cascader-tab"), ref: _this.tabRef, key: tabIndex }, _this.renderOptions(options, tabIndex)));
            }),
            depth <= 3 && options.length
                ? Array(getTreeDepth(options) - tabs.length)
                    .fill(1)
                    .map(function (item, index) { return (React__default.createElement("div", { className: cx("Cascader-tab"), key: index })); })
                : null));
    };
    Cascader.prototype.render = function () {
        var _a = this.props; _a.classPrefix; var cx = _a.classnames, className = _a.className, onClose = _a.onClose; _a.valueField; var __ = _a.translate;
        return (React__default.createElement("div", { className: cx("Cascader", className) },
            React__default.createElement("div", { className: cx("Cascader-btnGroup") },
                React__default.createElement(Button, { className: cx("Cascader-btnCancel"), level: "text", onClick: onClose }, __('cancel')),
                React__default.createElement(Button, { className: cx("Cascader-btnConfirm"), level: "text", onClick: this.confirm, disabled: this.state.disableConfirm }, __('confirm'))),
            this.renderTabs()));
    };
    Cascader.defaultProps = {
        labelField: 'label',
        valueField: 'value'
    };
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Number]),
        __metadata("design:returntype", void 0)
    ], Cascader.prototype, "handleTabSelect", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], Cascader.prototype, "getOptionParent", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Object]),
        __metadata("design:returntype", Object)
    ], Cascader.prototype, "dealParentSelect", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], Cascader.prototype, "flattenTreeWithLeafNodes", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Boolean)
    ], Cascader.prototype, "adjustOptionSelect", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Number)
    ], Cascader.prototype, "getSelectedChildNum", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], Cascader.prototype, "dealOptionDisable", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Object]),
        __metadata("design:returntype", void 0)
    ], Cascader.prototype, "dealChildrenSelect", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Number]),
        __metadata("design:returntype", void 0)
    ], Cascader.prototype, "onSelect", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Number]),
        __metadata("design:returntype", void 0)
    ], Cascader.prototype, "onNextClick", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Object)
    ], Cascader.prototype, "getSubmitOptions", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], Cascader.prototype, "confirm", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], Cascader.prototype, "getSelectedOptions", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Number]),
        __metadata("design:returntype", void 0)
    ], Cascader.prototype, "renderOption", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Number]),
        __metadata("design:returntype", void 0)
    ], Cascader.prototype, "renderOptions", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], Cascader.prototype, "renderTabs", null);
    return Cascader;
}(React__default.Component));
var Cascader$1 = themeable(Cascader);

export { Cascader, Cascader$1 as default };
