// amis-ui/esm/components/Checkbox.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

import { __extends, __decorate, __metadata } from 'tslib';
import React__default from 'react';
import { autobind, themeable } from 'amis-core';

/**
 * @file Checkbox
 * @author fex
 */
var preventEvent = function (e) { return e.stopPropagation(); };
var Checkbox = /** @class */ (function (_super) {
    __extends(Checkbox, _super);
    function Checkbox() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Checkbox.prototype.handleCheck = function (e) {
        var _a = this.props, trueValue = _a.trueValue, falseValue = _a.falseValue, onChange = _a.onChange;
        if (!onChange) {
            return;
        }
        onChange(e.currentTarget.checked ? trueValue : falseValue, e.nativeEvent.shiftKey);
    };
    Checkbox.prototype.render = function () {
        var _a;
        var _b = this.props, size = _b.size, className = _b.className, cx = _b.classnames, value = _b.value, label = _b.label, partial = _b.partial, trueValue = _b.trueValue, children = _b.children, disabled = _b.disabled, description = _b.description, readOnly = _b.readOnly, checked = _b.checked, type = _b.type, name = _b.name, labelClassName = _b.labelClassName, optionType = _b.optionType;
        var _checked = typeof checked !== 'undefined'
            ? checked
            : typeof value === 'undefined'
                ? value
                : value == trueValue;
        return (React__default.createElement("label", { className: cx("Checkbox Checkbox--".concat(type), className, (_a = {
                    'Checkbox--full': !partial
                },
                // 'Checkbox--partial': partial
                _a["Checkbox--".concat(size)] = size,
                _a['Checkbox--button'] = optionType === 'button',
                _a['Checkbox--button--checked'] = optionType === 'button' && _checked,
                _a['Checkbox--button--disabled--unchecked'] = optionType === 'button' && disabled && !_checked,
                _a['Checkbox--button--disabled--checked'] = optionType === 'button' && disabled && _checked,
                _a)) },
            React__default.createElement("input", { type: type, checked: typeof checked !== 'undefined'
                    ? checked
                    : typeof value === 'undefined'
                        ? value
                        : value == trueValue, onChange: this.handleCheck, onClick: preventEvent // 当点击 i 的时候，这个地方也会触发 click，很奇怪，干脆禁掉
                , disabled: disabled, readOnly: readOnly, name: name }),
            React__default.createElement("i", null),
            React__default.createElement("span", { className: cx(labelClassName) }, children || label),
            description ? (React__default.createElement("div", { className: cx('Checkbox-desc') }, description)) : null));
    };
    Checkbox.defaultProps = {
        trueValue: true,
        falseValue: false,
        type: 'checkbox'
    };
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], Checkbox.prototype, "handleCheck", null);
    return Checkbox;
}(React__default.Component));
var Checkbox$1 = themeable(Checkbox);

export { Checkbox, Checkbox$1 as default };
