// amis-ui/esm/components/CityArea.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

import { __read, __assign } from 'tslib';
import React__default, { memo, useState, useEffect } from 'react';
import Picker from './Picker.js';
import ResultBox from './ResultBox.js';
import useSetState from '../hooks/use-set-state.js';
import useUpdateEffect from '../hooks/use-update-effect.js';
import 'lodash/pick';
import { themeable, localeable, uncontrollable } from 'amis-core';
import PopUp from './PopUp.js';

var CityArea = memo(function (props) {
    var _a;
    var _b = props.joinValues, joinValues = _b === void 0 ? true : _b, _c = props.extractValue, extractValue = _c === void 0 ? true : _c, _d = props.delimiter, delimiter = _d === void 0 ? ',' : _d, _e = props.allowCity, allowCity = _e === void 0 ? true : _e, _f = props.allowDistrict, allowDistrict = _f === void 0 ? true : _f, _g = props.allowStreet, allowStreet = _g === void 0 ? false : _g, 
    // 默认北京东城区
    _h = props.value, 
    // 默认北京东城区
    value = _h === void 0 ? 110101 : _h, cx = props.classnames, style = props.style, __ = props.translate, _j = props.disabled, disabled = _j === void 0 ? false : _j, popOverContainer = props.popOverContainer, useMobileUI = props.useMobileUI;
    var _k = __read(useState([]), 2), values = _k[0], setValues = _k[1];
    var _l = __read(useState(''), 2), street = _l[0], setStreet = _l[1];
    var _m = __read(useState(), 2), confirmValues = _m[0], setConfirmValues = _m[1];
    var _o = __read(useSetState(), 2), db = _o[0], updateDb = _o[1];
    var _p = __read(useSetState({
        columns: []
    }), 2), state = _p[0], updateState = _p[1];
    var _q = __read(useState(false), 2), isOpened = _q[0], setIsOpened = _q[1];
    var onChange = function (columnValues, columnIndex) {
        var _a, _b, _c, _d, _e;
        // 清空后面的值
        while (columnValues[columnIndex++]) {
            columnValues[columnIndex++] = -1;
        }
        var _f = __read(columnValues, 3), provience = _f[0], city = _f[1], district = _f[2];
        if (city === -1) {
            city = (_b = (_a = db.city) === null || _a === void 0 ? void 0 : _a[provience]) === null || _b === void 0 ? void 0 : _b[0];
        }
        if (district === -1) {
            district = (_e = (_d = (_c = db.district) === null || _c === void 0 ? void 0 : _c[provience]) === null || _d === void 0 ? void 0 : _d[city]) === null || _e === void 0 ? void 0 : _e[0];
        }
        var tempValues = [provience, city, district];
        if (!allowDistrict) {
            tempValues.splice(2, 1);
        }
        if (!allowCity) {
            tempValues.splice(1, 1);
        }
        setValues(tempValues);
    };
    var propsChange = function () {
        var onChange = props.onChange;
        var _a = __read(values, 3), province = _a[0], city = _a[1], district = _a[2];
        var code = allowDistrict && district
            ? district
            : allowCity && city
                ? city
                : province;
        if (typeof extractValue === 'undefined' ? joinValues : extractValue) {
            code
                ? onChange(allowStreet && street
                    ? [code, street].join(delimiter)
                    : String(code))
                : onChange('');
        }
        else {
            onChange({
                code: code,
                provinceCode: province,
                province: db[province],
                cityCode: city,
                city: db[city],
                districtCode: district,
                district: db[district],
                street: street
            });
        }
    };
    var onConfirm = function () {
        var confirmValues = values.map(function (item) { return ({
            text: db[item],
            value: item
        }); });
        setConfirmValues(confirmValues);
        propsChange();
        setIsOpened(false);
    };
    var onCancel = function () {
        setIsOpened(false);
        if (props.onCancel)
            props.onCancel();
    };
    var getPropsValue = function () {
        var _a;
        // 最后一项的值、默认北京东城区
        var code = (value && value.code) ||
            (typeof value === 'number' && value) ||
            (typeof value === 'string' && /(\d{6})/.test(value) && RegExp.$1) ||
            110101;
        var values = [];
        if (code && db[code]) {
            code = parseInt(code, 10);
            var provinceCode = code - (code % 10000);
            var cityCode = code - (code % 100);
            if (db[provinceCode]) {
                values[0] = provinceCode;
            }
            if (db[cityCode] && allowCity) {
                values[1] = cityCode;
            }
            else if (~((_a = db.city[provinceCode]) === null || _a === void 0 ? void 0 : _a.indexOf(code)) && allowCity) {
                values[1] = code;
            }
            if (code % 100 && allowDistrict) {
                values[2] = code;
            }
            setValues(values);
            if (props.value) {
                var confirmValues_1 = values.map(function (item) { return ({
                    text: db[item],
                    value: item
                }); });
                setConfirmValues(confirmValues_1);
            }
        }
    };
    var updateColumns = function () {
        if (!db) {
            return;
        }
        var _a = __read(values, 3), provience = _a[0], city = _a[1], district = _a[2];
        var provienceColumn = db.province.map(function (code) {
            return { text: db[code], value: code, disabled: disabled };
        });
        var cityColumn = city
            ? db.city[provience].map(function (code) {
                return { text: db[code], value: code, disabled: disabled };
            })
            : [];
        var districtColumn = city && district
            ? db.district[provience][city].map(function (code) {
                return { text: db[code], value: code, disabled: disabled };
            })
            : [];
        var columns = [
            { options: provienceColumn },
            { options: cityColumn },
            { options: districtColumn }
        ];
        if (!allowDistrict || !allowCity) {
            columns.splice(2, 1);
        }
        if (!allowCity) {
            columns.splice(1, 1);
        }
        updateState({ columns: columns });
    };
    var loadDb = function () {
        import('./CityDB.js').then(function (db) {
            updateDb(__assign(__assign({}, db.default), { province: db.province, city: db.city, district: db.district }));
        });
    };
    useEffect(function () {
        loadDb();
    }, []);
    useEffect(function () {
        db && (props.value || isOpened) && getPropsValue();
    }, [db, isOpened, props.value]);
    useEffect(function () {
        street && propsChange();
    }, [street]);
    useUpdateEffect(function () {
        values.length && updateColumns();
    }, [values]);
    var result = (_a = confirmValues === null || confirmValues === void 0 ? void 0 : confirmValues.filter(function (item) { return item === null || item === void 0 ? void 0 : item.value; })) === null || _a === void 0 ? void 0 : _a.map(function (item) { return item.text; }).join(delimiter);
    return (React__default.createElement("div", { className: cx("CityArea"), style: style },
        React__default.createElement(ResultBox, { className: cx('CityArea-Input', isOpened ? 'is-active' : ''), allowInput: false, result: result, onResultChange: function () { }, onResultClick: function () { return setIsOpened(!isOpened); }, placeholder: __('Condition.cond_placeholder'), useMobileUI: useMobileUI }),
        allowStreet && values[0] ? (React__default.createElement("input", { className: cx('CityArea-Input'), value: street, onChange: function (e) {
                return setStreet(e.currentTarget.value);
            }, placeholder: __('City.street'), disabled: disabled })) : null,
        React__default.createElement(PopUp, { className: cx("CityArea-popup"), container: popOverContainer, isShow: isOpened, showConfirm: true, onConfirm: onConfirm, onHide: onCancel },
            React__default.createElement(Picker, { className: 'CityArea-picker', columns: state.columns, onChange: onChange, showToolbar: false, labelField: "text", itemHeight: 40, value: values, classnames: props.classnames, classPrefix: props.classPrefix }))));
});
var CityArea$1 = themeable(localeable(uncontrollable(CityArea, {
    value: 'onChange'
})));

export { CityArea$1 as default };
