// amis-ui/esm/components/DatePicker.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

import { __extends, __spreadArray, __read, __assign } from 'tslib';
import React__default from 'react';
import moment from 'moment';
import 'moment/locale/zh-cn';
import { Icon } from './icons.js';
import { isMobile, ucFirst, Overlay, PopOver, themeable, localeable } from 'amis-core';
import PopUp from './PopUp.js';
import Calendar from './calendar/Calendar.js';
import CalendarMobile from './CalendarMobile.js';
import Input from './Input.js';

/**
 * @file DatePicker
 * @description 时间选择器组件
 * @author fex
 */
var availableShortcuts = {
    now: {
        label: 'Date.now',
        date: function (now) {
            return now;
        }
    },
    today: {
        label: 'Date.today',
        date: function (now) {
            return now.startOf('day');
        }
    },
    yesterday: {
        label: 'Date.yesterday',
        date: function (now) {
            return now.add(-1, 'days').startOf('day');
        }
    },
    thisweek: {
        label: 'Date.monday',
        date: function (now) {
            return now.startOf('week').startOf('day');
        }
    },
    thismonth: {
        label: 'Date.startOfMonth',
        date: function (now) {
            return now.startOf('month');
        }
    },
    prevmonth: {
        label: 'Date.startOfLastMonth',
        date: function (now) {
            return now.startOf('month').add(-1, 'month');
        }
    },
    prevquarter: {
        label: 'Date.startOfLastQuarter',
        date: function (now) {
            return now.startOf('quarter').add(-1, 'quarter');
        }
    },
    thisquarter: {
        label: 'Date.startOfQuarter',
        date: function (now) {
            return now.startOf('quarter');
        }
    },
    tomorrow: {
        label: 'Date.tomorrow',
        date: function (now) {
            return now.add(1, 'days').startOf('day');
        }
    },
    endofthisweek: {
        label: 'Date.endOfWeek',
        date: function (now) {
            return now.endOf('week');
        }
    },
    endofthismonth: {
        label: 'Date.endOfMonth',
        date: function (now) {
            return now.endOf('month');
        }
    },
    endoflastmonth: {
        label: 'Date.endOfLastMonth',
        date: function (now) {
            return now.add(-1, 'month').endOf('month');
        }
    }
};
var advancedShortcuts = [
    {
        regexp: /^(\d+)hoursago$/,
        resolve: function (__, _, hours) {
            return {
                label: __('Date.hoursago', { hours: hours }),
                date: function (now) {
                    return now.subtract(hours, 'hours');
                }
            };
        }
    },
    {
        regexp: /^(\d+)hourslater$/,
        resolve: function (__, _, hours) {
            return {
                label: __('Date.hourslater', { hours: hours }),
                date: function (now) {
                    return now.add(hours, 'hours');
                }
            };
        }
    },
    {
        regexp: /^(\d+)daysago$/,
        resolve: function (__, _, days) {
            return {
                label: __('Date.daysago', { days: days }),
                date: function (now) {
                    return now.subtract(days, 'days');
                }
            };
        }
    },
    {
        regexp: /^(\d+)dayslater$/,
        resolve: function (__, _, days) {
            return {
                label: __('Date.dayslater', { days: days }),
                date: function (now) {
                    return now.add(days, 'days');
                }
            };
        }
    },
    {
        regexp: /^(\d+)weeksago$/,
        resolve: function (__, _, weeks) {
            return {
                label: __('Date.weeksago', { weeks: weeks }),
                date: function (now) {
                    return now.subtract(weeks, 'weeks');
                }
            };
        }
    },
    {
        regexp: /^(\d+)weekslater$/,
        resolve: function (__, _, weeks) {
            return {
                label: __('Date.weekslater', { weeks: weeks }),
                date: function (now) {
                    return now.add(weeks, 'weeks');
                }
            };
        }
    },
    {
        regexp: /^(\d+)monthsago$/,
        resolve: function (__, _, months) {
            return {
                label: __('Date.monthsago', { months: months }),
                date: function (now) {
                    return now.subtract(months, 'months');
                }
            };
        }
    },
    {
        regexp: /^(\d+)monthslater$/,
        resolve: function (__, _, months) {
            return {
                label: __('Date.monthslater', { months: months }),
                date: function (now) {
                    return now.add(months, 'months');
                }
            };
        }
    },
    {
        regexp: /^(\d+)quartersago$/,
        resolve: function (__, _, quarters) {
            return {
                label: __('Date.quartersago', { quarters: quarters }),
                date: function (now) {
                    return now.subtract(quarters, 'quarters');
                }
            };
        }
    },
    {
        regexp: /^(\d+)quarterslater$/,
        resolve: function (__, _, quarters) {
            return {
                label: __('Date.quarterslater', { quarters: quarters }),
                date: function (now) {
                    return now.add(quarters, 'quarters');
                }
            };
        }
    }
];
function normalizeValue(value, format) {
    if (!value || value === '0') {
        return undefined;
    }
    var v = moment(value, format, true);
    return v.isValid() ? v : undefined;
}
var DatePicker = /** @class */ (function (_super) {
    __extends(DatePicker, _super);
    function DatePicker(props) {
        var _this = this;
        var _a;
        _this = _super.call(this, props) || this;
        _this.state = {
            isOpened: false,
            isFocused: false,
            value: normalizeValue(_this.props.value, _this.props.format),
            inputValue: ((_a = normalizeValue(_this.props.value, _this.props.format)) === null || _a === void 0 ? void 0 : _a.format(_this.props.inputFormat)) || ''
        };
        _this.domRef = function (ref) {
            _this.dom = ref;
        };
        _this.inputRef = React__default.createRef();
        _this.handleChange = _this.handleChange.bind(_this);
        _this.selectRannge = _this.selectRannge.bind(_this);
        _this.checkIsValidDate = _this.checkIsValidDate.bind(_this);
        _this.open = _this.open.bind(_this);
        _this.close = _this.close.bind(_this);
        _this.handleFocus = _this.handleFocus.bind(_this);
        _this.handleBlur = _this.handleBlur.bind(_this);
        _this.clearValue = _this.clearValue.bind(_this);
        _this.handleClick = _this.handleClick.bind(_this);
        _this.handleKeyPress = _this.handleKeyPress.bind(_this);
        _this.getParent = _this.getParent.bind(_this);
        _this.getTarget = _this.getTarget.bind(_this);
        _this.handlePopOverClick = _this.handlePopOverClick.bind(_this);
        _this.renderShortCuts = _this.renderShortCuts.bind(_this);
        _this.inputChange = _this.inputChange.bind(_this);
        _this.onInputBlur = _this.onInputBlur.bind(_this);
        return _this;
    }
    DatePicker.prototype.componentDidMount = function () {
        var _a, _b;
        (_b = (_a = this.props) === null || _a === void 0 ? void 0 : _a.onRef) === null || _b === void 0 ? void 0 : _b.call(_a, this);
        var _c = this.props, value = _c.value, format = _c.format, inputFormat = _c.inputFormat;
        if (value) {
            var valueCache = normalizeValue(value, format);
            this.inputValueCache = (valueCache === null || valueCache === void 0 ? void 0 : valueCache.format(inputFormat)) || '';
        }
    };
    DatePicker.prototype.componentDidUpdate = function (prevProps) {
        var _a;
        var props = this.props;
        var prevValue = prevProps.value;
        if (prevValue !== props.value) {
            var newState = {
                value: normalizeValue(props.value, props.format)
            };
            newState.inputValue =
                ((_a = newState.value) === null || _a === void 0 ? void 0 : _a.format(this.props.inputFormat)) || '';
            this.inputValueCache = newState.inputValue;
            this.setState(newState);
        }
    };
    DatePicker.prototype.focus = function () {
        if (!this.dom) {
            return;
        }
        this.dom.focus();
    };
    DatePicker.prototype.handleFocus = function (e) {
        this.setState({
            isFocused: true
        });
        var onFocus = this.props.onFocus;
        onFocus && onFocus(e);
    };
    DatePicker.prototype.handleBlur = function (e) {
        this.setState({
            isFocused: false
        });
        var onBlur = this.props.onBlur;
        onBlur && onBlur(e);
    };
    DatePicker.prototype.handleKeyPress = function (e) {
        if (e.key === ' ') {
            this.handleClick();
            e.preventDefault();
        }
    };
    DatePicker.prototype.handleClick = function () {
        this.state.isOpened ? this.close() : this.open();
    };
    DatePicker.prototype.handlePopOverClick = function (e) {
        e.stopPropagation();
        e.preventDefault();
    };
    DatePicker.prototype.open = function (fn) {
        if (this.props.disabled) {
            return;
        }
        this.setState({
            isOpened: true
        }, fn);
        var input = this.inputRef.current;
        input && input.focus();
    };
    DatePicker.prototype.close = function () {
        this.setState({
            isOpened: false
        });
    };
    DatePicker.prototype.clearValue = function (e) {
        e.preventDefault();
        e.stopPropagation();
        var onChange = this.props.onChange;
        onChange('');
        this.setState({ inputValue: '' });
    };
    // 清空
    DatePicker.prototype.clear = function () {
        var onChange = this.props.onChange;
        onChange('');
        this.setState({ inputValue: '' });
    };
    // 重置
    DatePicker.prototype.reset = function (resetValue) {
        var _a;
        if (!resetValue) {
            return;
        }
        var _b = this.props, format = _b.format, inputFormat = _b.inputFormat, onChange = _b.onChange;
        onChange(resetValue);
        this.setState({
            inputValue: (_a = normalizeValue(resetValue, format)) === null || _a === void 0 ? void 0 : _a.format(inputFormat || '')
        });
    };
    DatePicker.prototype.handleChange = function (value) {
        var _a = this.props, onChange = _a.onChange, format = _a.format, minDate = _a.minDate, maxDate = _a.maxDate, dateFormat = _a.dateFormat, inputFormat = _a.inputFormat, timeFormat = _a.timeFormat, closeOnSelect = _a.closeOnSelect, utc = _a.utc; _a.viewMode;
        if (!moment.isMoment(value)) {
            return;
        }
        if (minDate && value && value.isBefore(minDate, 'second')) {
            value = minDate;
        }
        else if (maxDate && value && value.isAfter(maxDate, 'second')) {
            value = maxDate;
        }
        onChange(utc ? moment.utc(value).format(format) : value.format(format));
        if (closeOnSelect && dateFormat && !timeFormat) {
            this.close();
        }
        this.setState({
            inputValue: utc
                ? moment.utc(value).format(inputFormat)
                : value.format(inputFormat)
        });
    };
    // 手动输入日期
    DatePicker.prototype.inputChange = function (e) {
        var _a = this.props, onChange = _a.onChange, inputFormat = _a.inputFormat, format = _a.format, utc = _a.utc, minDate = _a.minDate, maxDate = _a.maxDate;
        var value = e.currentTarget.value;
        this.setState({ inputValue: value });
        if (value === '') {
            onChange('');
        }
        else {
            // 将输入的格式转成正则匹配，比如 YYYY-MM-DD HH:mm:ss 改成 \d\d\d\d\-
            // 只有匹配成功才更新
            var inputCheckRegex = new RegExp(inputFormat.replace(/[ymdhs]/gi, '\\d').replace(/-/gi, '\\-'));
            if (inputCheckRegex.test(value)) {
                var newDate = moment(value, inputFormat);
                var dateValue = utc
                    ? moment.utc(newDate).format(format)
                    : newDate.format(format);
                // 判断大小值是否是合法的日期，并且当前日期在范围内
                var isMinDateValid = (minDate === null || minDate === void 0 ? void 0 : minDate.isValid())
                    ? newDate.isSameOrAfter(minDate)
                    : true;
                var isMaxDateValid = (maxDate === null || maxDate === void 0 ? void 0 : maxDate.isValid())
                    ? newDate.isSameOrBefore(maxDate)
                    : true;
                // 小于 0 的日期丢弃
                if (!dateValue.startsWith('-') && isMinDateValid && isMaxDateValid) {
                    onChange(dateValue);
                }
            }
        }
    };
    DatePicker.prototype.onInputBlur = function () {
        this.setState({
            inputValue: this.inputValueCache
        });
    };
    DatePicker.prototype.selectRannge = function (item) {
        var closeOnSelect = this.props.closeOnSelect;
        var now = moment();
        this.handleChange(item.date(now));
        closeOnSelect && this.close();
    };
    DatePicker.prototype.checkIsValidDate = function (currentDate) {
        var _a = this.props, minDate = _a.minDate, maxDate = _a.maxDate;
        if (minDate && currentDate.isBefore(minDate, 'day')) {
            return false;
        }
        else if (maxDate && currentDate.isAfter(maxDate, 'day')) {
            return false;
        }
        return true;
    };
    DatePicker.prototype.getTarget = function () {
        return this.dom;
    };
    DatePicker.prototype.getParent = function () {
        return this.dom;
    };
    DatePicker.prototype.getAvailableShortcuts = function (key) {
        if (availableShortcuts[key]) {
            return availableShortcuts[key];
        }
        var __ = this.props.translate;
        for (var i = 0, len = advancedShortcuts.length; i < len; i++) {
            var item = advancedShortcuts[i];
            var m = item.regexp.exec(key);
            if (m) {
                return item.resolve.apply(item, __spreadArray([__], __read(m), false));
            }
        }
        return null;
    };
    DatePicker.prototype.renderShortCuts = function (shortcuts) {
        var _this = this;
        if (!shortcuts) {
            return null;
        }
        var _a = this.props; _a.classPrefix; var cx = _a.classnames;
        var shortcutArr;
        if (typeof shortcuts === 'string') {
            shortcutArr = shortcuts.split(',');
        }
        else {
            shortcutArr = shortcuts;
        }
        var __ = this.props.translate;
        return (React__default.createElement("ul", { className: cx("DatePicker-shortcuts") }, shortcutArr.map(function (item) {
            if (!item) {
                return null;
            }
            var shortcut = {};
            if (typeof item === 'string') {
                shortcut = _this.getAvailableShortcuts(item);
                shortcut.key = item;
            }
            else if (item.date) {
                shortcut = __assign(__assign({}, item), { date: function () { return item.date; } });
            }
            return (React__default.createElement("li", { className: cx("DatePicker-shortcut"), onClick: function () { return _this.selectRannge(shortcut); }, key: shortcut.key || shortcut.label },
                React__default.createElement("a", null, __(shortcut.label))));
        })));
    };
    DatePicker.prototype.render = function () {
        var _a;
        var _b = this.props, ns = _b.classPrefix, cx = _b.classnames, className = _b.className, popoverClassName = _b.popoverClassName, value = _b.value, placeholder = _b.placeholder, disabled = _b.disabled, inputFormat = _b.inputFormat, dateFormat = _b.dateFormat, timeFormat = _b.timeFormat, viewMode = _b.viewMode, timeConstraints = _b.timeConstraints, popOverContainer = _b.popOverContainer, clearable = _b.clearable, shortcuts = _b.shortcuts; _b.utc; var overlayPlacement = _b.overlayPlacement, locale = _b.locale, format = _b.format, borderMode = _b.borderMode, embed = _b.embed, minDate = _b.minDate, useMobileUI = _b.useMobileUI, maxDate = _b.maxDate, schedules = _b.schedules, largeMode = _b.largeMode, scheduleClassNames = _b.scheduleClassNames, todayActiveStyle = _b.todayActiveStyle, onScheduleClick = _b.onScheduleClick, mobileCalendarMode = _b.mobileCalendarMode, label = _b.label;
        var __ = this.props.translate;
        var isOpened = this.state.isOpened;
        var date = this.state.value;
        var calendarMobile = (React__default.createElement(CalendarMobile, { isDatePicker: true, timeFormat: timeFormat, inputFormat: inputFormat, startDate: date, defaultDate: date, minDate: minDate, maxDate: maxDate, dateFormat: dateFormat, embed: embed, viewMode: viewMode, close: this.close, confirm: this.handleChange, footerExtra: this.renderShortCuts(shortcuts), showViewMode: viewMode === 'quarters' || viewMode === 'months' ? 'years' : 'months', timeConstraints: timeConstraints }));
        var CalendarMobileTitle = (React__default.createElement("div", { className: "".concat(ns, "CalendarMobile-title") }, label && typeof label === 'string' ? label : __('Calendar.datepicker')));
        var useCalendarMobile = useMobileUI &&
            isMobile() &&
            ['days', 'months', 'quarters'].indexOf(viewMode) > -1;
        if (embed) {
            var schedulesData = undefined;
            if (schedules && Array.isArray(schedules)) {
                // 设置日程颜色
                var index_1 = 0;
                schedulesData = schedules.map(function (schedule) {
                    var className = schedule.className;
                    if (!className && scheduleClassNames) {
                        className = scheduleClassNames[index_1];
                        index_1++;
                        if (index_1 >= scheduleClassNames.length) {
                            index_1 = 0;
                        }
                    }
                    return __assign(__assign({}, schedule), { className: className });
                });
            }
            return (React__default.createElement("div", { className: cx("DateCalendar", {
                    'is-disabled': disabled,
                    'ScheduleCalendar': schedulesData,
                    'ScheduleCalendar-large': largeMode
                }, className) },
                React__default.createElement(Calendar, { value: date, onChange: this.handleChange, requiredConfirm: false, dateFormat: dateFormat, timeFormat: timeFormat, isValidDate: this.checkIsValidDate, viewMode: viewMode, timeConstraints: timeConstraints, input: false, onClose: this.close, locale: locale, minDate: minDate, maxDate: maxDate, 
                    // utc={utc}
                    schedules: schedulesData, largeMode: largeMode, todayActiveStyle: todayActiveStyle, onScheduleClick: onScheduleClick, embed: embed, useMobileUI: useMobileUI })));
        }
        return (React__default.createElement("div", { tabIndex: 0, onKeyPress: this.handleKeyPress, onFocus: this.handleFocus, onBlur: this.handleBlur, className: cx("DatePicker", (_a = {
                    'is-disabled': disabled,
                    'is-focused': !disabled && this.state.isFocused
                },
                _a["DatePicker--border".concat(ucFirst(borderMode))] = borderMode,
                _a['is-mobile'] = useMobileUI && isMobile(),
                _a), className), ref: this.domRef, onClick: this.handleClick },
            React__default.createElement(Input, { className: cx('DatePicker-input'), onChange: this.inputChange, onBlur: this.onInputBlur, ref: this.inputRef, placeholder: placeholder, autoComplete: "off", value: this.state.inputValue || '', disabled: disabled }),
            clearable && !disabled && normalizeValue(value, format) ? (React__default.createElement("a", { className: cx("DatePicker-clear"), onClick: this.clearValue },
                React__default.createElement(Icon, { icon: "input-clear", className: "icon" }))) : null,
            React__default.createElement("a", { className: cx("DatePicker-toggler") },
                React__default.createElement(Icon, { icon: viewMode === 'time' ? 'clock' : 'date', className: "icon", iconContent: viewMode === 'time'
                        ? 'DatePicker-toggler-clock'
                        : 'DatePicker-toggler-date' })),
            !(useMobileUI && isMobile()) && isOpened ? (React__default.createElement(Overlay, { target: this.getTarget, container: popOverContainer || this.getParent, rootClose: false, placement: overlayPlacement, show: true },
                React__default.createElement(PopOver, { classPrefix: ns, className: cx("DatePicker-popover", popoverClassName), onHide: this.close, overlay: true, onClick: this.handlePopOverClick },
                    this.renderShortCuts(shortcuts),
                    React__default.createElement(Calendar, { value: date, onChange: this.handleChange, requiredConfirm: viewMode === 'time', dateFormat: dateFormat, inputFormat: inputFormat, timeFormat: timeFormat, isValidDate: this.checkIsValidDate, viewMode: viewMode, timeConstraints: timeConstraints, input: false, onClose: this.close, locale: locale, minDate: minDate, maxDate: maxDate, useMobileUI: useMobileUI })))) : null,
            useMobileUI && isMobile() ? (mobileCalendarMode === 'calendar' && useCalendarMobile ? (React__default.createElement(PopUp, { isShow: isOpened, className: cx("".concat(ns, "CalendarMobile-pop")), onHide: this.close, header: CalendarMobileTitle }, calendarMobile)) : (React__default.createElement(PopUp, { className: cx("".concat(ns, "DatePicker-popup DatePicker-mobile")), container: popOverContainer, isShow: isOpened, showClose: false, onHide: this.handleClick },
                React__default.createElement(Calendar, { value: date, onChange: this.handleChange, requiredConfirm: false, dateFormat: dateFormat, inputFormat: inputFormat, timeFormat: timeFormat, isValidDate: this.checkIsValidDate, viewMode: viewMode, timeConstraints: timeConstraints, input: false, onClose: this.close, locale: locale, minDate: minDate, maxDate: maxDate, useMobileUI: useMobileUI })))) : null));
    };
    DatePicker.defaultProps = {
        viewMode: 'days',
        shortcuts: '',
        closeOnSelect: true,
        overlayPlacement: 'auto',
        scheduleClassNames: [
            'bg-warning',
            'bg-danger',
            'bg-success',
            'bg-info',
            'bg-secondary'
        ]
    };
    return DatePicker;
}(React__default.Component));
var DatePicker$1 = themeable(localeable(DatePicker));

export { DatePicker, DatePicker$1 as default };
