// amis-ui/esm/components/ListMenu.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

import { __extends, __assign } from 'tslib';
import { themeable, localeable } from 'amis-core';
import React__default from 'react';

var ListMenu = /** @class */ (function (_super) {
    __extends(ListMenu, _super);
    function ListMenu() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ListMenu.prototype.renderItem = function (result, option, optionIndex) {
        var _this = this;
        var _a = this.props, cx = _a.classnames, itemRender = _a.itemRender, disabled = _a.disabled, getItemProps = _a.getItemProps, highlightIndex = _a.highlightIndex, selectedOptions = _a.selectedOptions, onSelect = _a.onSelect;
        if (Array.isArray(option.children) && option.children.length) {
            var stackResult = {
                items: [],
                index: result.index
            };
            result.items.push(React__default.createElement("div", { className: cx('ListMenu-group'), key: optionIndex },
                React__default.createElement("div", { className: cx('ListMenu-groupLabel') }, itemRender(option)),
                option.children.reduce(function (result, option, index) {
                    return _this.renderItem(result, option, index);
                }, stackResult).items));
            result.index = stackResult.index;
            return result;
        }
        var index = result.index++;
        result.items.push(React__default.createElement("div", __assign({ className: cx('ListMenu-item', option.className, disabled || option.disabled ? 'is-disabled' : '', index === highlightIndex ? 'is-highlight' : '', ~(selectedOptions || []).indexOf(option) ? 'is-active' : ''), key: index, onClick: onSelect ? function (e) { return onSelect(e, option); } : undefined }, getItemProps({
            item: option,
            index: index
        })),
            React__default.createElement("div", { className: cx('ListMenu-itemLabel') }, itemRender(option))));
        return result;
    };
    ListMenu.prototype.render = function () {
        var _this = this;
        var _a = this.props, cx = _a.classnames, options = _a.options, placeholder = _a.placeholder, prefix = _a.prefix, children = _a.children;
        var __ = this.props.translate;
        return (React__default.createElement("div", { className: cx('ListMenu') },
            prefix,
            Array.isArray(options) && options.length ? (options.reduce(function (result, option, index) {
                return _this.renderItem(result, option, index);
            }, {
                items: [],
                index: 0
            }).items) : (React__default.createElement("span", { className: cx('ListMenu-placeholder') }, __(placeholder))),
            children));
    };
    ListMenu.defaultProps = {
        placeholder: 'placeholder.noOption',
        itemRender: function (option) { return React__default.createElement(React__default.Fragment, null, option.label); },
        getItemProps: function (props) { return null; }
    };
    return ListMenu;
}(React__default.Component));
var ListMenu$1 = themeable(localeable(ListMenu));

export { ListMenu, ListMenu$1 as default };
