// amis-ui/esm/components/Modal.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

import { __extends, __rest, __assign, __decorate, __metadata } from 'tslib';
import React__default from 'react';
import Transition, { ENTERING, ENTERED, EXITING, EXITED } from 'react-transition-group/Transition';
import Portal from 'react-overlays/Portal';
import { current, removeModal, addModal } from './ModalManager.js';
import { getScrollbarWidth, themeable, localeable, autobind } from 'amis-core';
import { Icon } from './icons.js';

/**
 * @file Modal
 * @description
 * @author fex
 */
var _a, _b;
var fadeStyles = (_a = {},
    _a[ENTERING] = 'in',
    _a[ENTERED] = 'in',
    _a[EXITING] = 'out',
    _a);
var contentFadeStyles = (_b = {},
    _b[ENTERING] = 'in',
    _b[ENTERED] = '',
    _b[EXITING] = 'out',
    _b);
var Modal = /** @class */ (function (_super) {
    __extends(Modal, _super);
    function Modal() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isRootClosed = false;
        _this.handleEnter = function () {
            document.body.classList.add("is-modalOpened");
            if (window.innerWidth - document.documentElement.clientWidth > 0 ||
                document.body.scrollHeight > document.body.clientHeight) {
                var scrollbarWidth = getScrollbarWidth();
                document.body.style.width = "calc(100% - ".concat(scrollbarWidth, "px)");
            }
        };
        _this.handleEntered = function () {
            var onEntered = _this.props.onEntered;
            onEntered && onEntered();
        };
        _this.handleExited = function () {
            var onExited = _this.props.onExited;
            onExited && onExited();
            setTimeout(function () {
                if (!document.querySelector('.amis-dialog-widget')) {
                    document.body.classList.remove("is-modalOpened");
                    document.body.style.width = '';
                }
            }, 200);
        };
        _this.modalRef = function (ref) {
            _this.modalDom = ref;
            var ns = _this.props.classPrefix;
            if (ref) {
                addModal(_this);
                ref.classList.add("".concat(ns, "Modal--").concat(current(), "th"));
            }
            else {
                removeModal(_this);
            }
        };
        return _this;
    }
    Modal.prototype.componentDidMount = function () {
        if (this.props.show) {
            this.handleEnter();
            this.handleEntered();
        }
        document.body.addEventListener('click', this.handleRootClickCapture, true);
        document.body.addEventListener('click', this.handleRootClick);
    };
    Modal.prototype.componentWillUnmount = function () {
        if (this.props.show) {
            this.handleExited();
        }
        document.body.removeEventListener('click', this.handleRootClick);
        document.body.removeEventListener('click', this.handleRootClickCapture, true);
    };
    Modal.prototype.handleRootClickCapture = function (e) {
        var target = e.target;
        var _a = this.props, closeOnOutside = _a.closeOnOutside, ns = _a.classPrefix;
        var isLeftButton = (e.button === 1 && window.event !== null) || e.button === 0;
        this.isRootClosed = !!(isLeftButton &&
            closeOnOutside &&
            target &&
            this.modalDom &&
            ((!this.modalDom.contains(target) && !target.closest('[role=dialog]')) ||
                (target.matches(".".concat(ns, "Modal")) && target === this.modalDom))); // 干脆过滤掉来自弹框里面的点击
    };
    Modal.prototype.handleRootClick = function (e) {
        var onHide = this.props.onHide;
        this.isRootClosed && !e.defaultPrevented && onHide(e);
    };
    Modal.prototype.render = function () {
        var _this = this;
        var _a = this.props, className = _a.className, contentClassName = _a.contentClassName, children = _a.children, container = _a.container, show = _a.show, size = _a.size, overlay = _a.overlay, width = _a.width, height = _a.height, cx = _a.classnames;
        return (React__default.createElement(Transition, { mountOnEnter: true, unmountOnExit: true, appear: true, in: show, timeout: 500, onEnter: this.handleEnter, onExited: this.handleExited, onEntered: this.handleEntered }, function (status) {
            var _a;
            return (React__default.createElement(Portal, { container: container },
                React__default.createElement("div", { ref: _this.modalRef, role: "dialog", className: cx("amis-dialog-widget Modal", (_a = {},
                        _a["Modal--".concat(size)] = size,
                        _a), className) },
                    overlay ? (React__default.createElement("div", { className: cx("Modal-overlay", fadeStyles[status]) })) : null,
                    React__default.createElement("div", { className: cx("Modal-content", contentClassName, contentFadeStyles[status]), style: { width: width, height: height } }, status === EXITED ? null : children))));
        }));
    };
    Modal.defaultProps = {
        container: document.body,
        size: '',
        overlay: true
    };
    Modal.Header = themeable(localeable(function (_a) {
        var cx = _a.classnames, className = _a.className, showCloseButton = _a.showCloseButton, onClose = _a.onClose, children = _a.children; _a.classPrefix; var __ = _a.translate; _a.forwardedRef; var rest = __rest(_a, ["classnames", "className", "showCloseButton", "onClose", "children", "classPrefix", "translate", "forwardedRef"]);
        return (React__default.createElement("div", __assign({}, rest, { className: cx('Modal-header', className) }),
            showCloseButton !== false ? (React__default.createElement("a", { "data-tooltip": __('Dialog.close'), "data-position": "left", onClick: onClose, className: cx('Modal-close') },
                React__default.createElement(Icon, { icon: "close", className: "icon" }))) : null,
            children));
    }));
    Modal.Title = themeable(function (_a) {
        var cx = _a.classnames, className = _a.className, children = _a.children; _a.classPrefix; _a.forwardedRef; var rest = __rest(_a, ["classnames", "className", "children", "classPrefix", "forwardedRef"]);
        return (React__default.createElement("div", __assign({}, rest, { className: cx('Modal-title', className) }), children));
    });
    Modal.Body = themeable(function (_a) {
        var cx = _a.classnames, className = _a.className, children = _a.children; _a.classPrefix; _a.forwardedRef; var rest = __rest(_a, ["classnames", "className", "children", "classPrefix", "forwardedRef"]);
        return (React__default.createElement("div", __assign({}, rest, { className: cx('Modal-body', className) }), children));
    });
    Modal.Footer = themeable(function (_a) {
        var cx = _a.classnames, className = _a.className, children = _a.children; _a.classPrefix; _a.forwardedRef; var rest = __rest(_a, ["classnames", "className", "children", "classPrefix", "forwardedRef"]);
        return (React__default.createElement("div", __assign({}, rest, { className: cx('Modal-footer', className) }), children));
    });
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [MouseEvent]),
        __metadata("design:returntype", void 0)
    ], Modal.prototype, "handleRootClickCapture", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [MouseEvent]),
        __metadata("design:returntype", void 0)
    ], Modal.prototype, "handleRootClick", null);
    return Modal;
}(React__default.Component));
var FinalModal = themeable(localeable(Modal));

export { Modal, FinalModal as default };
