// amis-ui/esm/components/NumberInput.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

import { __extends, __assign, __decorate, __metadata } from 'tslib';
import React__default from 'react';
import isInteger from 'lodash/isInteger';
import InputNumber from 'rc-input-number';
import getMiniDecimal, { getNumberPrecision, toFixed, num2str } from '@rc-component/mini-decimal';
import { Icon } from './icons.js';
import { ucFirst, isNumeric, autobind, themeable } from 'amis-core';

var NumberInput = /** @class */ (function (_super) {
    __extends(NumberInput, _super);
    function NumberInput(props) {
        var _this = _super.call(this, props) || this;
        /**
         * 是否是 bigNumber，如果输入的内容是字符串就自动开启
         */
        _this.isBig = false;
        // 严格判断大数模式，因为初始化value为empty string时，修改value值格式仍然为string
        _this.isBig = !!props.big;
        return _this;
    }
    NumberInput.prototype.componentDidUpdate = function (prevProps) {
        var isBig = !!this.props.big;
        if (!!(prevProps === null || prevProps === void 0 ? void 0 : prevProps.big) !== isBig) {
            this.isBig = isBig;
        }
    };
    NumberInput.prototype.handleChange = function (value) {
        var _a = this.props, min = _a.min, max = _a.max, step = _a.step, precision = _a.precision, resetValue = _a.resetValue, clearValueOnEmpty = _a.clearValueOnEmpty, onChange = _a.onChange;
        var finalPrecision = NumberInput.normalizePrecision(precision, step);
        var result = NumberInput.normalizeValue(value, min, max, finalPrecision, resetValue, clearValueOnEmpty, this.isBig);
        onChange === null || onChange === void 0 ? void 0 : onChange(result);
    };
    NumberInput.prototype.handleFocus = function (e) {
        var onFocus = this.props.onFocus;
        onFocus && onFocus(e);
    };
    NumberInput.prototype.handleBlur = function (e) {
        var onBlur = this.props.onBlur;
        onBlur && onBlur(e);
    };
    NumberInput.prototype.handleEnhanceModeChange = function (action) {
        var _a = this.props, value = _a.value, _b = _a.step, step = _b === void 0 ? 1 : _b, disabled = _a.disabled, readOnly = _a.readOnly, precision = _a.precision;
        // value为undefined会导致溢出错误
        var val = value || 0;
        if (disabled || readOnly) {
            return;
        }
        if (isNaN(Number(step)) || !Number(step)) {
            return;
        }
        var stepDecimal = getMiniDecimal(step);
        if (action !== 'add') {
            stepDecimal = stepDecimal.negate();
        }
        var target = getMiniDecimal(val).add(stepDecimal.toString());
        var getPrecision = function (numStr) {
            if (precision != null && precision >= 0) {
                return precision;
            }
            return Math.max(getNumberPrecision(numStr), getNumberPrecision(Number(step) || 1));
        };
        var triggerValueUpdate = function (newValue, userTyping) {
            var updateValue = newValue;
            var numStr = updateValue.toString();
            var mergedPrecision = getPrecision(numStr);
            if (mergedPrecision >= 0) {
                updateValue = getMiniDecimal(toFixed(numStr, '.', mergedPrecision));
            }
            return updateValue;
        };
        var updatedValue = triggerValueUpdate(target);
        if (this.isBig) {
            this.handleChange(updatedValue.toString());
        }
        else {
            val = Number(updatedValue.toString());
            this.handleChange(val);
        }
    };
    NumberInput.prototype.renderBase = function () {
        var _a;
        var _b = this.props, className = _b.className, ns = _b.classPrefix, cx = _b.classnames, value = _b.value, step = _b.step, precision = _b.precision, max = _b.max, min = _b.min, disabled = _b.disabled, placeholder = _b.placeholder, showSteps = _b.showSteps, formatter = _b.formatter, parser = _b.parser, borderMode = _b.borderMode, readOnly = _b.readOnly, displayMode = _b.displayMode, inputRef = _b.inputRef, keyboard = _b.keyboard;
        var precisionProps = {
            precision: NumberInput.normalizePrecision(precision, step)
        };
        return (React__default.createElement(InputNumber, __assign({ className: cx(className, showSteps === false ? 'no-steps' : '', displayMode === 'enhance' ? 'Number--enhance-input' : '', (_a = {},
                _a["Number--border".concat(ucFirst(borderMode))] = borderMode,
                _a)), ref: inputRef, readOnly: readOnly, prefixCls: "".concat(ns, "Number"), value: value, step: step, max: max, min: min, formatter: formatter, parser: parser, onChange: this.handleChange, disabled: disabled, placeholder: placeholder, onFocus: this.handleFocus, onBlur: this.handleBlur, stringMode: this.isBig ? true : false, keyboard: keyboard }, precisionProps)));
    };
    NumberInput.prototype.render = function () {
        var _a;
        var _this = this;
        var _b = this.props, cx = _b.classnames, value = _b.value, max = _b.max, min = _b.min, disabled = _b.disabled, showSteps = _b.showSteps, borderMode = _b.borderMode, readOnly = _b.readOnly, displayMode = _b.displayMode;
        return (React__default.createElement(React__default.Fragment, null, displayMode === 'enhance' ? (React__default.createElement("div", { className: cx('Number--enhance', disabled ? 'Number--enhance-disabled' : '', showSteps === false ? 'Number--enhance-no-steps' : '', (_a = {},
                _a["Number--enhance-border".concat(ucFirst(borderMode))] = borderMode,
                _a)) },
            React__default.createElement("div", { className: cx('Number--enhance-left-icon', value && value === min ? 'Number--enhance-border-min' : '', disabled ? 'Number--enhance-border-disabled' : '', readOnly ? 'Number--enhance-border-readOnly' : ''), onClick: function () { return _this.handleEnhanceModeChange('subtract'); } },
                React__default.createElement(Icon, { icon: "minus", className: "icon", wrapClassName: cx('InputNumber-enhance-minus icon'), iconContent: "InputNumber-enhance-minus" })),
            this.renderBase(),
            React__default.createElement("div", { className: cx('Number--enhance-right-icon', value && value === max ? 'Number--enhance-border-max' : '', disabled ? 'Number--enhance-border-disabled' : '', readOnly ? 'Number--enhance-border-readOnly' : ''), onClick: function () { return _this.handleEnhanceModeChange('add'); } },
                React__default.createElement(Icon, { icon: "plus", className: "icon", wrapClassName: cx('InputNumber-enhance-plus icon'), iconContent: "InputNumber-enhance-plus" })))) : (this.renderBase())));
    };
    NumberInput.defaultProps = {
        step: 1,
        readOnly: false,
        borderMode: 'full',
        resetValue: ''
    };
    /**
     * 处理value值
     *
     * @param value value 值
     * @param min 最小值
     * @param max 最大值
     * @param precision 精度
     * @param resetValue 重置值
     * @param isBig 是否为大数模式
     */
    NumberInput.normalizeValue = function (value, min, max, precision, resetValue, clearValueOnEmpty, isBig) {
        /**
         * 输入不合法时重置为resetValue
         * 若resetValue为不合法数字，直接清空输入
         * 若resetValue为数字，则需要处理max，min，precision，保证抛出的值满足条件
         */
        if (!isNumeric(value)) {
            if (!isNumeric(resetValue)) {
                return clearValueOnEmpty ? undefined : '';
            }
            value = resetValue;
        }
        // 处理max & min
        if (typeof value === 'number') {
            if (typeof min === 'number') {
                value = Math.max(value, min);
            }
            if (typeof max === 'number') {
                value = Math.min(value, max);
            }
        }
        // 处理string类型输入
        if (typeof value === 'string') {
            var val = getMiniDecimal(value);
            if (typeof min !== 'undefined') {
                var minValue = getMiniDecimal(min);
                if (val.lessEquals(minValue)) {
                    value = min;
                }
            }
            if (typeof max !== 'undefined') {
                var maxValue = getMiniDecimal(max);
                if (maxValue.lessEquals(val)) {
                    value = max;
                }
            }
        }
        /**
         * 非大数模式下，如果精度不满足要求，需要处理value值，遵循四舍五入的处理规则
         */
        if (!isBig && getNumberPrecision(value) !== precision) {
            value = getMiniDecimal(toFixed(num2str(value), '.', precision)).toNumber();
        }
        return value;
    };
    /**
     * 获取精度，合法的精度为0和正整数，不合法的精度统一转化为0
     * 若设置了step，则会基于step的精度生成，最终使用更高的精度
     *
     * @param precision 精度
     * @param step 步长
     */
    NumberInput.normalizePrecision = function (precision, step) {
        if (typeof precision === 'number' &&
            isInteger(precision) &&
            precision >= 0) {
            return Math.max(precision, getNumberPrecision(step !== null && step !== void 0 ? step : 1));
        }
        // 如果设置了step，就基于step和precision，选取更高精度
        if (step != null) {
            return Math.max(0, getNumberPrecision(step));
        }
        return 0;
    };
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], NumberInput.prototype, "handleChange", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], NumberInput.prototype, "handleFocus", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], NumberInput.prototype, "handleBlur", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], NumberInput.prototype, "handleEnhanceModeChange", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], NumberInput.prototype, "renderBase", null);
    return NumberInput;
}(React__default.Component));
var NumberInput$1 = themeable(NumberInput);

export { NumberInput, NumberInput$1 as default };
