// amis-ui/esm/components/TabsTransferPicker.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

import { __extends, __rest, __assign, __decorate, __metadata } from 'tslib';
import { mapTree, autobind, themeable, localeable, uncontrollable } from 'amis-core';
import React__default from 'react';
import ResultBox from './ResultBox.js';
import { Icon } from './icons.js';
import PickerContainer from './PickerContainer.js';
import TabsTransfer from './TabsTransfer.js';

var TransferPicker = /** @class */ (function (_super) {
    __extends(TransferPicker, _super);
    function TransferPicker() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.optionModified = false;
        return _this;
    }
    TransferPicker.prototype.handleConfirm = function (value) {
        var _a, _b;
        (_b = (_a = this.props).onChange) === null || _b === void 0 ? void 0 : _b.call(_a, value, this.optionModified);
        this.optionModified = false;
    };
    TransferPicker.prototype.onFoucs = function () {
        var _a, _b;
        (_b = (_a = this.props).onFocus) === null || _b === void 0 ? void 0 : _b.call(_a);
    };
    TransferPicker.prototype.onBlur = function () {
        var _a, _b;
        (_b = (_a = this.props).onBlur) === null || _b === void 0 ? void 0 : _b.call(_a);
    };
    TransferPicker.prototype.render = function () {
        var _this = this;
        var _a = this.props, cx = _a.classnames, value = _a.value, __ = _a.translate, disabled = _a.disabled, className = _a.className, onChange = _a.onChange, size = _a.size, rest = __rest(_a, ["classnames", "value", "translate", "disabled", "className", "onChange", "size"]);
        return (React__default.createElement(PickerContainer, { title: __('Select.placeholder'), onFocus: this.onFoucs, onClose: this.onBlur, bodyRender: function (_a) {
                _a.onClose; var value = _a.value, onChange = _a.onChange, setState = _a.setState, states = __rest(_a, ["onClose", "value", "onChange", "setState"]);
                return (React__default.createElement(TabsTransfer, __assign({}, rest, states, { value: value, onChange: function (value, optionModified) {
                        if (optionModified) {
                            var options = mapTree(rest.options, function (item) {
                                return (value.find(function (a) { return a.value === item.value; }) || item);
                            });
                            _this.optionModified = true;
                            setState({ options: options, value: value });
                        }
                        else {
                            onChange(value);
                        }
                    } })));
            }, value: value, onConfirm: this.handleConfirm, size: size }, function (_a) {
            var onClick = _a.onClick, isOpened = _a.isOpened;
            return (React__default.createElement(ResultBox, { className: cx('TransferPicker', className, isOpened ? 'is-active' : ''), allowInput: false, result: value, onResultChange: onChange, onResultClick: onClick, placeholder: __('Select.placeholder'), disabled: disabled },
                React__default.createElement("span", { className: cx('TransferPicker-icon') },
                    React__default.createElement(Icon, { icon: "pencil", className: "icon" }))));
        }));
    };
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], TransferPicker.prototype, "handleConfirm", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], TransferPicker.prototype, "onFoucs", null);
    __decorate([
        autobind,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], TransferPicker.prototype, "onBlur", null);
    return TransferPicker;
}(React__default.Component));
var TabsTransferPicker = themeable(localeable(uncontrollable(TransferPicker, {
    value: 'onChange'
})));

export { TransferPicker, TabsTransferPicker as default };
