// amis-ui/esm/components/virtual-list/constants.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

var _a, _b, _c, _d, _e;
var ALIGNMENT;
(function (ALIGNMENT) {
    ALIGNMENT["AUTO"] = "auto";
    ALIGNMENT["START"] = "start";
    ALIGNMENT["CENTER"] = "center";
    ALIGNMENT["END"] = "end";
})(ALIGNMENT || (ALIGNMENT = {}));
var DIRECTION;
(function (DIRECTION) {
    DIRECTION["HORIZONTAL"] = "horizontal";
    DIRECTION["VERTICAL"] = "vertical";
})(DIRECTION || (DIRECTION = {}));
var SCROLL_CHANGE_REASON;
(function (SCROLL_CHANGE_REASON) {
    SCROLL_CHANGE_REASON["OBSERVED"] = "observed";
    SCROLL_CHANGE_REASON["REQUESTED"] = "requested";
})(SCROLL_CHANGE_REASON || (SCROLL_CHANGE_REASON = {}));
var scrollProp = (_a = {},
    _a[DIRECTION.VERTICAL] = 'scrollTop',
    _a[DIRECTION.HORIZONTAL] = 'scrollLeft',
    _a);
var sizeProp = (_b = {},
    _b[DIRECTION.VERTICAL] = 'height',
    _b[DIRECTION.HORIZONTAL] = 'width',
    _b);
var positionProp = (_c = {},
    _c[DIRECTION.VERTICAL] = 'top',
    _c[DIRECTION.HORIZONTAL] = 'left',
    _c);
var marginProp = (_d = {},
    _d[DIRECTION.VERTICAL] = 'marginTop',
    _d[DIRECTION.HORIZONTAL] = 'marginLeft',
    _d);
var oppositeMarginProp = (_e = {},
    _e[DIRECTION.VERTICAL] = 'marginBottom',
    _e[DIRECTION.HORIZONTAL] = 'marginRight',
    _e);

export { ALIGNMENT, DIRECTION, SCROLL_CHANGE_REASON, marginProp, oppositeMarginProp, positionProp, scrollProp, sizeProp };
