// amis-ui/lib/components/Alert.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var React = require('react');
var ReactDOM = require('react-dom');
var Modal = require('./Modal.js');
var Button = require('./Button.js');
var amisCore = require('amis-core');
var Html = require('./Html.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

/**
 * @file Alert
 * @author fex
 */
var Alert = /** @class */ (function (_super) {
    tslib.__extends(Alert, _super);
    function Alert(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            show: false,
            title: '',
            content: '',
            confirm: false
        };
        _this.close = _this.close.bind(_this);
        _this.handleConfirm = _this.handleConfirm.bind(_this);
        _this.handleCancel = _this.handleCancel.bind(_this);
        _this.modalRef = _this.modalRef.bind(_this);
        _this.handleFormSubmit = _this.handleFormSubmit.bind(_this);
        _this.scopeRef = _this.scopeRef.bind(_this);
        if (!props.isolate) {
            _this.originInstance = Alert.instance;
            Alert.instance = _this;
        }
        return _this;
    }
    Alert.getInstance = function () {
        if (!Alert.instance) {
            console.warn('Alert 组件应该没有被渲染，所以隐性的渲染到 body 了');
            var container = document.body;
            var div = document.createElement('div');
            container.appendChild(div);
            ReactDOM.render(React__default["default"].createElement(FinnalAlert, null), div);
        }
        return Alert.instance;
    };
    Alert.prototype.componentDidMount = function () {
        this._body && (this._body.innerHTML = this.state.content);
    };
    Alert.prototype.componentDidUpdate = function (prevProps, prevState) {
        if (prevState.content !== this.state.content) {
            this._body && (this._body.innerHTML = this.state.content);
        }
    };
    Alert.prototype.componentWillUnmount = function () {
        if (Alert.instance === this) {
            Alert.instance = this.originInstance || null;
            this.originInstance = null;
        }
    };
    Alert.prototype.scopeRef = function (schemaSope) {
        this.schemaSope = schemaSope;
    };
    Alert.prototype.handleConfirm = function () {
        var _a;
        var form = (_a = this.schemaSope) === null || _a === void 0 ? void 0 : _a.getComponentByName('form');
        if (form) {
            form.doAction({ type: 'submit' });
        }
        else {
            this.close(true);
        }
    };
    Alert.prototype.handleCancel = function () {
        this.close(false);
    };
    Alert.prototype.close = function (confirmed) {
        var _this = this;
        var isConfirm = this.state.confirm || this.state.prompt;
        this.setState({
            show: false,
            prompt: false,
            confirm: false
        }, isConfirm ? function () { return _this._resolve(confirmed); } /*this._reject()*/ : undefined);
    };
    Alert.prototype.alert = function (content, title) {
        this.setState({
            title: title,
            content: content,
            show: true,
            confirm: false
        });
    };
    Alert.prototype.confirm = function (content, title, confirmText, cancelText) {
        var _this = this;
        this.setState({
            title: title,
            content: content,
            show: true,
            confirm: true,
            confirmText: confirmText,
            cancelText: cancelText
        });
        return new Promise(function (resolve) {
            _this._resolve = resolve;
        });
    };
    Alert.prototype.prompt = function (controls, defaultValue, title, confirmText) {
        var _this = this;
        if (title === void 0) { title = 'placeholder.enter'; }
        if (confirmText === void 0) { confirmText = 'confirm'; }
        if (typeof controls === 'string') {
            // 兼容浏览器标准用法。
            controls = [
                {
                    name: 'text',
                    label: controls,
                    type: 'text'
                }
            ];
            if (typeof defaultValue === 'string') {
                defaultValue = {
                    text: defaultValue
                };
            }
        }
        else if (!Array.isArray(controls)) {
            controls = [controls];
        }
        this.setState({
            title: title,
            controls: controls,
            show: true,
            prompt: true,
            value: defaultValue,
            confirmText: confirmText
        });
        return new Promise(function (resolve) {
            _this._resolve = resolve;
        });
    };
    Alert.prototype.modalRef = function (ref) {
        this._modal = ref;
    };
    Alert.prototype.handleFormSubmit = function (values) {
        this.close(values);
    };
    Alert.prototype.render = function () {
        var _a, _b, _c;
        var _d = this.props, container = _d.container, cancelText = _d.cancelText, confirmText = _d.confirmText, title = _d.title, confirmBtnLevel = _d.confirmBtnLevel, alertBtnLevel = _d.alertBtnLevel, cx = _d.classnames;
        var theme = this.props.theme || 'cxd';
        if (theme === 'default') {
            theme = 'cxd';
        }
        var __ = this.props.translate;
        var finalTitle = __((_a = this.state.title) !== null && _a !== void 0 ? _a : title);
        var finalConfirmText = __((_b = this.state.confirmText) !== null && _b !== void 0 ? _b : confirmText);
        var finalCancelText = __((_c = this.state.cancelText) !== null && _c !== void 0 ? _c : cancelText);
        return (React__default["default"].createElement(Modal["default"], { show: this.state.show, onHide: this.handleCancel, container: container, ref: this.modalRef, closeOnEsc: true },
            finalTitle ? (React__default["default"].createElement("div", { className: cx('Modal-header') },
                React__default["default"].createElement("div", { className: cx('Modal-title') }, finalTitle))) : null,
            React__default["default"].createElement("div", { className: cx('Modal-body') }, this.state.prompt ? (renderForm(this.state.controls, this.state.value, this.handleFormSubmit, this.scopeRef, theme)) : (React__default["default"].createElement(Html["default"], { html: this.state.content }))),
            finalConfirmText ? (React__default["default"].createElement("div", { className: cx('Modal-footer') },
                this.state.confirm || this.state.prompt ? (React__default["default"].createElement(Button["default"], { onClick: this.handleCancel }, __(finalCancelText))) : null,
                React__default["default"].createElement(Button["default"], { level: this.state.confirm || this.state.prompt
                        ? confirmBtnLevel
                        : alertBtnLevel, onClick: this.handleConfirm }, finalConfirmText))) : null));
    };
    Alert.instance = null;
    Alert.defaultProps = {
        confirmText: 'confirm',
        cancelText: 'cancel',
        title: 'Alert.info',
        alertBtnLevel: 'primary',
        confirmBtnLevel: 'danger'
    };
    return Alert;
}(React__default["default"].Component));
var renderSchemaFn;
function setRenderSchemaFn(fn) {
    renderSchemaFn = fn;
}
function renderForm(controls, value, callback, scopeRef, theme) {
    if (value === void 0) { value = {}; }
    return renderSchemaFn === null || renderSchemaFn === void 0 ? void 0 : renderSchemaFn(controls, value, callback, scopeRef, theme);
}
var alert = function (content, title) { return Alert.getInstance().alert(content, title); };
var confirm = function (content, title, confirmText, cancelText) {
    return Alert.getInstance().confirm(content, title, confirmText, cancelText);
};
var prompt = function (controls, defaultvalue, title, confirmText) {
    return Alert.getInstance().prompt(controls, defaultvalue, title, confirmText);
};
var FinnalAlert = amisCore.themeable(amisCore.localeable(Alert));

exports.Alert = Alert;
exports.FinnalAlert = FinnalAlert;
exports.alert = alert;
exports.confirm = confirm;
exports["default"] = FinnalAlert;
exports.prompt = prompt;
exports.setRenderSchemaFn = setRenderSchemaFn;
