// amis-ui/lib/components/ConfirmBox.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var React = require('react');
var Modal = require('./Modal.js');
var Button = require('./Button.js');
var Drawer = require('./Drawer.js');
var amisCore = require('amis-core');
var Spinner = require('./Spinner.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ConfirmBox(_a) {
    var _this = this;
    var type = _a.type, size = _a.size, closeOnEsc = _a.closeOnEsc, show = _a.show, onCancel = _a.onCancel, title = _a.title, showTitle = _a.showTitle, headerClassName = _a.headerClassName, __ = _a.translate, children = _a.children, showFooter = _a.showFooter, onConfirm = _a.onConfirm, beforeConfirm = _a.beforeConfirm, popOverContainer = _a.popOverContainer, position = _a.position, resizable = _a.resizable, cx = _a.classnames, className = _a.className, bodyClassName = _a.bodyClassName, footerClassName = _a.footerClassName;
    var _b = tslib.__read(React__default["default"].useState(), 2), loading = _b[0], setLoading = _b[1];
    var _c = tslib.__read(React__default["default"].useState(), 2), error = _c[0], setError = _c[1];
    var bodyRef = React__default["default"].useRef();
    var handleConfirm = React__default["default"].useCallback(function () { return tslib.__awaiter(_this, void 0, void 0, function () {
        var ret, _a, e_1;
        var _b, _c;
        return tslib.__generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    setError('');
                    setLoading(true);
                    _d.label = 1;
                case 1:
                    _d.trys.push([1, 7, 8, 9]);
                    if (!beforeConfirm) return [3 /*break*/, 3];
                    return [4 /*yield*/, (beforeConfirm === null || beforeConfirm === void 0 ? void 0 : beforeConfirm(bodyRef.current))];
                case 2:
                    _a = _d.sent();
                    return [3 /*break*/, 5];
                case 3: return [4 /*yield*/, ((_c = (_b = bodyRef.current) === null || _b === void 0 ? void 0 : _b.submit) === null || _c === void 0 ? void 0 : _c.call(_b))];
                case 4:
                    _a = _d.sent();
                    _d.label = 5;
                case 5:
                    ret = _a;
                    if (ret === false) {
                        return [2 /*return*/];
                    }
                    else if (typeof ret === 'string') {
                        setError(ret);
                        return [2 /*return*/];
                    }
                    return [4 /*yield*/, (onConfirm === null || onConfirm === void 0 ? void 0 : onConfirm(ret))];
                case 6:
                    _d.sent();
                    return [3 /*break*/, 9];
                case 7:
                    e_1 = _d.sent();
                    setError(e_1.message);
                    return [3 /*break*/, 9];
                case 8:
                    setLoading(false);
                    return [7 /*endfinally*/];
                case 9: return [2 /*return*/];
            }
        });
    }); }, [onConfirm, beforeConfirm]);
    React__default["default"].useEffect(function () {
        show && setError('');
    }, [show]);
    function renderDialog() {
        return (React__default["default"].createElement(Modal["default"], { size: size, closeOnEsc: closeOnEsc, show: show, onHide: onCancel, container: popOverContainer, className: className },
            showTitle !== false && title ? (React__default["default"].createElement(Modal["default"].Header, { onClose: onCancel, className: headerClassName }, title)) : null,
            React__default["default"].createElement(Modal["default"].Body, { className: bodyClassName }, typeof children === 'function'
                ? children({
                    bodyRef: bodyRef,
                    loading: loading
                })
                : children),
            (showFooter !== null && showFooter !== void 0 ? showFooter : true) ? (React__default["default"].createElement(Modal["default"].Footer, { className: footerClassName },
                loading || error ? (React__default["default"].createElement("div", { className: cx('Dialog-info') },
                    React__default["default"].createElement(Spinner["default"], { size: "sm", key: "info", show: loading }),
                    error ? (React__default["default"].createElement("span", { className: cx('Dialog-error') }, error)) : null)) : null,
                React__default["default"].createElement(Button["default"], { disabled: loading, onClick: onCancel }, __('cancel')),
                React__default["default"].createElement(Button["default"], { disabled: loading, onClick: handleConfirm, level: "primary" }, __('confirm')))) : null));
    }
    function renderDrawer() {
        return (React__default["default"].createElement(Drawer["default"], { size: size, closeOnEsc: closeOnEsc, show: show, onHide: onCancel, container: popOverContainer, position: position, resizable: resizable, showCloseButton: false, className: className },
            showTitle !== false && title ? (React__default["default"].createElement("div", { className: cx('Drawer-header', headerClassName) },
                React__default["default"].createElement("div", { className: cx('Drawer-title') }, title))) : null,
            React__default["default"].createElement("div", { className: cx('Drawer-body', bodyClassName) }, typeof children === 'function'
                ? children({
                    bodyRef: bodyRef
                })
                : children),
            (showFooter !== null && showFooter !== void 0 ? showFooter : true) ? (React__default["default"].createElement("div", { className: cx('Drawer-footer', footerClassName) },
                loading || error ? (React__default["default"].createElement("div", { className: cx('Drawer-info') },
                    React__default["default"].createElement(Spinner["default"], { size: "sm", key: "info", show: loading }),
                    error ? (React__default["default"].createElement("span", { className: cx('Drawer-error') }, error)) : null)) : null,
                React__default["default"].createElement(Button["default"], { onClick: handleConfirm, level: "primary" }, __('confirm')),
                React__default["default"].createElement(Button["default"], { onClick: onCancel }, __('cancel')))) : null));
    }
    return type === 'drawer' ? renderDrawer() : renderDialog();
}
ConfirmBox.defaultProps = {
    type: 'dialog',
    position: 'right'
};
var ConfirmBox$1 = amisCore.localeable(amisCore.themeable(ConfirmBox));

exports.ConfirmBox = ConfirmBox;
exports["default"] = ConfirmBox$1;
