// amis-ui/lib/components/InputTable.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var amisCore = require('amis-core');
var React = require('react');
var reactHookForm = require('react-hook-form');
var useSubForm = require('../hooks/use-sub-form.js');
var Button = require('./Button.js');
var FormField = require('./FormField.js');
var icons = require('./icons.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function InputTable(_a) {
    var _this = this;
    var _b, _c;
    var control = _a.control, name = _a.name, wrap = _a.wrap, mode = _a.mode, label = _a.label, labelAlign = _a.labelAlign, labelClassName = _a.labelClassName, description = _a.description, fieldClassName = _a.fieldClassName, className = _a.className, __ = _a.translate, cx = _a.classnames, removable = _a.removable, columns = _a.columns, addable = _a.addable, addButtonText = _a.addButtonText, addButtonClassName = _a.addButtonClassName, scaffold = _a.scaffold, minLength = _a.minLength, maxLength = _a.maxLength, isRequired = _a.isRequired, rules = _a.rules;
    var subForms = React__default["default"].useRef({});
    var subFormRef = React__default["default"].useCallback(function (subform, index) {
        if (subform) {
            subForms.current[index] = subform;
        }
        else {
            delete subForms.current[index];
        }
    }, [subForms]);
    var finalRules = tslib.__assign({}, rules);
    if (isRequired) {
        finalRules.required = true;
    }
    if (minLength) {
        finalRules.minLength = minLength;
    }
    if (maxLength) {
        finalRules.maxLength = maxLength;
    }
    finalRules.validate = React__default["default"].useCallback(function (items) { return tslib.__awaiter(_this, void 0, void 0, function () {
        var map, result, _a, _b, key, valid, e_1_1;
        var e_1, _c;
        return tslib.__generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    map = subForms.current;
                    if (!(typeof (rules === null || rules === void 0 ? void 0 : rules.validate) === 'function')) return [3 /*break*/, 2];
                    return [4 /*yield*/, rules.validate(items)];
                case 1:
                    result = _d.sent();
                    if (result) {
                        return [2 /*return*/, result];
                    }
                    _d.label = 2;
                case 2:
                    _d.trys.push([2, 7, 8, 9]);
                    _a = tslib.__values(Object.keys(map)), _b = _a.next();
                    _d.label = 3;
                case 3:
                    if (!!_b.done) return [3 /*break*/, 6];
                    key = _b.value;
                    return [4 /*yield*/, (function (methods) {
                            return new Promise(function (resolve) {
                                methods.handleSubmit(function () { return resolve(true); }, function () { return resolve(false); })();
                            });
                        })(map[key])];
                case 4:
                    valid = _d.sent();
                    if (!valid) {
                        return [2 /*return*/, __('validateFailed')];
                    }
                    _d.label = 5;
                case 5:
                    _b = _a.next();
                    return [3 /*break*/, 3];
                case 6: return [3 /*break*/, 9];
                case 7:
                    e_1_1 = _d.sent();
                    e_1 = { error: e_1_1 };
                    return [3 /*break*/, 9];
                case 8:
                    try {
                        if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                    }
                    finally { if (e_1) throw e_1.error; }
                    return [7 /*endfinally*/];
                case 9: return [2 /*return*/];
            }
        });
    }); }, [subForms]);
    var _d = reactHookForm.useFieldArray({
        control: control,
        name: name,
        rules: finalRules
    }), fields = _d.fields, append = _d.append; _d.update; var remove = _d.remove;
    if (!Array.isArray(columns)) {
        columns = [];
    }
    var errors = reactHookForm.useFormState({
        control: control
    }).errors;
    var trigger = reactHookForm.useFormContext().trigger;
    // useFieldArray 的 update 会更新行 id，导致重新渲染
    // 正在编辑中的元素失去焦点，所以自己写一个
    var lightUpdate = React__default["default"].useCallback(function (index, value) {
        var arr = control._getFieldArray(name);
        arr[index] = tslib.__assign({}, value);
        control._updateFieldArray(name, arr);
        trigger(name);
        control._subjects.watch.next({});
    }, [control]);
    function renderBody() {
        return (React__default["default"].createElement("div", { className: cx("Table", className) },
            React__default["default"].createElement("div", { className: cx("Table-contentWrap") },
                React__default["default"].createElement("table", { className: cx("Table-table") },
                    React__default["default"].createElement("thead", null,
                        React__default["default"].createElement("tr", null,
                            columns.map(function (item, index) { return (React__default["default"].createElement("th", { key: index, className: item.className }, item.thRender ? item.thRender() : item.title)); }),
                            React__default["default"].createElement("th", { key: "operation" }, __('Table.operation')))),
                    React__default["default"].createElement("tbody", null, fields.length ? (fields.map(function (field, index) { return (React__default["default"].createElement("tr", { key: field.id },
                        React__default["default"].createElement(InputTableRow, { key: "columns", control: control, update: lightUpdate, index: index, value: field, columns: columns, translate: __, classnames: cx, formRef: subFormRef }),
                        React__default["default"].createElement("td", { key: "operation" },
                            React__default["default"].createElement(Button["default"], { level: "link", key: "delete", disabled: removable === false ||
                                    !!(minLength && fields.length <= minLength), className: cx('Table-delBtn'), onClick: function () { return remove(index); } }, __('delete'))))); })) : (React__default["default"].createElement("tr", null,
                        React__default["default"].createElement("td", { colSpan: columns.length + 1 },
                            React__default["default"].createElement(icons.Icon, { icon: "desk-empty", className: cx('Table-placeholder-empty-icon', 'icon') }),
                            __('placeholder.noData'))))))),
            addable !== false && (!maxLength || fields.length < maxLength) ? (React__default["default"].createElement("div", { className: cx("InputTable-toolbar") },
                React__default["default"].createElement(Button["default"], { className: cx(addButtonClassName), onClick: function () { return append(tslib.__assign({}, scaffold)); }, size: "sm" },
                    React__default["default"].createElement(icons.Icon, { icon: "plus", className: "icon" }),
                    React__default["default"].createElement("span", null, __(addButtonText || 'add'))))) : null));
    }
    return wrap === false ? (renderBody()) : (React__default["default"].createElement(FormField["default"], { className: fieldClassName, label: label, labelAlign: labelAlign, labelClassName: labelClassName, description: description, mode: mode, hasError: !!((_b = errors[name]) === null || _b === void 0 ? void 0 : _b.message), errors: (_c = errors[name]) === null || _c === void 0 ? void 0 : _c.message }, renderBody()));
}
function InputTableRow(_a) {
    var value = _a.value, columns = _a.columns, index = _a.index, translate = _a.translate, update = _a.update, formRef = _a.formRef; _a.classnames;
    var methods = useSubForm["default"](value, translate, function (data) { return update(index, data); });
    React__default["default"].useEffect(function () {
        formRef === null || formRef === void 0 ? void 0 : formRef(methods, index);
        return function () {
            formRef === null || formRef === void 0 ? void 0 : formRef(null, index);
        };
    }, [methods]);
    return (React__default["default"].createElement(React__default["default"].Fragment, null, columns.map(function (item, index) { return (React__default["default"].createElement("td", { key: index, className: item.className }, item.tdRender(methods, index))); })));
}
var InputTable$1 = amisCore.themeable(amisCore.localeable(InputTable));

exports.InputTable = InputTable;
exports.InputTableRow = InputTableRow;
exports["default"] = InputTable$1;
