// amis-ui/lib/components/Layout.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var amisCore = require('amis-core');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

/**
 * @file Layout
 * @description 页面布局，支持左边栏、顶部、内容区域布局。
 * @author fex
 *
 * @param 参数说明：
 * * children 会渲染在内容区。
 * * header 头部区域
 * * aside 边栏
 * * asideClassName 边栏附加样式class
 * * footer 页脚
 * * folder 是否收起边栏
 * * asideFixed 边栏是否为固定模式，如果是会用 position:fixed 来定位.
 * * className 附件的样式名
 * * contentClassName 内容区域附加样式名称
 */
function Layout(_a) {
    var _b;
    var header = _a.header, headerClassName = _a.headerClassName, aside = _a.aside, asideClassName = _a.asideClassName, children = _a.children, className = _a.className, contentClassName = _a.contentClassName, folded = _a.folded, asideFixed = _a.asideFixed, headerFixed = _a.headerFixed, footer = _a.footer, offScreen = _a.offScreen, size = _a.size, boxed = _a.boxed, cx = _a.classnames, bodyClassName = _a.bodyClassName;
    var body = (React__default["default"].createElement("div", { className: cx("Layout-body", contentClassName) }, children));
    if (aside) {
        body = (React__default["default"].createElement("div", { className: cx('Layout-content'), role: "main" }, body));
    }
    React__default["default"].useEffect(function () {
        bodyClassName && document.body.classList.add(bodyClassName);
        return function () {
            bodyClassName && document.body.classList.remove(bodyClassName);
        };
    }, [bodyClassName]);
    return (React__default["default"].createElement("div", { className: cx("Layout", className, (_b = {
                'Layout--boxed': boxed,
                'Layout--withAside': !!aside,
                'Layout--headerFixed': header ? headerFixed : false,
                'Layout--asideFixed': aside ? asideFixed : false,
                'Layout--folded': folded,
                'Layout--offScreen': offScreen
            },
            _b["Layout--".concat(size)] = size,
            _b['Layout--noFooter'] = !footer,
            _b)) },
        header ? (React__default["default"].createElement("div", { className: cx('Layout-header', headerClassName) }, header)) : null,
        aside ? (React__default["default"].createElement("div", { className: cx("Layout-aside", asideClassName) },
            React__default["default"].createElement("div", { className: cx('Layout-asideWrap') },
                React__default["default"].createElement("div", { id: "asideInner", className: cx('Layout-asideInner') }, aside)))) : null,
        body,
        footer ? (React__default["default"].createElement("footer", { className: cx('Layout-footer'), role: "footer" }, footer)) : null));
}
Layout.defaultProps = {
    // asideWide: false,
    asideFixed: true,
    asideClassName: '',
    headerFixed: true,
    offScreen: false,
    footer: false
};
var Layout$1 = amisCore.themeable(Layout);

exports.Layout = Layout;
exports["default"] = Layout$1;
