// amis-ui/lib/components/Radios.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var React = require('react');
var amisCore = require('amis-core');
var Checkbox = require('./Checkbox.js');
var Button = require('./Button.js');
var Select = require('./Select.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

/**
 * @file Radios
 * @description
 * @author fex
 *
 * @param 参数说明：
 * options: [
 *   {
 *      label: '显示的名字',
 *      value: '值',
 *      disabled: false
 *   }
 * ]
 *
 */
var Radios = /** @class */ (function (_super) {
    tslib.__extends(Radios, _super);
    function Radios() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Radios.prototype.toggleOption = function (option) {
        var _a = this.props, value = _a.value, onChange = _a.onChange, valueField = _a.valueField, clearable = _a.clearable, delimiter = _a.delimiter, options = _a.options;
        var valueArray = Select.value2array(value, {
            multiple: false,
            delimiter: delimiter,
            valueField: valueField,
            options: options
        });
        var idx = valueArray.indexOf(option);
        if (~idx) {
            clearable && valueArray.splice(idx, 1);
        }
        else {
            valueArray = [option];
        }
        var newValue = valueArray[0];
        onChange && onChange(newValue);
    };
    Radios.prototype.renderGroup = function (option, index, valueArray) {
        var _this = this;
        var _a = this.props, cx = _a.classnames; _a.optionType; _a.classPrefix;
        return (React__default["default"].createElement("div", { key: index, className: cx('RadiosControl-group', option.className) },
            React__default["default"].createElement("label", { className: cx('RadiosControl-groupLabel', option.labelClassName) }, option.label),
            option.children && option.children.length
                ? option.children.map(function (option, index) {
                    return _this.renderItem(option, index, valueArray);
                })
                : null));
    };
    Radios.prototype.renderItem = function (option, index, valueArray) {
        var _this = this;
        if (option.children) {
            return this.renderGroup(option, index, valueArray);
        }
        var _a = this.props, disabled = _a.disabled, inline = _a.inline, itemClassName = _a.itemClassName, cx = _a.classnames, labelClassName = _a.labelClassName, labelField = _a.labelField, optionType = _a.optionType, level = _a.level, btnActiveLevel = _a.btnActiveLevel; _a.classPrefix;
        if (optionType === 'button') {
            var active = !!~valueArray.indexOf(option);
            return (React__default["default"].createElement(Button["default"], { key: index, active: active, onClick: function () { return _this.toggleOption(option); }, className: cx(itemClassName, option.className), disabled: disabled || option.disabled, level: (active ? btnActiveLevel : '') || level },
                React__default["default"].createElement("span", null, "".concat(option[labelField || 'label']))));
        }
        return (React__default["default"].createElement(Checkbox["default"], { type: "radio", key: index, onChange: function () { return _this.toggleOption(option); }, checked: !!~valueArray.indexOf(option), className: cx(itemClassName, option.className), disabled: disabled || option.disabled, description: option.description, inline: inline, labelClassName: labelClassName }, "".concat(option[labelField || 'label'])));
    };
    Radios.prototype.render = function () {
        var _this = this;
        var _a = this.props, value = _a.value, options = _a.options, className = _a.className, style = _a.style, cx = _a.classnames, placeholder = _a.placeholder, columnsCount = _a.columnsCount; _a.joinValues; _a.extractValue; _a.disabled; var inline = _a.inline, delimiter = _a.delimiter, valueField = _a.valueField;
        var valueArray = Select.value2array(value, {
            multiple: false,
            delimiter: delimiter,
            valueField: valueField,
            options: options
        });
        var body = [];
        if (options) {
            body = options.map(function (option, key) {
                return _this.renderItem(option, key, valueArray);
            });
        }
        if (!inline) {
            body = amisCore.columnsSplit(body, cx, columnsCount);
        }
        return (React__default["default"].createElement("div", { className: className, style: style }, body && body.length ? body : placeholder));
    };
    Radios.defaultProps = {
        type: 'radio',
        optionType: 'default',
        btnActiveLevel: 'primary',
        resetValue: '',
        inline: true,
        joinValues: true,
        clearable: false,
        columnsCount: 1 // 一行显示一个
    };
    return Radios;
}(React__default["default"].Component));
var Radios$1 = amisCore.themeable(amisCore.uncontrollable(Radios, {
    value: 'onChange'
}));

exports.Radios = Radios;
exports["default"] = Radios$1;
