// amis-ui/lib/components/TableSelection.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var Selection = require('./Selection.js');
var amisCore = require('amis-core');
var React = require('react');
var Checkbox = require('./Checkbox.js');
var index = require('./virtual-list/index.js');
var _ = require('lodash');
var AutoSizer = require('./virtual-list/AutoSizer.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var TableSelection = /** @class */ (function (_super) {
    tslib.__extends(TableSelection, _super);
    function TableSelection(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            rowRenderScope: null,
            colsWidth: [],
            tableWidth: 0
        };
        return _this;
    }
    TableSelection.prototype.getColumns = function () {
        var columns = this.props.columns;
        if (!Array.isArray(columns) || !columns.length) {
            columns = [{ label: 'Label', name: 'label' }];
        }
        return columns;
    };
    TableSelection.prototype.renderTHead = function () {
        var _a = this.props, options = _a.options, cx = _a.classnames, value = _a.value, disabled = _a.disabled, option2value = _a.option2value, multiple = _a.multiple;
        var columns = this.getColumns();
        var valueArray = Selection.BaseSelection.value2array(value, options, option2value);
        var availableOptions = options.filter(function (option) { return !option.disabled; });
        var partialChecked = false;
        var allChecked = !!availableOptions.length;
        availableOptions.forEach(function (option) {
            var isIn = !!~valueArray.indexOf(option);
            if (isIn && !partialChecked) {
                partialChecked = true;
            }
            else if (!isIn && allChecked) {
                allChecked = false;
            }
        });
        return (React__default["default"].createElement(React__default["default"].Fragment, null,
            React__default["default"].createElement("thead", null,
                React__default["default"].createElement("tr", null,
                    multiple && Array.isArray(options) && options.length ? (React__default["default"].createElement("th", { className: cx('Table-checkCell') },
                        React__default["default"].createElement(Checkbox["default"], { key: "checkbox", size: "sm", disabled: disabled, onChange: this.toggleAll, checked: partialChecked, partial: partialChecked && !allChecked }))) : null,
                    columns.map(function (column, index) { return (React__default["default"].createElement("th", { key: index }, column.label)); })))));
    };
    TableSelection.prototype.renderTr = function (_a) {
        var _this = this;
        var option = _a.option, rowIndex = _a.rowIndex, valueArray = _a.valueArray, columns = _a.columns, styles = _a.styles;
        var _b = this.props, cx = _b.classnames, cellRender = _b.cellRender, disabled = _b.disabled, multiple = _b.multiple; _b.translate; var itemClassName = _b.itemClassName, resultMode = _b.resultMode;
        var checked = valueArray.indexOf(option) !== -1;
        return (React__default["default"].createElement("tr", { style: styles !== null && styles !== void 0 ? styles : {}, key: rowIndex, 
            /** 被ResultTableList引用，如果设置click事件，会导致错误删除结果列表的内容，先加一个开关判断 */
            onClick: resultMode
                ? amisCore.noop
                : function (e) { return e.defaultPrevented || _this.toggleOption(option); }, className: cx(itemClassName, option.className, disabled || option.disabled ? 'is-disabled' : '', !!~valueArray.indexOf(option) ? 'is-active' : '') },
            multiple ? (React__default["default"].createElement("td", { className: cx('Table-checkCell'), key: "checkbox", onClick: function (e) {
                    e.stopPropagation();
                    _this.toggleOption(option);
                } },
                React__default["default"].createElement(Checkbox["default"], { size: "sm", checked: checked, disabled: disabled }))) : null,
            columns.map(function (column, colIndex) { return (React__default["default"].createElement("td", { key: colIndex }, cellRender(column, option, colIndex, rowIndex))); })));
    };
    TableSelection.prototype.renderTBody = function () {
        var _this = this;
        var _a = this.props, options = _a.options, placeholder = _a.placeholder, value = _a.value, option2value = _a.option2value, __ = _a.translate;
        var columns = this.getColumns();
        var valueArray = Selection.BaseSelection.value2array(value, options, option2value);
        return (React__default["default"].createElement("tbody", null, Array.isArray(options) && options.length ? (options.map(function (option, rowIndex) {
            return _this.renderTr({ option: option, rowIndex: rowIndex, valueArray: valueArray, columns: columns });
        })) : (React__default["default"].createElement("tr", null,
            React__default["default"].createElement("td", { colSpan: columns.length }, __(placeholder))))));
    };
    TableSelection.prototype.tableHeadRef = function (ref) {
        ref && (this.ref = ref);
    };
    TableSelection.prototype.handleVirtualTableResize = function (_a) {
        var width = _a.width;
        if (width && width === this.state.width) {
            return;
        }
        var widths = {};
        this.ref &&
            _.forEach(this.ref.querySelectorAll('thead>tr:last-child>th'), function (item, index) {
                widths[index] = item.getBoundingClientRect().width;
            });
        var colsWidth = [];
        Object.keys(widths)
            .filter(function (key) { return !isNaN(Number(key)); })
            .sort()
            .forEach(function (key) {
            colsWidth.push(widths[key]);
        });
        this.setState({ colsWidth: colsWidth, tableWidth: width });
    };
    TableSelection.prototype.renderVirtualTable = function () {
        var _this = this;
        var _a = this.props, options = _a.options, value = _a.value, cx = _a.classnames, option2value = _a.option2value; _a.translate; var _b = _a.itemHeight, itemHeight = _b === void 0 ? 30 : _b, virtualListHeight = _a.virtualListHeight;
        var columns = this.getColumns();
        var valueArray = Selection.BaseSelection.value2array(value, options, option2value);
        var _c = this.state.rowRenderScope || {}, _d = _c.startIndex, startIndex = _d === void 0 ? 0 : _d, _e = _c.stopIndex, stopIndex = _e === void 0 ? 10 : _e;
        var tableList = null;
        if (startIndex !== undefined && stopIndex !== undefined) {
            var trs = [];
            for (var index$1 = startIndex; index$1 <= stopIndex; index$1++) {
                var option = options[index$1];
                if (!option) {
                    break;
                }
                trs.push(this.renderTr({
                    option: option,
                    rowIndex: index$1,
                    valueArray: valueArray,
                    columns: columns,
                    styles: {
                        height: "".concat(itemHeight, "px")
                    }
                }));
            }
            tableList = (React__default["default"].createElement("table", { className: cx('Table-table'), style: {
                    marginTop: (startIndex || 0) * itemHeight + 'px'
                } },
                this.state.colsWidth.length ? (React__default["default"].createElement("colgroup", null, this.state.colsWidth.map(function (colWidth, index) { return (React__default["default"].createElement("col", { style: { width: "".concat(colWidth, "px") }, key: "col-".concat(index) })); }))) : null,
                React__default["default"].createElement("tbody", null, trs)));
        }
        return (React__default["default"].createElement("div", { className: cx('Table-content', 'is-virtual') },
            React__default["default"].createElement("table", { className: cx('Table-table'), ref: this.tableHeadRef.bind(this) }, this.renderTHead()),
            React__default["default"].createElement("div", { className: cx('Table-content-virtual') },
                React__default["default"].createElement(AutoSizer["default"], { minHeight: virtualListHeight, onResize: this.handleVirtualTableResize.bind(this) }, function (_a) {
                    var height = _a.height;
                    return (React__default["default"].createElement(index["default"], { onItemsRendered: function (res) {
                            if (!_.isEqual(_this.state.rowRenderScope, res)) {
                                // 需要延后执行，否则报 warning
                                setTimeout(function () {
                                    _this.setState({
                                        rowRenderScope: res
                                    });
                                });
                            }
                        }, height: height, itemCount: options.length, itemSize: itemHeight, WrapperComponent: "div", InnerComponent: "div", prefix: tableList, innerStyleFilter: function (styles) { return (tslib.__assign(tslib.__assign({}, styles), { position: 'absolute', top: 0, minWidth: undefined, width: '1px', visibility: 'hidden' })); }, renderItem: function () { return null; } }));
                }))));
    };
    TableSelection.prototype.render = function () {
        var _a = this.props, className = _a.className, cx = _a.classnames, options = _a.options, _b = _a.virtualThreshold, virtualThreshold = _b === void 0 ? 1000 : _b;
        var table = Array.isArray(options) && options.length > virtualThreshold ? (this.renderVirtualTable()) : (React__default["default"].createElement("div", { className: cx('Table-content') },
            React__default["default"].createElement("table", { className: cx('Table-table') },
                this.renderTHead(),
                this.renderTBody())));
        return React__default["default"].createElement("div", { className: cx('TableSelection', className) }, table);
    };
    TableSelection.defaultProps = tslib.__assign(tslib.__assign({}, Selection.BaseSelection.defaultProps), { cellRender: function (column, option, colIndex, rowIndex) { return React__default["default"].createElement("span", null, amisCore.resolveVariable(column.name, option)); } });
    return TableSelection;
}(Selection.BaseSelection));
var TableCheckboxes = amisCore.themeable(amisCore.localeable(amisCore.uncontrollable(TableSelection, {
    value: 'onChange'
})));

exports.TableSelection = TableSelection;
exports["default"] = TableCheckboxes;
