// office-viewer/esm/util/dom.js
import { __spreadArray, __read } from '../node_modules/tslib/tslib.es6.js';

/**
 * dom 相关的操作
 */
/**
 * 样式对象转成 css 文本
 */
function styleToText(style) {
    if (style === void 0) { style = {}; }
    var text = '';
    for (var key in style) {
        var value = style[key];
        if (value != null && value !== '') {
            text += "".concat(key, ": ").concat(value, ";\n");
        }
    }
    return text;
}
/**
 * 设置元素样式
 */
function setStyle(el, style) {
    for (var key in style) {
        var value = style[key];
        if (value != null && value !== '') {
            el.style.setProperty(key, style[key]);
        }
    }
}
/**
 * 创建元素
 */
function createElement(tagName) {
    return document.createElement(tagName);
}
/**
 * 添加子节点，会做一些判断避免报错
 */
function appendChild(parent, child) {
    if (parent && child) {
        parent.appendChild(child);
    }
}
/**
 * 添加 css 类
 */
function addClassName(el, className) {
    if (el && className) {
        el.classList.add(className);
    }
}
/**
 * 批量添加 css 类
 */
function addClassNames(el, classNames) {
    var _a;
    if (el && classNames) {
        (_a = el.classList).add.apply(_a, __spreadArray([], __read(classNames), false));
    }
}

export { addClassName, addClassNames, appendChild, createElement, setStyle, styleToText };
