// amis/lib/renderers/OfficeViewer.js
/**
 * amis v2.9.0
 * Copyright 2018-2023 baidu
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var React = require('react');
var amisCore = require('amis-core');
var amisUi = require('amis-ui');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

/**
 * office 文件预览
 */
var OfficeViewer = /** @class */ (function (_super) {
    tslib.__extends(OfficeViewer, _super);
    function OfficeViewer(props) {
        var _this = _super.call(this, props) || this;
        _this.rootElement = React__default["default"].createRef();
        return _this;
    }
    OfficeViewer.prototype.componentDidMount = function () {
        var _a;
        if ((_a = this.rootElement) === null || _a === void 0 ? void 0 : _a.current) {
            this.renderWord();
        }
    };
    OfficeViewer.prototype.componentDidUpdate = function (prevProps) {
        var _a;
        var props = this.props;
        if (amisCore.isApiOutdated(prevProps.src, props.src, prevProps.data, props.data)) {
            this.renderWord();
        }
        if (props.name) {
            if (prevProps.data[props.name] !== props.data[props.name]) {
                this.renderWord();
            }
        }
        // 这个变量替换只会更新变化的部分，所以性能还能接受
        (_a = this.word) === null || _a === void 0 ? void 0 : _a.updateVariable();
    };
    /**
     * 接收动作事件
     */
    OfficeViewer.prototype.doAction = function (action, args, throwErrors) {
        var _a, _b;
        var actionType = action === null || action === void 0 ? void 0 : action.actionType;
        if (actionType === 'saveAs') {
            (_a = this.word) === null || _a === void 0 ? void 0 : _a.download((args === null || args === void 0 ? void 0 : args.name) || this.fileName);
        }
        if (actionType === 'print') {
            (_b = this.word) === null || _b === void 0 ? void 0 : _b.print();
        }
    };
    /**
     * 执行变量替换
     */
    OfficeViewer.prototype.evalVar = function (text, data) {
        var localData = this.props.data;
        return amisCore.resolveVariable(text, amisCore.createObject(localData, data));
    };
    OfficeViewer.prototype.renderWord = function () {
        return tslib.__awaiter(this, void 0, void 0, function () {
            var _a, src, name;
            return tslib.__generator(this, function (_b) {
                _a = this.props, src = _a.src, name = _a.name;
                if (src) {
                    this.renderRemoteWord();
                }
                else if (name) {
                    this.renderFormFile();
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * 渲染远端文件
     */
    OfficeViewer.prototype.renderRemoteWord = function () {
        return tslib.__awaiter(this, void 0, void 0, function () {
            var _a, wordOptions, env, src, data, display, finalSrc, response;
            var _this = this;
            return tslib.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this.props, wordOptions = _a.wordOptions, env = _a.env, src = _a.src, data = _a.data, display = _a.display;
                        finalSrc = src
                            ? amisCore.resolveVariableAndFilter(src, data, '| raw')
                            : undefined;
                        if (typeof finalSrc === 'string') {
                            this.fileName = finalSrc.split('/').pop();
                        }
                        if (!finalSrc) {
                            console.warn('file src is empty');
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, env.fetcher(finalSrc, data, {
                                responseType: 'arraybuffer'
                            })];
                    case 1:
                        response = _b.sent();
                        Promise.resolve().then(function() {return new Promise(function(fullfill) {require(['office-viewer'], function(mod) {fullfill(tslib.__importStar(mod))})})}).then(function (officeViewer) { return tslib.__awaiter(_this, void 0, void 0, function () {
                            var Word, word;
                            var _a;
                            return tslib.__generator(this, function (_b) {
                                Word = officeViewer.Word;
                                word = new Word(response.data, tslib.__assign(tslib.__assign({}, wordOptions), { data: data, evalVar: this.evalVar.bind(this) }));
                                if (display !== false) {
                                    word.render((_a = this.rootElement) === null || _a === void 0 ? void 0 : _a.current);
                                }
                                this.word = word;
                                return [2 /*return*/];
                            });
                        }); });
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 渲染本地文件，用于预览 input-file
     */
    OfficeViewer.prototype.renderFormFile = function () {
        var _this = this;
        var _a = this.props, wordOptions = _a.wordOptions, name = _a.name, data = _a.data, display = _a.display;
        var file = data[name];
        if (file instanceof File) {
            var reader_1 = new FileReader();
            reader_1.onload = function (_e) {
                var data = reader_1.result;
                Promise.resolve().then(function() {return new Promise(function(fullfill) {require(['office-viewer'], function(mod) {fullfill(tslib.__importStar(mod))})})}).then(function (officeViewer) { return tslib.__awaiter(_this, void 0, void 0, function () {
                    var Word, word;
                    var _a;
                    return tslib.__generator(this, function (_b) {
                        Word = officeViewer.Word;
                        word = new Word(data, tslib.__assign(tslib.__assign({}, wordOptions), { evalVar: this.evalVar.bind(this) }));
                        if (display !== false) {
                            word.render((_a = this.rootElement) === null || _a === void 0 ? void 0 : _a.current);
                        }
                        return [2 /*return*/];
                    });
                }); });
            };
            reader_1.readAsArrayBuffer(file);
        }
    };
    OfficeViewer.prototype.render = function () {
        var _a = this.props, cx = _a.classnames; _a.translate; var className = _a.className, _b = _a.loading, loading = _b === void 0 ? false : _b, src = _a.src, name = _a.name, loadingConfig = _a.loadingConfig;
        return (React__default["default"].createElement("div", { ref: this.rootElement, className: cx('Office-Viewer', className) },
            (!src || !name) && (React__default["default"].createElement("svg", { width: "100%", height: "100", xmlns: "http://www.w3.org/2000/svg" },
                React__default["default"].createElement("rect", { x: "0", y: "0", width: "100%", height: "100", style: { fill: '#F7F7F9' } }),
                React__default["default"].createElement("text", { x: "50%", y: "50%", fontSize: "18", textAnchor: "middle", alignmentBaseline: "middle", fontFamily: "monospace, sans-serif", fill: "#555555" }, "office viewer"))),
            React__default["default"].createElement(amisUi.Spinner, { overlay: true, key: "info", show: loading, loadingConfig: loadingConfig })));
    };
    return OfficeViewer;
}(React__default["default"].Component));
/** @class */ ((function (_super) {
    tslib.__extends(OfficeViewerRenderer, _super);
    function OfficeViewerRenderer(props, context) {
        var _this = _super.call(this, props) || this;
        var scoped = context;
        scoped.registerComponent(_this);
        return _this;
    }
    OfficeViewerRenderer.prototype.componentWillUnmount = function () {
        var _a;
        (_a = _super.prototype.componentWillUnmount) === null || _a === void 0 ? void 0 : _a.call(this);
        var scoped = this.context;
        scoped.unRegisterComponent(this);
    };
    OfficeViewerRenderer.contextType = amisCore.ScopedContext;
    OfficeViewerRenderer = tslib.__decorate([
        amisCore.Renderer({
            type: 'office-viewer'
        }),
        tslib.__metadata("design:paramtypes", [Object, Object])
    ], OfficeViewerRenderer);
    return OfficeViewerRenderer;
})(OfficeViewer));

exports["default"] = OfficeViewer;
