// amis-ui/lib/components/calendar/DaysView.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var moment = require('moment');
var React = require('react');
var Downshift = require('downshift');
var findIndex = require('lodash/findIndex');
var extend = require('lodash/extend');
var merge = require('lodash/merge');
var omit = require('lodash/omit');
var each = require('lodash/each');
var kebabCase = require('lodash/kebabCase');
var amisCore = require('amis-core');
var Picker = require('../Picker.js');
var icons = require('../icons.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var moment__default = /*#__PURE__*/_interopDefaultLegacy(moment);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var Downshift__default = /*#__PURE__*/_interopDefaultLegacy(Downshift);
var findIndex__default = /*#__PURE__*/_interopDefaultLegacy(findIndex);
var extend__default = /*#__PURE__*/_interopDefaultLegacy(extend);
var merge__default = /*#__PURE__*/_interopDefaultLegacy(merge);
var omit__default = /*#__PURE__*/_interopDefaultLegacy(omit);
var each__default = /*#__PURE__*/_interopDefaultLegacy(each);
var kebabCase__default = /*#__PURE__*/_interopDefaultLegacy(kebabCase);

// 最早基于 react-datetime 2.16.2 版本，后来大部分都自己写了
var CustomDaysView = /** @class */ (function (_super) {
    tslib.__extends(CustomDaysView, _super);
    function CustomDaysView(props) {
        var _this = _super.call(this, props) || this;
        _this.timeConstraints = {
            hours: {
                min: 0,
                max: 23,
                step: 1
            },
            minutes: {
                min: 0,
                max: 59,
                step: 1
            },
            seconds: {
                min: 0,
                max: 59,
                step: 1
            },
            milliseconds: {
                min: 0,
                max: 999,
                step: 1
            }
        };
        _this.updateSelectedDate = function (event) {
            // need confirm
            if (_this.props.requiredConfirm) {
                var viewDate = _this.props.viewDate.clone();
                var currentDate = _this.props.selectedDate || viewDate;
                var target = event.target;
                var modifier = 0;
                if (~target.className.indexOf('rdtNew')) {
                    modifier = 1;
                }
                if (~target.className.indexOf('rdtOld')) {
                    modifier = -1;
                }
                viewDate
                    .month(viewDate.month() + modifier)
                    .date(parseInt(target.getAttribute('data-value'), 10))
                    .hours(currentDate.hours())
                    .minutes(currentDate.minutes())
                    .seconds(currentDate.seconds())
                    .milliseconds(currentDate.milliseconds());
                _this.props.setDateTimeState({
                    viewDate: viewDate,
                    selectedDate: viewDate.clone()
                });
                return;
            }
            _this.props.updateSelectedDate(event, true);
        };
        _this.showTime = function () {
            var _a = _this.props, selectedDate = _a.selectedDate, viewDate = _a.viewDate, timeFormat = _a.timeFormat;
            return (React__default["default"].createElement("div", { key: "stb", className: "rdtShowTime" }, (selectedDate || viewDate || moment__default["default"]()).format(timeFormat)));
        };
        _this.setTime = function (type, value) {
            var date = (_this.props.selectedDate || _this.props.viewDate).clone();
            date[type](value);
            _this.props.setDateTimeState({
                viewDate: date.clone(),
                selectedDate: date.clone()
            });
            if (!_this.props.requiredConfirm) {
                _this.props.onChange(date);
            }
        };
        _this.scrollToTop = function (type, value, i, label) {
            var _a, _b;
            var elf = document.getElementById("".concat(_this.state.uniqueTag, "-").concat(i, "-input"));
            var _c = _this.timeConstraints[type], min = _c.min, step = _c.step;
            var offset = (value - min) / step;
            var height = 28; /** 单个选项的高度 */
            (_b = (_a = elf === null || elf === void 0 ? void 0 : elf.parentNode) === null || _a === void 0 ? void 0 : _a.scrollTo) === null || _b === void 0 ? void 0 : _b.call(_a, {
                top: offset * height,
                behavior: label === 'init' ? 'auto' : 'smooth'
            });
        };
        _this.confirm = function () {
            var _a, _b;
            var date = (_this.props.selectedDate || _this.props.viewDate).clone();
            // 如果 minDate 是可用的，且比当前日期晚，则用 minDate
            if (((_a = _this.props.minDate) === null || _a === void 0 ? void 0 : _a.isValid()) && ((_b = _this.props.minDate) === null || _b === void 0 ? void 0 : _b.isAfter(date))) {
                date = _this.props.minDate.clone();
            }
            _this.props.setDateTimeState({
                selectedDate: date
            });
            _this.props.onChange(date);
            _this.props.onClose && _this.props.onClose();
        };
        _this.cancel = function () {
            _this.props.onClose && _this.props.onClose();
        };
        _this.renderDay = function (props, currentDate) {
            var todayActiveStyle = props.todayActiveStyle; /** 只有today才会传入这个属性 */
            var _a = _this.props, cx = _a.classnames, __ = _a.translate;
            var injectedProps = omit__default["default"](props, ['todayActiveStyle']);
            /** 某些情况下需要用inline style覆盖动态class，需要hack important的样式 */
            var todayDomRef = function (node) {
                if (todayActiveStyle && node) {
                    each__default["default"](todayActiveStyle, function (value, key) {
                        var _a, _b;
                        if (typeof value === 'string' && !!~value.indexOf('!important')) {
                            (_b = (_a = node === null || node === void 0 ? void 0 : node.style) === null || _a === void 0 ? void 0 : _a.setProperty) === null || _b === void 0 ? void 0 : _b.call(_a, kebabCase__default["default"](key), String(value)
                                .replace(/\!important/, '')
                                .trim(), 'important');
                        }
                    });
                }
            };
            if (_this.props.schedules) {
                var schedule_1 = [];
                _this.props.schedules.forEach(function (item) {
                    /** 时间全部统一到当天的00:00:00再做比较 */
                    var currentDateBegin = currentDate.startOf('day');
                    var startTime = moment__default["default"](item.startTime).startOf('day');
                    var endTime = moment__default["default"](item.endTime).startOf('day');
                    if (currentDateBegin.isSameOrAfter(startTime) &&
                        currentDateBegin.isSameOrBefore(endTime)) {
                        schedule_1.push(item);
                    }
                });
                if (schedule_1.length > 0) {
                    // 日程数据
                    var scheduleData_1 = {
                        scheduleData: schedule_1.map(function (item) {
                            return tslib.__assign(tslib.__assign({}, item), { time: moment__default["default"](item.startTime).format('YYYY-MM-DD HH:mm:ss') +
                                    ' - ' +
                                    moment__default["default"](item.endTime).format('YYYY-MM-DD HH:mm:ss') });
                        }),
                        currentDate: currentDate
                    };
                    // 放大模式
                    if (_this.props.largeMode) {
                        var showSchedule_1 = [];
                        for (var i = 0; i < schedule_1.length; i++) {
                            if (showSchedule_1.length > 3) {
                                break;
                            }
                            if (moment__default["default"](schedule_1[i].startTime).isSame(currentDate, 'day')) {
                                showSchedule_1.push(schedule_1[i]);
                            }
                            else if (currentDate.weekday() === 0) {
                                var width = Math.min(moment__default["default"](schedule_1[i].endTime).diff(currentDate, 'days') + 1, 7);
                                // 周一重新设置日程
                                showSchedule_1.push(tslib.__assign(tslib.__assign({}, schedule_1[i]), { width: width, startTime: moment__default["default"](currentDate), endTime: moment__default["default"](currentDate).add(width - 1, 'days') }));
                                schedule_1[i].height === undefined && (schedule_1[i].height = 0);
                            }
                            else {
                                // 生成空白格占位
                                showSchedule_1.push({
                                    width: 1,
                                    className: 'bg-transparent',
                                    content: '',
                                    height: schedule_1[i].height
                                });
                            }
                        }
                        [0, 1, 2].forEach(function (i) {
                            // 排序
                            var tempIndex = findIndex__default["default"](showSchedule_1, function (item) { return item.height === i; });
                            if (tempIndex === -1) {
                                tempIndex = findIndex__default["default"](showSchedule_1, function (item) { return item.height === undefined; });
                            }
                            if (tempIndex > -1 && tempIndex !== i) {
                                var temp = showSchedule_1[i];
                                showSchedule_1[i] = showSchedule_1[tempIndex];
                                showSchedule_1[tempIndex] = temp;
                            }
                            if (showSchedule_1[i] && showSchedule_1[i].height === undefined) {
                                showSchedule_1[i].height = i;
                            }
                        });
                        // 最多展示3个
                        showSchedule_1 = showSchedule_1.slice(0, 3);
                        var scheduleDiv = showSchedule_1.map(function (item, index) {
                            var diffDays = moment__default["default"](item.endTime).diff(moment__default["default"](item.startTime), 'days');
                            /* 存在天数跨度小于1，但是横跨2天的case */
                            if (diffDays <= 0) {
                                diffDays = moment__default["default"](item.endTime)
                                    .endOf('day')
                                    .diff(moment__default["default"](item.startTime).startOf('day'), 'days');
                            }
                            /* 前面的计算结果是闭区间，所以最终结果要补足1 */
                            diffDays += 1;
                            var width = item.width ||
                                Math.min(diffDays, 7 - moment__default["default"](item.startTime).weekday());
                            return (React__default["default"].createElement("div", { key: injectedProps.key + 'content' + index, className: cx('ScheduleCalendar-large-schedule-content', item.className), style: { width: width + '00%' }, onClick: function () {
                                    return _this.props.onScheduleClick &&
                                        _this.props.onScheduleClick(scheduleData_1);
                                } },
                                React__default["default"].createElement("div", { className: cx('ScheduleCalendar-text-overflow') }, item.content)));
                        });
                        return (React__default["default"].createElement("td", tslib.__assign({}, injectedProps),
                            React__default["default"].createElement("div", { className: cx('ScheduleCalendar-large-day-wrap') },
                                React__default["default"].createElement("div", { className: cx('ScheduleCalendar-large-schedule-header') },
                                    React__default["default"].createElement("span", null, currentDate.date())),
                                scheduleDiv,
                                schedule_1.length > 3 && (React__default["default"].createElement("div", { className: cx('ScheduleCalendar-large-schedule-footer') },
                                    schedule_1.length - 3,
                                    " ",
                                    __('more'))))));
                    }
                    // 正常模式
                    var ScheduleIcon = (React__default["default"].createElement("span", { className: cx('ScheduleCalendar-icon', schedule_1[0].className), onClick: function () {
                            return _this.props.onScheduleClick &&
                                _this.props.onScheduleClick(scheduleData_1);
                        } }));
                    return (React__default["default"].createElement("td", tslib.__assign({}, injectedProps),
                        React__default["default"].createElement("span", { style: todayActiveStyle, ref: todayDomRef },
                            currentDate.date(),
                            ScheduleIcon)));
                }
            }
            return (React__default["default"].createElement("td", tslib.__assign({}, injectedProps),
                React__default["default"].createElement("span", { style: todayActiveStyle, ref: todayDomRef }, currentDate.date())));
        };
        _this.renderTimes = function () {
            var _a = _this.props, timeFormat = _a.timeFormat, selectedDate = _a.selectedDate, viewDate = _a.viewDate, isEndDate = _a.isEndDate, cx = _a.classnames;
            var date = selectedDate || (isEndDate ? viewDate.endOf('day') : viewDate);
            var inputs = [];
            var timeConstraints = _this.timeConstraints;
            timeFormat.split(':').forEach(function (format, i) {
                var type = /h/i.test(format)
                    ? 'hours'
                    : /m/.test(format)
                        ? 'minutes'
                        : /s/.test(format)
                            ? 'seconds'
                            : '';
                if (type) {
                    var min_1 = timeConstraints[type].min;
                    var max_1 = timeConstraints[type].max;
                    var options_1 = _this.computedTimeOptions(type);
                    var formatMap_1 = {
                        hours: 'HH',
                        minutes: 'mm',
                        seconds: 'ss'
                    };
                    inputs.push(React__default["default"].createElement(Downshift__default["default"], { key: i + 'input', inputValue: date.format(formatMap_1[type]) }, function (_a) {
                        var getInputProps = _a.getInputProps, openMenu = _a.openMenu, closeMenu = _a.closeMenu;
                        getInputProps({
                            onFocus: function () { return openMenu(); },
                            onChange: function (e) {
                                return _this.setTime(type, Math.max(min_1, Math.min(parseInt(e.currentTarget.value.replace(/\D/g, ''), 10) || 0, max_1)));
                            }
                        });
                        return (React__default["default"].createElement("div", { className: cx('CalendarInputWrapper', 'CalendarInputWrapperMT') },
                            React__default["default"].createElement("div", { className: cx('CalendarInput-sugs', type === 'hours'
                                    ? 'CalendarInput-sugsHours'
                                    : 'CalendarInput-sugsTimes'), id: "".concat(_this.state.uniqueTag, "-").concat(i, "-input") }, options_1.map(function (option) {
                                var _a;
                                return (React__default["default"].createElement("div", { key: option.value, className: cx('CalendarInput-sugsItem', {
                                        'is-highlight': selectedDate
                                            ? option.value === date.format(formatMap_1[type])
                                            : option.value === ((_a = options_1 === null || options_1 === void 0 ? void 0 : options_1[0]) === null || _a === void 0 ? void 0 : _a.value)
                                    }), onClick: function () {
                                        _this.setTime(type, parseInt(option.value, 10));
                                        _this.scrollToTop(type, parseInt(option.value, 10), i);
                                        closeMenu();
                                    } }, option.value));
                            }))));
                    }));
                    inputs.push(React__default["default"].createElement("span", { key: i + 'divider' }));
                }
            });
            inputs.length && inputs.pop();
            return (React__default["default"].createElement("div", { className: cx('CalendarTimesWrapper') },
                _this.showTime(),
                React__default["default"].createElement("div", { className: cx('CalendarInputsWrapper') }, inputs)));
        };
        _this.renderFooter = function () {
            if (!_this.props.requiredConfirm) {
                return null;
            }
            var _a = _this.props, __ = _a.translate, cx = _a.classnames;
            return (React__default["default"].createElement("tfoot", { key: "tf" },
                React__default["default"].createElement("tr", null,
                    React__default["default"].createElement("td", { colSpan: 7 }, _this.props.requiredConfirm ? (React__default["default"].createElement("div", { key: "button", className: "rdtActions" },
                        React__default["default"].createElement("a", { className: cx('Button', 'Button--default'), onClick: _this.cancel }, __('cancel')),
                        React__default["default"].createElement("a", { className: cx('Button', 'Button--primary', 'm-l-sm'), onClick: _this.confirm }, __('confirm')))) : null))));
        };
        _this.onPickerConfirm = function (value) {
            _this.props.onConfirm && _this.props.onConfirm(value, _this.state.types);
        };
        _this.onPickerChange = function (value, index) {
            var _a = _this.props, selectedDate = _a.selectedDate, viewDate = _a.viewDate;
            // 变更年份、月份的时候，需要更新columns
            if (index === 1 || index === 0) {
                var currentDate = (selectedDate || viewDate || moment__default["default"]()).clone();
                // 只需计算year 、month
                var selectDate = amisCore.convertArrayValueToMoment(value, ['year', 'month'], currentDate);
                var dateBoundary = _this.props.getDateBoundary(selectDate);
                _this.setState({
                    columns: _this.props.getColumns(_this.state.types, dateBoundary),
                    pickerValue: value
                });
            }
        };
        _this.renderPicker = function () {
            var __ = _this.props.translate;
            var title = _this.state.types.length > 3 ? __('Date.titleTime') : __('Date.titleDate');
            return (React__default["default"].createElement(Picker["default"], { translate: _this.props.translate, locale: _this.props.locale, title: title, columns: _this.state.columns, value: _this.state.pickerValue, onChange: _this.onPickerChange, onConfirm: _this.onPickerConfirm, onClose: _this.cancel }));
        };
        var selectedDate = props.selectedDate; props.viewDate; var timeFormat = props.timeFormat;
        var currentDate = selectedDate || moment__default["default"]();
        var types = ['year', 'month', 'date'];
        timeFormat.split(':').forEach(function (format) {
            var type = /h/i.test(format)
                ? 'hours'
                : /m/.test(format)
                    ? 'minutes'
                    : /s/.test(format)
                        ? 'seconds'
                        : '';
            type && types.push(type);
        });
        var dateBoundary = _this.props.getDateBoundary(currentDate);
        var columns = _this.props.getColumns(types, dateBoundary);
        _this.state = {
            columns: columns,
            types: types,
            pickerValue: currentDate.toArray(),
            uniqueTag: new Date().valueOf()
        };
        if (_this.props.timeConstraints) {
            _this.timeConstraints = merge__default["default"](_this.timeConstraints, props.timeConstraints);
        }
        return _this;
    }
    CustomDaysView.prototype.getDaysOfWeek = function (locale) {
        var days = locale._weekdaysMin, first = locale.firstDayOfWeek(), dow = [], i = 0;
        days.forEach(function (day) {
            dow[(7 + i++ - first) % 7] = day;
        });
        return dow;
    };
    CustomDaysView.prototype.alwaysValidDate = function () {
        return 1;
    };
    CustomDaysView.prototype.renderDays = function () {
        var date = this.props.viewDate, selected = this.props.selectedDate && this.props.selectedDate.clone(), prevMonth = date.clone().subtract(1, 'months'), currentYear = date.year(), currentMonth = date.month(), weeks = [], days = [], renderer = this.props.renderDay || this.renderDay, isValid = this.props.isValidDate || this.alwaysValidDate, classes, isDisabled, dayProps, currentDate;
        var _a = this.props.todayActiveStyle, todayActiveStyle = _a === void 0 ? {} : _a;
        // Go to the last week of the previous month
        prevMonth.date(prevMonth.daysInMonth()).startOf('week');
        var lastDay = prevMonth.clone().add(42, 'd');
        while (prevMonth.isBefore(lastDay)) {
            classes = 'rdtDay';
            currentDate = prevMonth.clone();
            if ((prevMonth.year() === currentYear &&
                prevMonth.month() < currentMonth) ||
                prevMonth.year() < currentYear)
                classes += ' rdtOld';
            else if ((prevMonth.year() === currentYear &&
                prevMonth.month() > currentMonth) ||
                prevMonth.year() > currentYear)
                classes += ' rdtNew';
            if (selected && prevMonth.isSame(selected, 'day'))
                classes += ' rdtActive';
            if (prevMonth.isSame(moment__default["default"](), 'day'))
                classes += ' rdtToday';
            isDisabled = !isValid(currentDate, selected);
            if (isDisabled)
                classes += ' rdtDisabled';
            dayProps = extend__default["default"]({
                'key': prevMonth.format('M_D'),
                'data-value': prevMonth.date(),
                'className': classes
            }, classes.includes('rdtToday') ? { todayActiveStyle: todayActiveStyle } : {});
            if (!isDisabled)
                dayProps.onClick = this.updateSelectedDate;
            days.push(renderer(dayProps, currentDate, selected));
            if (days.length === 7) {
                weeks.push(React__default["default"].createElement('tr', { key: prevMonth.format('M_D') }, days));
                days = [];
            }
            prevMonth.add(1, 'd');
        }
        return weeks;
    };
    CustomDaysView.prototype.componentDidMount = function () {
        var _this = this;
        var _a = this.props, timeFormat = _a.timeFormat, selectedDate = _a.selectedDate, viewDate = _a.viewDate, isEndDate = _a.isEndDate;
        var formatMap = {
            hours: 'HH',
            minutes: 'mm',
            seconds: 'ss'
        };
        var date = selectedDate || (isEndDate ? viewDate.endOf('day') : viewDate);
        timeFormat.split(':').forEach(function (format, i) {
            var type = /h/i.test(format)
                ? 'hours'
                : /m/.test(format)
                    ? 'minutes'
                    : /s/.test(format)
                        ? 'seconds'
                        : '';
            if (type) {
                _this.scrollToTop(type, parseInt(date.format(formatMap[type]), 10), i, 'init');
            }
        });
    };
    /** 时间选择器数据源 */
    CustomDaysView.prototype.computedTimeOptions = function (timeScale) {
        var _a;
        var _b = (_a = this.timeConstraints) === null || _a === void 0 ? void 0 : _a[timeScale], min = _b.min, max = _b.max, step = _b.step;
        return Array.from({ length: max - min + 1 }, function (item, index) {
            var value = (index + min)
                .toString()
                .padStart(timeScale !== 'milliseconds' ? 2 : 3, '0');
            return index % step === 0 ? { label: value, value: value } : undefined;
        }).filter(function (item) { return !!item; });
    };
    CustomDaysView.prototype.render = function () {
        var _a = this.props, date = _a.viewDate, useMobileUI = _a.useMobileUI, embed = _a.embed, timeFormat = _a.timeFormat; _a.classnames;
        var locale = date.localeData();
        var __ = this.props.translate;
        if (amisCore.isMobile() && useMobileUI && !embed) {
            return React__default["default"].createElement("div", { className: "rdtYears" }, this.renderPicker());
        }
        var tableChildren = [
            this.props.hideHeader ? null : (React__default["default"].createElement("thead", { key: "th" },
                React__default["default"].createElement("tr", null,
                    React__default["default"].createElement("th", { colSpan: 7 },
                        React__default["default"].createElement("div", { className: "rdtHeader" },
                            React__default["default"].createElement("a", { className: "rdtPrev", onClick: this.props.subtractTime(1, 'years') },
                                React__default["default"].createElement(icons.Icon, { icon: "right-double-arrow", className: "icon date-icon-arrow-left" })),
                            React__default["default"].createElement("a", { className: "rdtPrev", onClick: this.props.subtractTime(1, 'months') },
                                React__default["default"].createElement(icons.Icon, { icon: "right-arrow", className: "icon date-icon-arrow-left" })),
                            React__default["default"].createElement("div", { className: "rdtCenter" },
                                React__default["default"].createElement("a", { className: "rdtSwitch", onClick: this.props.showView('years') }, date.format(__('dateformat.year'))),
                                React__default["default"].createElement("a", { className: "rdtSwitch", onClick: this.props.showView('months') }, date.format(__('MMM')))),
                            React__default["default"].createElement("a", { className: "rdtNext", onClick: this.props.addTime(1, 'months') },
                                React__default["default"].createElement(icons.Icon, { icon: "right-arrow", className: "icon date-icon-arrow" })),
                            React__default["default"].createElement("a", { className: "rdtNext", onClick: this.props.addTime(1, 'years') },
                                React__default["default"].createElement(icons.Icon, { icon: "right-double-arrow", className: "icon date-icon-arrow" }))),
                        React__default["default"].createElement("div", { className: "header-line" }))),
                React__default["default"].createElement("tr", null, this.getDaysOfWeek(locale).map(function (day, index) { return (React__default["default"].createElement("th", { key: day + index, className: "dow" }, day)); })))),
            React__default["default"].createElement("tbody", { key: "tb" }, this.renderDays())
        ];
        return (React__default["default"].createElement(React__default["default"].Fragment, null,
            React__default["default"].createElement("div", { className: timeFormat ? 'rdtDays' : '' },
                React__default["default"].createElement("table", { className: timeFormat ? 'rdtDaysPart' : '' }, tableChildren),
                timeFormat ? (React__default["default"].createElement("div", { className: timeFormat.toLowerCase().indexOf('s') > 0
                        ? 'rdtTimePartWithS'
                        : 'rdtTimePart' }, this.renderTimes())) : null),
            React__default["default"].createElement("table", null, this.renderFooter())));
    };
    return CustomDaysView;
}(React__default["default"].Component));
var CustomDaysView$1 = amisCore.localeable(CustomDaysView);

exports.CustomDaysView = CustomDaysView;
exports["default"] = CustomDaysView$1;
