// amis-ui/lib/components/table/Cell.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var React = require('react');
var amisCore = require('amis-core');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

/**
 * @file table/BodyCell
 * @author fex
 */
var zIndex = 1;
var BodyCell = /** @class */ (function (_super) {
    tslib.__extends(BodyCell, _super);
    function BodyCell() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BodyCell.prototype.render = function () {
        var _a;
        var _b = this.props, fixed = _b.fixed, rowSpan = _b.rowSpan, colSpan = _b.colSpan, key = _b.key, children = _b.children, className = _b.className, column = _b.column, style = _b.style, groupId = _b.groupId, depth = _b.depth, Component = _b.wrapperComponent, cx = _b.classnames;
        return (React__default["default"].createElement(Component, { key: key || null, rowSpan: rowSpan && rowSpan > 1 ? rowSpan : null, colSpan: colSpan && colSpan > 1 ? colSpan : null, className: cx('Table-cell', className, (_a = {},
                _a[cx("Table-cell-fix-".concat(fixed))] = fixed,
                _a["text-".concat(column === null || column === void 0 ? void 0 : column.align)] = column === null || column === void 0 ? void 0 : column.align,
                _a)), style: fixed ? tslib.__assign({ position: 'sticky', zIndex: zIndex }, style) : tslib.__assign({}, style), "data-group-id": groupId || null, "data-depth": depth || null }, children));
    };
    BodyCell.defaultProps = {
        fixed: '',
        wrapperComponent: 'td',
        rowSpan: null,
        colSpan: null
    };
    return BodyCell;
}(React__default["default"].Component));
var Cell = amisCore.themeable(amisCore.localeable(BodyCell));

exports.BodyCell = BodyCell;
exports["default"] = Cell;
