// amis-ui/lib/components/virtual-list/constants.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _a, _b, _c, _d, _e;
exports.ALIGNMENT = void 0;
(function (ALIGNMENT) {
    ALIGNMENT["AUTO"] = "auto";
    ALIGNMENT["START"] = "start";
    ALIGNMENT["CENTER"] = "center";
    ALIGNMENT["END"] = "end";
})(exports.ALIGNMENT || (exports.ALIGNMENT = {}));
exports.DIRECTION = void 0;
(function (DIRECTION) {
    DIRECTION["HORIZONTAL"] = "horizontal";
    DIRECTION["VERTICAL"] = "vertical";
})(exports.DIRECTION || (exports.DIRECTION = {}));
exports.SCROLL_CHANGE_REASON = void 0;
(function (SCROLL_CHANGE_REASON) {
    SCROLL_CHANGE_REASON["OBSERVED"] = "observed";
    SCROLL_CHANGE_REASON["REQUESTED"] = "requested";
})(exports.SCROLL_CHANGE_REASON || (exports.SCROLL_CHANGE_REASON = {}));
var scrollProp = (_a = {},
    _a[exports.DIRECTION.VERTICAL] = 'scrollTop',
    _a[exports.DIRECTION.HORIZONTAL] = 'scrollLeft',
    _a);
var sizeProp = (_b = {},
    _b[exports.DIRECTION.VERTICAL] = 'height',
    _b[exports.DIRECTION.HORIZONTAL] = 'width',
    _b);
var positionProp = (_c = {},
    _c[exports.DIRECTION.VERTICAL] = 'top',
    _c[exports.DIRECTION.HORIZONTAL] = 'left',
    _c);
var marginProp = (_d = {},
    _d[exports.DIRECTION.VERTICAL] = 'marginTop',
    _d[exports.DIRECTION.HORIZONTAL] = 'marginLeft',
    _d);
var oppositeMarginProp = (_e = {},
    _e[exports.DIRECTION.VERTICAL] = 'marginBottom',
    _e[exports.DIRECTION.HORIZONTAL] = 'marginRight',
    _e);

exports.marginProp = marginProp;
exports.oppositeMarginProp = oppositeMarginProp;
exports.positionProp = positionProp;
exports.scrollProp = scrollProp;
exports.sizeProp = sizeProp;
