// office-viewer/esm/openxml/word/Section.js
import { __values } from '../../node_modules/tslib/tslib.es6.js';
import { parseSize } from '../../parse/parseSize.js';

/**
 * section 属性
 * word 文档是按 section 组织的
 * 参考了 docx 里的类型定义
 */
var Section = /** @class */ (function () {
    function Section() {
        this.properties = {};
        this.children = [];
    }
    Section.prototype.addChild = function (child) {
        this.children.push(child);
    };
    Section.parsePr = function (element) {
        var e_1, _a;
        var properties = {};
        try {
            for (var _b = __values(element.children), _c = _b.next(); !_c.done; _c = _b.next()) {
                var child = _c.value;
                var tagName = child.tagName;
                switch (tagName) {
                    case 'w:pgSz':
                        properties.pageSize = {
                            width: parseSize(child, 'w:w'),
                            height: parseSize(child, 'w:h')
                        };
                        break;
                    case 'w:pgMar':
                        properties.pageMargin = {
                            left: parseSize(child, 'w:left'),
                            right: parseSize(child, 'w:right'),
                            top: parseSize(child, 'w:top'),
                            bottom: parseSize(child, 'w:bottom'),
                            header: parseSize(child, 'w:header'),
                            footer: parseSize(child, 'w:footer'),
                            gutter: parseSize(child, 'w:gutter')
                        };
                        break;
                    default:
                        break;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return properties;
    };
    return Section;
}());

export { Section };
