// office-viewer/esm/render/renderNumbering.js
import { __assign } from '../node_modules/tslib/tslib.es6.js';
import { createElement } from '../util/dom.js';
import { ST_NumberFormat } from '../openxml/Types.js';
import { setElementStyle } from './setElementStyle.js';

/**
 * 整数转成罗马数字，来自 https://stackoverflow.com/a/32851198
 */
function romanize(num) {
    var lookup = {
        M: 1000,
        CM: 900,
        D: 500,
        CD: 400,
        C: 100,
        XC: 90,
        L: 50,
        XL: 40,
        X: 10,
        IX: 9,
        V: 5,
        IV: 4,
        I: 1
    };
    var roman = '';
    for (var key in lookup) {
        while (num >= lookup[key]) {
            roman += key;
            num -= lookup[key];
        }
    }
    return roman;
}
/**
 * 将数字转成列表格式，目前只支持少数几种格式，不支持的都直接返回数字
 */
function convertNumToFormat(numFmt, num) {
    switch (numFmt) {
        case ST_NumberFormat.decimal:
            return num.toString();
        case ST_NumberFormat.lowerLetter:
            return String.fromCharCode(96 + num);
        case ST_NumberFormat.upperLetter:
            return String.fromCharCode(64 + num);
        case ST_NumberFormat.lowerRoman:
            return romanize(num).toLowerCase();
        case ST_NumberFormat.upperRoman:
            return romanize(num).toUpperCase();
        case ST_NumberFormat.bullet:
            // 原本其实是用
            return '&bull;';
        default:
            return num.toString();
    }
}
/**
 * 渲染列表 http://officeopenxml.com/WPnumbering.php
 * 为了支持复杂场景，这里使用代码来直接生成内容
 */
function renderNumbering(p, word, numPr) {
    var numbering = word.numbering;
    var numId = numPr.numId;
    if (!numId) {
        console.warn('renderNumbering: numId is empty');
        return null;
    }
    if (!numbering) {
        console.warn('renderNumbering: numbering is empty');
        return null;
    }
    var num = numbering.nums[numId];
    if (!num) {
        console.warn('renderNumbering: num is empty');
        return null;
    }
    var abstractNum = numbering.abstractNums[num.abstractNumId];
    var lvls = abstractNum.lvls;
    if (num.lvlOverride) {
        lvls = __assign(__assign({}, lvls), num.lvlOverride.lvls);
    }
    var lvl = lvls[numPr.ilvl];
    if (!lvl) {
        console.warn('renderNumbering: lvl is empty');
        return null;
    }
    var ilvl = numPr.ilvl;
    var ilvlData = numbering.numData[numId];
    // 还不支持 http://officeopenxml.com/WPnumbering-restart.php
    if (!ilvlData[ilvl]) {
        ilvlData[ilvl] = lvl.start;
    }
    else {
        ilvlData[ilvl] += 1;
        // 加一之后，将比它大的都清空，这样才能每个级别重置
        for (var ilvIndex in ilvlData) {
            if (parseInt(ilvIndex) > parseInt(ilvl)) {
                ilvlData[ilvIndex] = 0;
            }
        }
    }
    var element = createElement('span');
    var lvlText = lvl.lvlText;
    // 格式化内容
    var level = parseInt(ilvl);
    // 获取每个层级的值
    var levelNums = [];
    for (var i = 0; i <= level; i++) {
        var listNumber = ilvlData[i];
        if (listNumber) {
            var numFmt = lvls[i].numFmt;
            var numText = convertNumToFormat(numFmt, listNumber);
            // 强制数字显示 http://webapp.docx4java.org/OnlineDemo/ecma376/WordML/isLgl.html
            // http://officeopenxml.com/WPnumbering-isLgl.php
            if (lvl.isLgl) {
                numText = String(listNumber);
            }
            levelNums.push(numText);
        }
    }
    for (var i = 0; i < levelNums.length; i++) {
        var levelNum = levelNums[i];
        lvlText = lvlText.replace("%".concat(i + 1), levelNum);
    }
    // 这个 pPr 似乎是影响父级的
    // http://webapp.docx4java.org/OnlineDemo/ecma376/WordML/pPr_6.html
    // This element specifies the paragraph properties which shall be applied as part of a given numbering level within the parent numbering definition.
    setElementStyle(word, p, lvl.pPr);
    setElementStyle(word, element, lvl.rPr);
    // 还不支持 lvlJc
    // 还不支持 image
    // http://officeopenxml.com/WPnumbering-imagesAsSymbol.php
    if (lvl.numFmt !== ST_NumberFormat.bullet ||
        word.renderOptions.bulletUseFont) {
        element.innerText = lvlText;
    }
    else {
        // 如果没有字体只能尽可能模拟了
        // 参考 https://www.w3schools.com/charsets/ref_utf_geometric.asp
        var bulletText = '&bull;';
        var unicodeString = lvlText.charCodeAt(0).toString(16).padStart(4, '0');
        if (unicodeString === 'f06e') {
            bulletText = '&#9632;';
        }
        else if (unicodeString === 'f075') {
            bulletText = '&#9670;';
        }
        else if (unicodeString === 'f0d8') {
            bulletText = '&#9658;';
        }
        element.innerHTML = bulletText;
    }
    if (lvl.suff === 'space') {
        element.innerHTML += ' ';
    }
    else if (lvl.suff === 'tab') {
        element.innerHTML += '&emsp;';
    }
    return element;
}

export { renderNumbering };
