// amis-ui/lib/components/condition-builder/Value.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var React = require('react');
var amisCore = require('amis-core');
var InputBox = require('../InputBox.js');
var NumberInput = require('../NumberInput.js');
var DatePicker = require('../DatePicker.js');
var Select = require('../Select.js');
var Switch = require('../Switch.js');
var Picker = require('../formula/Picker.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var Value = /** @class */ (function (_super) {
    tslib.__extends(Value, _super);
    function Value() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Value.prototype.render = function () {
        var _a;
        var _b = this.props, cx = _b.classnames, field = _b.field, value = _b.value, onChange = _b.onChange, op = _b.op, __ = _b.translate, data = _b.data, disabled = _b.disabled, formula = _b.formula, popOverContainer = _b.popOverContainer, renderEtrValue = _b.renderEtrValue;
        var input = undefined;
        if (formula) {
            // 如果配置了 formula 字段，则所有的输入变为 formula 形式
            formula = Object.assign(formula, {
                translate: __,
                classnames: cx,
                data: data,
                value: value !== null && value !== void 0 ? value : field.defaultValue,
                onChange: onChange,
                disabled: disabled
            });
            input = React__default["default"].createElement(Picker.FormulaPicker, tslib.__assign({}, formula));
        }
        else if (field.type === 'text') {
            input = (React__default["default"].createElement(InputBox["default"], { value: value !== null && value !== void 0 ? value : field.defaultValue, onChange: onChange, placeholder: __(field.placeholder), disabled: disabled }));
        }
        else if (field.type === 'number') {
            input = (React__default["default"].createElement(NumberInput["default"], { placeholder: __(field.placeholder) || __('NumberInput.placeholder'), step: field.step, min: field.minimum, max: field.maximum, precision: field.precision, value: value !== null && value !== void 0 ? value : field.defaultValue, onChange: onChange, disabled: disabled }));
        }
        else if (field.type === 'date') {
            input = (React__default["default"].createElement(DatePicker["default"], { placeholder: __(field.placeholder) || __('Date.placeholder'), format: field.format || 'YYYY-MM-DD', inputFormat: field.inputFormat || 'YYYY-MM-DD', value: value !== null && value !== void 0 ? value : field.defaultValue, onChange: onChange, timeFormat: "", disabled: disabled, popOverContainer: popOverContainer }));
        }
        else if (field.type === 'time') {
            input = (React__default["default"].createElement(DatePicker["default"], { viewMode: "time", placeholder: __(field.placeholder) || __('Time.placeholder'), format: field.format || 'HH:mm', inputFormat: field.inputFormat || 'HH:mm', value: value !== null && value !== void 0 ? value : field.defaultValue, onChange: onChange, dateFormat: "", timeFormat: field.format || 'HH:mm', disabled: disabled, popOverContainer: popOverContainer }));
        }
        else if (field.type === 'datetime') {
            input = (React__default["default"].createElement(DatePicker["default"], { placeholder: __(field.placeholder) || 'Time.placeholder', format: field.format || '', inputFormat: field.inputFormat || 'YYYY-MM-DD HH:mm', value: value !== null && value !== void 0 ? value : field.defaultValue, onChange: onChange, timeFormat: field.timeFormat || 'HH:mm', disabled: disabled, popOverContainer: popOverContainer }));
        }
        else if (field.type === 'select') {
            var autoComplete = field.autoComplete;
            input = (React__default["default"].createElement(Select.SelectWithRemoteOptions, { simpleValue: true, options: field.options, placeholder: __(field.placeholder) || 'Select.placeholder', source: field.source, autoComplete: autoComplete, searchable: field.searchable, value: (_a = value !== null && value !== void 0 ? value : field.defaultValue) !== null && _a !== void 0 ? _a : '', data: data, onChange: onChange, multiple: op === 'select_any_in' || op === 'select_not_any_in', disabled: disabled, popOverContainer: popOverContainer }));
        }
        else if (field.type === 'boolean') {
            input = (React__default["default"].createElement(Switch["default"], { value: value !== null && value !== void 0 ? value : field.defaultValue, onChange: onChange, disabled: disabled }));
        }
        else if (field.type === 'custom') {
            input = renderEtrValue
                ? renderEtrValue(tslib.__assign(tslib.__assign({}, field.value), { name: field.name }), {
                    data: data,
                    onChange: onChange,
                    value: value !== null && value !== void 0 ? value : field.defaultValue
                })
                : null;
        }
        else {
            var res = value !== null && value !== void 0 ? value : field.defaultValue;
            input = renderEtrValue
                ? renderEtrValue(field, {
                    data: data,
                    onChange: onChange,
                    value: res ? res[field.name] : res
                })
                : null;
        }
        return React__default["default"].createElement("div", { className: cx('CBValue') }, input);
    };
    return Value;
}(React__default["default"].Component));
var Value$1 = amisCore.themeable(amisCore.localeable(Value));

exports.Value = Value;
exports["default"] = Value$1;
