// amis-ui/lib/components/json-schema/Object.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var React = require('react');
var amisCore = require('amis-core');
var Button = require('../Button.js');
var icons = require('../icons.js');
var InputBox = require('../InputBox.js');
var InputBoxWithSuggestion = require('../InputBoxWithSuggestion.js');
var Select = require('../Select.js');
var Item = require('./Item.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function InputJSONSchemaObject(props) {
    var _a, _b;
    var cx = props.classnames, value = props.value, onChange = props.onChange, disabled = props.disabled, __ = props.translate, renderKey = props.renderKey, collapsable = props.collapsable, renderValue = props.renderValue;
    var buildMembers = React__default["default"].useCallback(function (schema, value) {
        var e_1, _a;
        var members = [];
        var required = Array.isArray(schema.required) ? schema.required : [];
        Object.keys(schema.properties || {}).forEach(function (key) {
            var child = schema.properties[key];
            members.push({
                key: amisCore.guid(),
                name: key,
                nameMutable: !required.includes(key),
                required: required.includes(key),
                schema: child
            });
        });
        var keys = Object.keys(value || {});
        var _loop_1 = function (key) {
            var exists = members.find(function (m) { return m.name === key; });
            if (!exists) {
                members.push({
                    key: amisCore.guid(),
                    name: key,
                    nameMutable: true,
                    schema: {
                        type: 'string'
                    }
                });
            }
        };
        try {
            for (var keys_1 = tslib.__values(keys), keys_1_1 = keys_1.next(); !keys_1_1.done; keys_1_1 = keys_1.next()) {
                var key = keys_1_1.value;
                _loop_1(key);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (keys_1_1 && !keys_1_1.done && (_a = keys_1.return)) _a.call(keys_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (!members.length) {
            members.push({
                key: amisCore.guid(),
                name: '',
                nameMutable: true,
                schema: {
                    type: 'string'
                }
            });
        }
        return members;
    }, []);
    var _c = tslib.__read(React__default["default"].useState(buildMembers(props.schema, props.value)), 2), members = _c[0], setMembers = _c[1];
    var membersRef = React__default["default"].useRef();
    membersRef.current = members;
    var _d = tslib.__read(React__default["default"].useState(collapsable ? true : false), 2), collapsed = _d[0], setCollapsed = _d[1];
    var toggleCollapsed = function () {
        setCollapsed(!collapsed);
    };
    var onMemberChange = function (member, memberValue) {
        var _a;
        var newValue = tslib.__assign(tslib.__assign({}, props.value), (_a = {}, _a[member.name] = memberValue, _a));
        onChange === null || onChange === void 0 ? void 0 : onChange(newValue);
    };
    var onMemberKeyChange = function (member, memberValue) {
        var _a, _b;
        var idx = members.indexOf(member);
        if (!~idx) {
            throw new Error('member object not found');
        }
        var newValue = tslib.__assign({}, props.value);
        var m = members.concat();
        m.splice(idx, 1, tslib.__assign(tslib.__assign({}, member), { schema: ((_b = (_a = props.schema) === null || _a === void 0 ? void 0 : _a.properties) === null || _b === void 0 ? void 0 : _b[memberValue]) || {
                type: 'string'
            }, name: memberValue, invalid: !memberValue ||
                members.some(function (a, b) { return a.name === memberValue && b !== idx; }) }));
        setMembers(m);
        newValue[memberValue] = newValue[member.name];
        delete newValue[member.name];
        onChange === null || onChange === void 0 ? void 0 : onChange(newValue);
    };
    var onMemberDelete = function (member) {
        var idx = members.indexOf(member);
        if (!~idx) {
            throw new Error('member object not found');
        }
        var m = members.concat();
        m.splice(idx, 1);
        setMembers(m);
        var newValue = tslib.__assign({}, props.value);
        delete newValue[member.name];
        onChange === null || onChange === void 0 ? void 0 : onChange(newValue);
    };
    React__default["default"].useEffect(function () {
        setMembers(buildMembers(props.schema, props.value));
    }, [JSON.stringify(props.schema)]);
    React__default["default"].useEffect(function () {
        var e_2, _a;
        var value = props.value;
        var m = membersRef.current.concat();
        var keys = Object.keys(value || {});
        var _loop_2 = function (key) {
            var exists = m.find(function (m) { return m.name === key; });
            if (!exists) {
                m.push({
                    key: amisCore.guid(),
                    name: key,
                    nameMutable: true,
                    schema: {
                        type: 'string'
                    }
                });
            }
        };
        try {
            for (var keys_2 = tslib.__values(keys), keys_2_1 = keys_2.next(); !keys_2_1.done; keys_2_1 = keys_2.next()) {
                var key = keys_2_1.value;
                _loop_2(key);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (keys_2_1 && !keys_2_1.done && (_a = keys_2.return)) _a.call(keys_2);
            }
            finally { if (e_2) throw e_2.error; }
        }
        if (m.length !== membersRef.current.length) {
            setMembers(m);
        }
    }, [JSON.stringify(props.value)]);
    var handleAdd = React__default["default"].useCallback(function () {
        var m = members.concat();
        m.push({
            key: amisCore.guid(),
            name: '',
            invalid: true,
            nameMutable: true
        });
        setMembers(m);
    }, [members]);
    var options = [];
    var properties = ((_a = props.schema) === null || _a === void 0 ? void 0 : _a.properties) || {};
    Object.keys(properties).forEach(function (key) {
        var _a;
        options.push({
            label: ((_a = properties[key]) === null || _a === void 0 ? void 0 : _a.title) || key,
            value: key
        });
    });
    // todo additionalProperties 还有其他格式
    var allowAdd = !(props.schema.additionalProperties === false &&
        options.every(function (o) { return members.find(function (m) { return m.name === o.value; }); }));
    var allowInput = props.schema.additionalProperties !== false;
    return (React__default["default"].createElement(React__default["default"].Fragment, null,
        collapsable ? (React__default["default"].createElement("a", { className: cx('JSONSchemaObject-caret', {
                'is-collapsed': collapsed
            }), onClick: toggleCollapsed },
            React__default["default"].createElement(icons.Icon, { icon: "caret", className: "icon" }))) : null,
        React__default["default"].createElement("div", { className: cx('JSONSchemaObject', {
                'is-expanded': collapsable && !collapsed
            }) },
            collapsed ? (renderValue ? (React__default["default"].createElement(Item.InputJSONSchemaItem, tslib.__assign({}, props, { value: value, onChange: onChange, schema: {
                    type: 'string'
                }, placeholder: (_b = props.schema) === null || _b === void 0 ? void 0 : _b.description }))) : null) : (members.map(function (member) {
                var _a, _b;
                var filtedOptions = options.filter(function (o) { return !members.find(function (m) { return m !== member && m.name === o.value; }); });
                return (React__default["default"].createElement("div", { key: member.key, className: cx('JSONSchemaMember') },
                    React__default["default"].createElement("div", { className: cx('JSONSchemaMember-key') }, member.nameMutable ? (React__default["default"].createElement(React__default["default"].Fragment, null, renderKey ? (renderKey(member.name, onMemberKeyChange.bind(null, member), member.schema, props)) : filtedOptions.length ? (allowInput ? (React__default["default"].createElement(InputBoxWithSuggestion["default"], { value: member.name, hasError: member.invalid, onChange: onMemberKeyChange.bind(null, member), clearable: false, placeholder: __('JSONSchema.key'), options: filtedOptions })) : (React__default["default"].createElement(Select["default"], { simpleValue: true, block: true, value: member.name, hasError: member.invalid, onChange: onMemberKeyChange.bind(null, member), clearable: false, placeholder: __('JSONSchema.key'), options: filtedOptions }))) : (React__default["default"].createElement(InputBox["default"], { value: member.name, hasError: member.invalid, onChange: onMemberKeyChange.bind(null, member), clearable: false, placeholder: __('JSONSchema.key') })))) : (React__default["default"].createElement("span", null,
                        member.required ? (React__default["default"].createElement("span", { className: cx("Form-star") }, "*")) : null,
                        ((_a = member.schema) === null || _a === void 0 ? void 0 : _a.title) || member.name))),
                    React__default["default"].createElement("div", { className: cx('JSONSchemaMember-value') },
                        React__default["default"].createElement(Item.InputJSONSchemaItem, tslib.__assign({}, props, { value: value === null || value === void 0 ? void 0 : value[member.name], onChange: onMemberChange.bind(null, member), schema: member.schema || {
                                type: 'string'
                            }, placeholder: (_b = member.schema) === null || _b === void 0 ? void 0 : _b.description, collapsable: true }))),
                    !member.required ? (React__default["default"].createElement(Button["default"], { className: cx('SchemaEditor-btn'), onClick: onMemberDelete.bind(null, member), iconOnly: true, disabled: disabled || !!(value === null || value === void 0 ? void 0 : value.$ref) },
                        React__default["default"].createElement(icons.Icon, { icon: "remove", className: "icon" }))) : null));
            })),
            allowAdd && !collapsed ? (React__default["default"].createElement(Button["default"], { level: "link", onClick: handleAdd, size: "xs", disabled: disabled }, __('JSONSchema.add_prop'))) : null)));
}

exports.InputJSONSchemaObject = InputJSONSchemaObject;
