// amis-core/lib/index.js
/**
 * amis-core v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var factory = require('./factory.js');
require('./renderers/builtin.js');
require('./renderers/register.js');
var api = require('./utils/api.js');
var attachmentAdpator = require('./utils/attachmentAdpator.js');
var autobind = require('./utils/autobind.js');
var ColorScale = require('./utils/ColorScale.js');
var columnsSplit = require('./utils/columnsSplit.js');
var dataMapping = require('./utils/dataMapping.js');
var DataSchema = require('./utils/DataSchema.js');
var DataScope = require('./utils/DataScope.js');
var date = require('./utils/date.js');
var debug = require('./utils/debug.js');
var dom = require('./utils/dom.js');
var errors = require('./utils/errors.js');
var escapeHtml = require('./utils/escapeHtml.js');
var filterSchema = require('./utils/filter-schema.js');
require('./utils/filter.js');
var formatDuration = require('./utils/formatDuration.js');
var formula = require('./utils/formula.js');
var getVariable = require('./utils/getVariable.js');
var grammar = require('./utils/grammar.js');
var handleAction = require('./utils/handleAction.js');
var helper = require('./utils/helper.js');
var highlight = require('./utils/highlight.js');
var icon = require('./utils/icon.js');
var image = require('./utils/image.js');
var isPureVariable = require('./utils/isPureVariable.js');
var jsonSchema2AmisSchema = require('./utils/json-schema-2-amis-schema.js');
var keyToPath = require('./utils/keyToPath.js');
var makeSorter = require('./utils/makeSorter.js');
var normalizeLink = require('./utils/normalizeLink.js');
var normalizeOptions = require('./utils/normalizeOptions.js');
var object = require('./utils/object.js');
var offset = require('./utils/offset.js');
var offsetParent = require('./utils/offsetParent.js');
var optionValueCompare = require('./utils/optionValueCompare.js');
var position = require('./utils/position.js');
var prettyBytes = require('./utils/prettyBytes.js');
var rendererEvent = require('./utils/renderer-event.js');
var replaceText = require('./utils/replaceText.js');
var resizeSensor = require('./utils/resize-sensor.js');
var resolveVariable = require('./utils/resolveVariable.js');
var resolveVariableAndFilter = require('./utils/resolveVariableAndFilter.js');
var resolveVariableAndFilterForAsync = require('./utils/resolveVariableAndFilterForAsync.js');
var RootClose = require('./utils/RootClose.js');
var scrollPosition = require('./utils/scrollPosition.js');
var SimpleMap = require('./utils/SimpleMap.js');
var string2regExp = require('./utils/string2regExp.js');
var style = require('./utils/style.js');
var tokenize = require('./utils/tokenize.js');
var tplBuiltin = require('./utils/tpl-builtin.js');
var tpl = require('./utils/tpl.js');
var uncontrollable = require('./utils/uncontrollable.js');
var validations = require('./utils/validations.js');
var toNumber = require('./utils/toNumber.js');
var decodeEntity = require('./utils/decodeEntity.js');
var styleHelper = require('./utils/style-helper.js');
var resolveCondition = require('./utils/resolveCondition.js');
var Animation = require('./utils/Animation.js');
var index = require('./store/index.js');
var mobxStateTree = require('mobx-state-tree');
var locale = require('./locale.js');
var Scoped = require('./Scoped.js');
var theme = require('./theme.js');
require('./actions/LoopAction.js');
require('./actions/BreakAction.js');
require('./actions/ContinueAction.js');
require('./actions/SwitchAction.js');
require('./actions/ParallelAction.js');
require('./actions/CustomAction.js');
require('./actions/BroadcastAction.js');
require('./actions/CmptAction.js');
require('./actions/AjaxAction.js');
require('./actions/CopyAction.js');
require('./actions/DialogAction.js');
require('./actions/DrawerAction.js');
require('./actions/EmailAction.js');
require('./actions/LinkAction.js');
require('./actions/ToastAction.js');
require('./actions/PageAction.js');
var Action = require('./actions/Action.js');
var Item = require('./renderers/Item.js');
var Options = require('./renderers/Options.js');
var Root = require('./Root.js');
var envOverwrite = require('./envOverwrite.js');
var env = require('./env.js');
var React = require('react');
var amisFormula = require('amis-formula');
var LazyComponent = require('./components/LazyComponent.js');
var Overlay = require('./components/Overlay.js');
var PopOver = require('./components/PopOver.js');
var Form = require('./renderers/Form.js');
var stripNumber = require('./utils/stripNumber.js');
var iRenderer = require('./store/iRenderer.js');
var service = require('./store/service.js');
var form = require('./store/form.js');
var combo = require('./store/combo.js');
var crud = require('./store/crud.js');
var table = require('./store/table.js');
var table2 = require('./store/table2.js');
var list = require('./store/list.js');
var modal = require('./store/modal.js');
var formItem = require('./store/formItem.js');
var pagination = require('./store/pagination.js');
var app = require('./store/app.js');
var node = require('./store/node.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

/** @license amis v@version
 *
 * Copyright Baidu
 *
 * This source code is licensed under the Apache license found in the
 * LICENSE file in the root directory of this source tree.
 */
var classPrefix = theme.getClassPrefix();
// @ts-ignore
var version = '2.9.0';
function render(schema, props, options, pathPrefix) {
    if (props === void 0) { props = {}; }
    if (options === void 0) { options = {}; }
    if (pathPrefix === void 0) { pathPrefix = ''; }
    var locale$1 = props.locale || locale.getDefaultLocale();
    // 兼容 locale 的不同写法
    locale$1 = locale$1.replace('_', '-');
    locale$1 = locale$1 === 'en' ? 'en-US' : locale$1;
    locale$1 = locale$1 === 'zh' ? 'zh-CN' : locale$1;
    locale$1 = locale$1 === 'cn' ? 'zh-CN' : locale$1;
    var translate = props.translate || locale.makeTranslator(locale$1);
    var store = factory.stores[options.session || 'global'];
    // 根据环境覆盖 schema，这个要在最前面做，不然就无法覆盖 validations
    envOverwrite.envOverwrite(schema, locale$1);
    if (!store) {
        options = tslib.__assign(tslib.__assign(tslib.__assign({}, factory.defaultOptions), options), { fetcher: options.fetcher
                ? api.wrapFetcher(options.fetcher, options.tracker)
                : factory.defaultOptions.fetcher, confirm: helper.promisify(options.confirm || factory.defaultOptions.confirm || window.confirm), locale: locale$1, translate: translate });
        if (options.enableAMISDebug) {
            // 因为里面还有 render
            setTimeout(function () {
                debug.enableDebug();
            }, 10);
        }
        store = index.RendererStore.create({}, options);
        factory.stores[options.session || 'global'] = store;
    }
    window.amisStore = store; // 为了方便 debug.
    var env$1 = mobxStateTree.getEnv(store);
    var theme$1 = props.theme || options.theme || 'cxd';
    if (theme$1 === 'default') {
        theme$1 = 'cxd';
    }
    env$1.theme = theme.getTheme(theme$1);
    if (props.locale !== undefined) {
        env$1.translate = translate;
        env$1.locale = locale$1;
    }
    // 默认将开启移动端原生 UI
    if (options.useMobileUI !== false) {
        props.useMobileUI = true;
    }
    schema = replaceText.replaceText(schema, options.replaceText, env$1.replaceTextIgnoreKeys);
    return (React__default["default"].createElement(env.EnvContext.Provider, { value: env$1 },
        React__default["default"].createElement(Root["default"], tslib.__assign({}, props, { schema: schema, pathPrefix: pathPrefix, rootStore: store, env: env$1, theme: theme$1, locale: locale$1, translate: translate }))));
}

exports.Renderer = factory.Renderer;
exports.addSchemaFilter = factory.addSchemaFilter;
exports.clearStoresCache = factory.clearStoresCache;
exports.extendDefaultEnv = factory.extendDefaultEnv;
exports.filterSchema = factory.filterSchema;
exports.getRendererByName = factory.getRendererByName;
exports.getRenderers = factory.getRenderers;
exports.registerRenderer = factory.registerRenderer;
exports.resolveRenderer = factory.resolveRenderer;
exports.unRegisterRenderer = factory.unRegisterRenderer;
exports.updateEnv = factory.updateEnv;
exports.buildApi = api.buildApi;
exports.clearApiCache = api.clearApiCache;
exports.getApiCache = api.getApiCache;
exports.isApiOutdated = api.isApiOutdated;
exports.isApiOutdatedWithData = api.isApiOutdatedWithData;
exports.isEffectiveApi = api.isEffectiveApi;
exports.isSameApi = api.isSameApi;
exports.isValidApi = api.isValidApi;
exports.jsFetcher = api.jsFetcher;
exports.jsonpFetcher = api.jsonpFetcher;
exports.normalizeApi = api.normalizeApi;
exports.normalizeApiResponseData = api.normalizeApiResponseData;
exports.responseAdaptor = api.responseAdaptor;
exports.setApiCache = api.setApiCache;
exports.str2AsyncFunction = api.str2AsyncFunction;
exports.str2function = api.str2function;
exports.wrapAdaptor = api.wrapAdaptor;
exports.wrapFetcher = api.wrapFetcher;
exports.attachmentAdpator = attachmentAdpator.attachmentAdpator;
exports.autobindMethod = autobind.autobindMethod;
exports.bind = autobind.bind;
exports.ColorScale = ColorScale.ColorScale;
exports.columnsSplit = columnsSplit.columnsSplit;
exports.dataMapping = dataMapping.dataMapping;
exports.resolveMapping = dataMapping.resolveMapping;
exports.resolveMappingObject = dataMapping.resolveMappingObject;
exports.DataSchema = DataSchema.DataSchema;
exports.DataScope = DataScope.DataScope;
exports.filterDate = date.filterDate;
exports.parseDuration = date.parseDuration;
exports.relativeValueRe = date.relativeValueRe;
exports.DebugWrapper = debug.DebugWrapper;
exports.debug = debug.debug;
exports.enableDebug = debug.enableDebug;
exports.warning = debug.warning;
exports.calculatePosition = dom.calculatePosition;
exports.getContainer = dom.getContainer;
exports.getStyleNumber = dom.getStyleNumber;
exports.ownerDocument = dom.ownerDocument;
exports.ServerError = errors.ServerError;
exports.escapeHtml = escapeHtml.escapeHtml;
exports.getExprProperties = filterSchema.getExprProperties;
exports.formatDuration = formatDuration.formatDuration;
exports.FormulaExec = formula.FormulaExec;
exports.formulaExec = formula.formulaExec;
exports.isExpression = formula.isExpression;
exports.isNeedFormula = formula.isNeedFormula;
exports.isNowFormula = formula.isNowFormula;
exports.registerFormulaExec = formula.registerFormulaExec;
exports.replaceExpression = formula.replaceExpression;
exports.getVariable = getVariable.getVariable;
exports.collectVariables = grammar.collectVariables;
exports.handleAction = handleAction.handleAction;
exports.JSONTraverse = helper.JSONTraverse;
exports.SkipOperation = helper.SkipOperation;
exports.__uri = helper.__uri;
exports.anyChanged = helper.anyChanged;
exports.autobind = helper.autobind;
exports.bulkBindFunctions = helper.bulkBindFunctions;
exports.camel = helper.camel;
exports.chainEvents = helper.chainEvents;
exports.chainFunctions = helper.chainFunctions;
exports.convertArrayValueToMoment = helper.convertArrayValueToMoment;
exports.detectPropValueChanged = helper.detectPropValueChanged;
exports.difference = helper.difference;
exports.eachTree = helper.eachTree;
exports.everyTree = helper.everyTree;
exports.filterTree = helper.filterTree;
exports.findIndex = helper.findIndex;
exports.findObjectsWithKey = helper.findObjectsWithKey;
exports.findTree = helper.findTree;
exports.findTreeIndex = helper.findTreeIndex;
exports.flattenTree = helper.flattenTree;
exports.flattenTreeWithLeafNodes = helper.flattenTreeWithLeafNodes;
exports.getLevelFromClassName = helper.getLevelFromClassName;
exports.getPropValue = helper.getPropValue;
exports.getRange = helper.getRange;
exports.getScrollParent = helper.getScrollParent;
exports.getScrollbarWidth = helper.getScrollbarWidth;
exports.getTree = helper.getTree;
exports.getTreeAncestors = helper.getTreeAncestors;
exports.getTreeDepth = helper.getTreeDepth;
exports.getTreeParent = helper.getTreeParent;
exports.getWidthRate = helper.getWidthRate;
exports.guid = helper.guid;
exports.hasAbility = helper.hasAbility;
exports.hasFile = helper.hasFile;
exports.hasOwnPropertyInPath = helper.hasOwnPropertyInPath;
exports.hasVisibleExpression = helper.hasVisibleExpression;
exports.hashCode = helper.hashCode;
exports.immutableExtends = helper.immutableExtends;
exports.injectPropsToObject = helper.injectPropsToObject;
exports.isArrayChildrenModified = helper.isArrayChildrenModified;
exports.isBreakpoint = helper.isBreakpoint;
exports.isClickOnInput = helper.isClickOnInput;
exports.isDisabled = helper.isDisabled;
exports.isEmpty = helper.isEmpty;
exports.isMobile = helper.isMobile;
exports.isNumeric = helper.isNumeric;
exports.isObjectShallowModified = helper.isObjectShallowModified;
exports.isSuperDataModified = helper.isSuperDataModified;
exports.isUnfolded = helper.isUnfolded;
exports.isVisible = helper.isVisible;
exports.lcFirst = helper.lcFirst;
exports.loadScript = helper.loadScript;
exports.makeColumnClassBuild = helper.makeColumnClassBuild;
exports.makeHorizontalDeeper = helper.makeHorizontalDeeper;
exports.mapObject = helper.mapObject;
exports.mapTree = helper.mapTree;
exports.noop = helper.noop;
exports.normalizeNodePath = helper.normalizeNodePath;
exports.object2formData = helper.object2formData;
exports.omitControls = helper.omitControls;
exports.padArr = helper.padArr;
exports.parseQuery = helper.parseQuery;
exports.pickEventsProps = helper.pickEventsProps;
exports.preventDefault = helper.preventDefault;
exports.promisify = helper.promisify;
exports.qsparse = helper.qsparse;
exports.qsstringify = helper.qsstringify;
exports.range = helper.range;
exports.removeHTMLTag = helper.removeHTMLTag;
exports.repeatCount = helper.repeatCount;
exports.rmUndefined = helper.rmUndefined;
exports.someTree = helper.someTree;
exports.sortArray = helper.sortArray;
exports.spliceTree = helper.spliceTree;
exports.syncDataFromSuper = helper.syncDataFromSuper;
exports.ucFirst = helper.ucFirst;
exports.until = helper.until;
exports.utils = helper;
exports.uuid = helper.uuid;
exports.uuidv4 = helper.uuidv4;
exports.visibilityFilter = helper.visibilityFilter;
exports.highlight = highlight.highlight;
exports.generateIcon = icon.generateIcon;
exports.getImageDimensions = image.getImageDimensions;
exports.toDataURL = image.toDataURL;
exports.isPureVariable = isPureVariable.isPureVariable;
exports.JSONSchme2AMisSchema = jsonSchema2AmisSchema.JSONSchme2AMisSchema;
exports.keyToPath = keyToPath.keyToPath;
exports.makeSorter = makeSorter.makeSorter;
exports.normalizeLink = normalizeLink.normalizeLink;
exports.normalizeOptions = normalizeOptions.normalizeOptions;
exports.cloneObject = object.cloneObject;
exports.createObject = object.createObject;
exports.deleteVariable = object.deleteVariable;
exports.extendObject = object.extendObject;
exports.isObject = object.isObject;
exports.pickValues = object.pickValues;
exports.setVariable = object.setVariable;
exports.offset = offset.offset;
exports.offsetParent = offsetParent.offsetParent;
exports.matchOptionValue = optionValueCompare.matchOptionValue;
exports.optionValueCompare = optionValueCompare.optionValueCompare;
exports.position = position.position;
exports.prettyBytes = prettyBytes.prettyBytes;
exports.bindEvent = rendererEvent.bindEvent;
exports.createRendererEvent = rendererEvent.createRendererEvent;
exports.dispatchEvent = rendererEvent.dispatchEvent;
exports.getRendererEventListeners = rendererEvent.getRendererEventListeners;
exports.resolveEventData = rendererEvent.resolveEventData;
exports.replaceText = replaceText.replaceText;
exports.getComputedStyle = resizeSensor.getComputedStyle;
exports.resizeSensor = resizeSensor.resizeSensor;
exports.resolveVariable = resolveVariable.resolveVariable;
exports.resolveVariableAndFilter = resolveVariableAndFilter.resolveVariableAndFilter;
exports.resolveVariableAndFilterForAsync = resolveVariableAndFilterForAsync.resolveVariableAndFilterForAsync;
exports.RootClose = RootClose.RootClose;
exports.scrollPosition = scrollPosition.scrollPosition;
exports.SimpleMap = SimpleMap.SimpleMap;
exports.string2regExp = string2regExp.string2regExp;
exports.buildStyle = style.buildStyle;
exports.tokenize = tokenize.tokenize;
exports.register = tplBuiltin.register;
exports.evalExpression = tpl.evalExpression;
exports.evalExpressionWithConditionBuilder = tpl.evalExpressionWithConditionBuilder;
exports.evalJS = tpl.evalJS;
exports.filter = tpl.filter;
exports.registerTplEnginer = tpl.registerTplEnginer;
exports.setCustomEvalExpression = tpl.setCustomEvalExpression;
exports.setCustomEvalJs = tpl.setCustomEvalJs;
exports.uncontrollable = uncontrollable.uncontrollable;
exports.addRule = validations.addRule;
exports.str2rules = validations.str2rules;
exports.validate = validations.validate;
exports.validateMessages = validations.validateMessages;
exports.validateObject = validations.validateObject;
exports.validations = validations.validations;
exports.toNumber = toNumber.toNumber;
exports.decodeEntity = decodeEntity.decodeEntity;
exports.addStyle = styleHelper.addStyle;
exports.findOrCreactStyle = styleHelper.findOrCreactStyle;
exports.formatStyle = styleHelper.formatStyle;
exports.insertCustomStyle = styleHelper.insertCustomStyle;
exports.insertStyle = styleHelper.insertStyle;
exports.valueMap = styleHelper.valueMap;
exports.getConditionComputers = resolveCondition.getConditionComputers;
exports.registerConditionComputer = resolveCondition.registerConditionComputer;
exports.resolveCondition = resolveCondition.resolveCondition;
exports.animation = Animation["default"];
exports.RegisterStore = index.RegisterStore;
exports.RendererStore = index.RendererStore;
exports.extendLocale = locale.extendLocale;
exports.getDefaultLocale = locale.getDefaultLocale;
exports.localeable = locale.localeable;
exports.makeTranslator = locale.makeTranslator;
exports.registerLocale = locale.register;
exports.removeLocaleData = locale.removeLocaleData;
exports.setDefaultLocale = locale.setDefaultLocale;
exports.Scoped = Scoped.HocScoped;
exports.ScopedContext = Scoped.ScopedContext;
exports.classnames = theme.classnames;
exports.getClassPrefix = theme.getClassPrefix;
exports.getTheme = theme.getTheme;
exports.makeClassnames = theme.makeClassnames;
exports.setDefaultTheme = theme.setDefaultTheme;
exports.theme = theme.theme;
exports.themeable = theme.themeable;
Object.defineProperty(exports, 'LoopStatus', {
    enumerable: true,
    get: function () { return Action.LoopStatus; }
});
exports.getActionByType = Action.getActionByType;
exports.registerAction = Action.registerAction;
exports.runAction = Action.runAction;
exports.runActions = Action.runActions;
exports.FormItem = Item.FormItem;
exports.FormItemWrap = Item.FormItemWrap;
exports.getFormItemByName = Item.getFormItemByName;
exports.registerFormItem = Item.registerFormItem;
exports.OptionsControl = Options.OptionsControl;
exports.registerOptionsControl = Options.registerOptionsControl;
exports.addRootWrapper = Root.addRootWrapper;
exports.EnvContext = env.EnvContext;
Object.defineProperty(exports, 'AsyncEvaluator', {
    enumerable: true,
    get: function () { return amisFormula.AsyncEvaluator; }
});
Object.defineProperty(exports, 'Evaluator', {
    enumerable: true,
    get: function () { return amisFormula.Evaluator; }
});
Object.defineProperty(exports, 'evaluate', {
    enumerable: true,
    get: function () { return amisFormula.evaluate; }
});
Object.defineProperty(exports, 'evaluateForAsync', {
    enumerable: true,
    get: function () { return amisFormula.evaluateForAsync; }
});
Object.defineProperty(exports, 'extendsFilters', {
    enumerable: true,
    get: function () { return amisFormula.extendsFilters; }
});
Object.defineProperty(exports, 'filters', {
    enumerable: true,
    get: function () { return amisFormula.filters; }
});
Object.defineProperty(exports, 'getFilters', {
    enumerable: true,
    get: function () { return amisFormula.getFilters; }
});
Object.defineProperty(exports, 'lexer', {
    enumerable: true,
    get: function () { return amisFormula.lexer; }
});
Object.defineProperty(exports, 'parse', {
    enumerable: true,
    get: function () { return amisFormula.parse; }
});
Object.defineProperty(exports, 'registerFilter', {
    enumerable: true,
    get: function () { return amisFormula.registerFilter; }
});
Object.defineProperty(exports, 'registerFunction', {
    enumerable: true,
    get: function () { return amisFormula.registerFunction; }
});
exports.LazyComponent = LazyComponent["default"];
exports.Overlay = Overlay["default"];
exports.PopOver = PopOver["default"];
exports.FormRenderer = Form.FormRenderer;
exports.stripNumber = stripNumber.stripNumber;
exports.iRendererStore = iRenderer.iRendererStore;
exports.ServiceStore = service.ServiceStore;
exports.FormStore = form.FormStore;
exports.ComboStore = combo.ComboStore;
exports.CRUDStore = crud.CRUDStore;
exports.TableStore = table.TableStore;
exports.TableStore2 = table2.TableStore2;
exports.ListStore = list.ListStore;
exports.ModalStore = modal.ModalStore;
exports.FormItemStore = formItem.FormItemStore;
exports.PaginationStore = pagination.PaginationStore;
exports.AppStore = app.AppStore;
exports.StoreNode = node.StoreNode;
exports.classPrefix = classPrefix;
exports.render = render;
exports.version = version;
