// office-viewer/esm/parse/parseSize.js
/**
 * 单位相关的解析，参考了 docxjs 里的实现
 */
var ptToPx = 1.3333;
var LengthUsage = {
    Dxa: { mul: ptToPx * 0.05, unit: 'px' },
    Emu: { mul: (ptToPx * 1) / 12700, unit: 'px' },
    FontSize: { mul: ptToPx * 0.5, unit: 'px' },
    Border: { mul: ptToPx * 0.125, unit: 'px' },
    Point: { mul: ptToPx * 1, unit: 'px' },
    Percent: { mul: 0.02, unit: '%' },
    LineHeight: { mul: 1 / 240, unit: '' },
    VmlEmu: { mul: 1 / 12700, unit: '' }
};
function convertLength(val, usage) {
    if (usage === void 0) { usage = LengthUsage.Dxa; }
    if (val == null || /.+(p[xt]|[%])$/.test(val)) {
        return val;
    }
    return "".concat((parseInt(val) * usage.mul).toFixed(2)).concat(usage.unit);
}
/**
 * 解析尺寸，返回 css 值
 *
 * @param element 节点
 * @param attrName 属性名
 * @param usage 类型
 */
function parseSize(element, attrName, usage) {
    if (usage === void 0) { usage = LengthUsage.Dxa; }
    var size = element.getAttribute(attrName);
    if (size) {
        return convertLength(String(size), usage);
    }
    return '';
}

export { LengthUsage, convertLength, parseSize };
