// office-viewer/esm/parse/parseRelationship.js
import { __values } from '../node_modules/tslib/tslib.es6.js';

/**
 * 解析 _rels/.rels 文件
 */
function parseRelationship(element, part) {
    var id = element.getAttribute('Id') || '';
    var type = element.getAttribute('Type') || '';
    var target = element.getAttribute('Target') || '';
    var targetMode = element.getAttribute('TargetMode') || '';
    return {
        id: id,
        type: type,
        target: target,
        targetMode: targetMode,
        part: part
    };
}
function parseRelationships(element, part) {
    var e_1, _a;
    var relationships = {};
    var relationshipElements = element.getElementsByTagName('Relationship');
    try {
        for (var relationshipElements_1 = __values(relationshipElements), relationshipElements_1_1 = relationshipElements_1.next(); !relationshipElements_1_1.done; relationshipElements_1_1 = relationshipElements_1.next()) {
            var relationshipElement = relationshipElements_1_1.value;
            var relationship = parseRelationship(relationshipElement, part);
            relationships[relationship.id] = relationship;
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (relationshipElements_1_1 && !relationshipElements_1_1.done && (_a = relationshipElements_1.return)) _a.call(relationshipElements_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return relationships;
}

export { parseRelationship, parseRelationships };
