// amis-ui/esm/components/table/Cell.js
/**
 * amis-ui v2.9.0
 * Copyright 2018-2023 fex
 */

import { __extends, __assign } from 'tslib';
import React__default from 'react';
import { themeable, localeable } from 'amis-core';

/**
 * @file table/BodyCell
 * @author fex
 */
var zIndex = 1;
var BodyCell = /** @class */ (function (_super) {
    __extends(BodyCell, _super);
    function BodyCell() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BodyCell.prototype.render = function () {
        var _a;
        var _b = this.props, fixed = _b.fixed, rowSpan = _b.rowSpan, colSpan = _b.colSpan, key = _b.key, children = _b.children, className = _b.className, column = _b.column, style = _b.style, groupId = _b.groupId, depth = _b.depth, Component = _b.wrapperComponent, cx = _b.classnames;
        return (React__default.createElement(Component, { key: key || null, rowSpan: rowSpan && rowSpan > 1 ? rowSpan : null, colSpan: colSpan && colSpan > 1 ? colSpan : null, className: cx('Table-cell', className, (_a = {},
                _a[cx("Table-cell-fix-".concat(fixed))] = fixed,
                _a["text-".concat(column === null || column === void 0 ? void 0 : column.align)] = column === null || column === void 0 ? void 0 : column.align,
                _a)), style: fixed ? __assign({ position: 'sticky', zIndex: zIndex }, style) : __assign({}, style), "data-group-id": groupId || null, "data-depth": depth || null }, children));
    };
    BodyCell.defaultProps = {
        fixed: '',
        wrapperComponent: 'td',
        rowSpan: null,
        colSpan: null
    };
    return BodyCell;
}(React__default.Component));
var Cell = themeable(localeable(BodyCell));

export { BodyCell, Cell as default };
