// amis-core/esm/utils/renderer-event.js
/**
 * amis-core v2.9.0
 * Copyright 2018-2023 fex
 */

import { __values, __awaiter, __generator, __assign } from 'tslib';
import { runActions } from '../actions/Action.js';
import { createObject } from './object.js';
import debounce from 'lodash/debounce';

var rendererEventListeners = [];
// 创建渲染器事件对象
function createRendererEvent(type, context) {
    var rendererEvent = {
        context: context,
        type: type,
        prevented: false,
        stoped: false,
        preventDefault: function () {
            rendererEvent.prevented = true;
        },
        stopPropagation: function () {
            rendererEvent.stoped = true;
        },
        get data() {
            return rendererEvent.context.data;
        },
        setData: function (data) {
            rendererEvent.context.data = data;
        }
    };
    return rendererEvent;
}
// 绑定事件
var bindEvent = function (renderer) {
    var e_1, _a;
    var _b, _c;
    if (!renderer) {
        return undefined;
    }
    var listeners = renderer.props.$schema.onEvent;
    if (listeners) {
        var _loop_1 = function (key) {
            var listener = rendererEventListeners.find(function (item) {
                return item.renderer === renderer && item.type === key;
            });
            if (listener === null || listener === void 0 ? void 0 : listener.executing) {
                (_c = (_b = listener === null || listener === void 0 ? void 0 : listener.debounceInstance) === null || _b === void 0 ? void 0 : _b.cancel) === null || _c === void 0 ? void 0 : _c.call(_b);
                rendererEventListeners = rendererEventListeners.filter(function (item) {
                    return !(item.renderer === listener.renderer && item.type === listener.type);
                });
                rendererEventListeners.push({
                    renderer: renderer,
                    type: key,
                    debounce: listener.debounce || null,
                    weight: listener.weight || 0,
                    actions: listener.actions
                });
            }
            if (!listener) {
                rendererEventListeners.push({
                    renderer: renderer,
                    type: key,
                    debounce: listeners[key].debounce || null,
                    weight: listeners[key].weight || 0,
                    actions: listeners[key].actions
                });
            }
        };
        try {
            // 暂存
            for (var _d = __values(Object.keys(listeners)), _e = _d.next(); !_e.done; _e = _d.next()) {
                var key = _e.value;
                _loop_1(key);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return function () {
            rendererEventListeners = rendererEventListeners.filter(function (item) { return item.renderer !== renderer; });
        };
    }
    return undefined;
};
// 触发事件
function dispatchEvent(e, renderer, scoped, data, broadcast) {
    var _a, _b, _c, _d, _e, _f;
    return __awaiter(this, void 0, void 0, function () {
        var unbindEvent, eventName, eventConfig, rendererEvent, listeners, executedCount, checkExecuted, _loop_2, listeners_1, listeners_1_1, listener, state_1, e_2_1;
        var e_2, _g;
        var _this = this;
        return __generator(this, function (_h) {
            switch (_h.label) {
                case 0:
                    unbindEvent = null;
                    eventName = typeof e === 'string' ? e : e.type;
                    (_c = (_b = (_a = renderer === null || renderer === void 0 ? void 0 : renderer.props) === null || _a === void 0 ? void 0 : _a.env) === null || _b === void 0 ? void 0 : _b.beforeDispatchEvent) === null || _c === void 0 ? void 0 : _c.call(_b, e, renderer, scoped, data, broadcast);
                    if (!broadcast) {
                        eventConfig = (_e = (_d = renderer === null || renderer === void 0 ? void 0 : renderer.props) === null || _d === void 0 ? void 0 : _d.onEvent) === null || _e === void 0 ? void 0 : _e[eventName];
                        if (!eventConfig) {
                            // 没命中也没关系
                            return [2 /*return*/, Promise.resolve()];
                        }
                        unbindEvent = bindEvent(renderer);
                    }
                    // 没有可处理的监听
                    if (!rendererEventListeners.length) {
                        return [2 /*return*/, Promise.resolve()];
                    }
                    rendererEvent = broadcast ||
                        createRendererEvent(eventName, {
                            env: (_f = renderer === null || renderer === void 0 ? void 0 : renderer.props) === null || _f === void 0 ? void 0 : _f.env,
                            nativeEvent: e,
                            data: data,
                            scoped: scoped
                        });
                    listeners = rendererEventListeners
                        .filter(function (item) {
                        return item.type === eventName &&
                            (broadcast ? true : item.renderer === renderer);
                    })
                        .sort(function (prev, next) {
                        return next.weight - prev.weight;
                    });
                    executedCount = 0;
                    checkExecuted = function () {
                        executedCount++;
                        if (executedCount === listeners.length) {
                            unbindEvent === null || unbindEvent === void 0 ? void 0 : unbindEvent();
                        }
                    };
                    _loop_2 = function (listener) {
                        var _j, _k, wait, _l, trailing, _m, leading, _o, maxWait, debounced_1;
                        return __generator(this, function (_p) {
                            switch (_p.label) {
                                case 0:
                                    _j = (listener === null || listener === void 0 ? void 0 : listener.debounce) || {}, _k = _j.wait, wait = _k === void 0 ? 100 : _k, _l = _j.trailing, trailing = _l === void 0 ? true : _l, _m = _j.leading, leading = _m === void 0 ? false : _m, _o = _j.maxWait, maxWait = _o === void 0 ? 10000 : _o;
                                    if (!(listener === null || listener === void 0 ? void 0 : listener.debounce)) return [3 /*break*/, 1];
                                    debounced_1 = debounce(function () { return __awaiter(_this, void 0, void 0, function () {
                                        return __generator(this, function (_a) {
                                            switch (_a.label) {
                                                case 0: return [4 /*yield*/, runActions(listener.actions, listener.renderer, rendererEvent)];
                                                case 1:
                                                    _a.sent();
                                                    checkExecuted();
                                                    return [2 /*return*/];
                                            }
                                        });
                                    }); }, wait, {
                                        trailing: trailing,
                                        leading: leading,
                                        maxWait: maxWait
                                    });
                                    rendererEventListeners.forEach(function (item) {
                                        // 找到事件队列中正在执行的事件加上标识，下次待执行队列就会把这个事件过滤掉
                                        if (item.renderer === listener.renderer &&
                                            listener.type === item.type) {
                                            item.executing = true;
                                            item.debounceInstance = debounced_1;
                                        }
                                    });
                                    debounced_1();
                                    return [3 /*break*/, 3];
                                case 1: return [4 /*yield*/, runActions(listener.actions, listener.renderer, rendererEvent)];
                                case 2:
                                    _p.sent();
                                    checkExecuted();
                                    _p.label = 3;
                                case 3:
                                    // 停止后续监听器执行
                                    if (rendererEvent.stoped) {
                                        return [2 /*return*/, "break"];
                                    }
                                    return [2 /*return*/];
                            }
                        });
                    };
                    _h.label = 1;
                case 1:
                    _h.trys.push([1, 6, 7, 8]);
                    listeners_1 = __values(listeners), listeners_1_1 = listeners_1.next();
                    _h.label = 2;
                case 2:
                    if (!!listeners_1_1.done) return [3 /*break*/, 5];
                    listener = listeners_1_1.value;
                    return [5 /*yield**/, _loop_2(listener)];
                case 3:
                    state_1 = _h.sent();
                    if (state_1 === "break")
                        return [3 /*break*/, 5];
                    _h.label = 4;
                case 4:
                    listeners_1_1 = listeners_1.next();
                    return [3 /*break*/, 2];
                case 5: return [3 /*break*/, 8];
                case 6:
                    e_2_1 = _h.sent();
                    e_2 = { error: e_2_1 };
                    return [3 /*break*/, 8];
                case 7:
                    try {
                        if (listeners_1_1 && !listeners_1_1.done && (_g = listeners_1.return)) _g.call(listeners_1);
                    }
                    finally { if (e_2) throw e_2.error; }
                    return [7 /*endfinally*/];
                case 8: return [2 /*return*/, Promise.resolve(rendererEvent)];
            }
        });
    });
}
var getRendererEventListeners = function () {
    return rendererEventListeners;
};
/**
 * 兼容历史配置，追加对应name的值
 * @param props
 * @param data
 * @param valueKey
 */
var resolveEventData = function (props, data, valueKey) {
    var _a, _b;
    return createObject(props.data, props.name && valueKey
        ? __assign(__assign({}, data), (_a = {}, _a[props.name] = data[valueKey], _a.__rendererData = __assign(__assign({}, props.data), (_b = {}, _b[props.name] = data[valueKey], _b)), _a)) : data);
};

export { bindEvent, createRendererEvent, dispatchEvent, getRendererEventListeners, resolveEventData };
