// amis-core/lib/utils/style-helper.js
/**
 * amis-core v2.9.0
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var helper = require('./helper.js');

var valueMap = {
    'marginTop': 'margin-top',
    'marginRight': 'margin-right',
    'marginBottom': 'margin-bottom',
    'marginLeft': 'margin-left',
    'paddingTop': 'padding-top',
    'paddingRight': 'padding-right',
    'paddingBottom': 'padding-bottom',
    'paddingLeft': 'padding-left',
    'top-border-width': 'border-top-width',
    'right-border-width': 'border-right-width',
    'bottom-border-width': 'border-bottom-width',
    'left-border-width': 'border-left-width',
    'top-border-style': 'border-top-style',
    'right-border-style': 'border-right-style',
    'bottom-border-style': 'border-bottom-style',
    'left-border-style': 'border-left-style',
    'top-border-color': 'border-top-color',
    'right-border-color': 'border-right-color',
    'bottom-border-color': 'border-bottom-color',
    'left-border-color': 'border-left-color',
    'fontSize': 'font-size',
    'fontWeight': 'font-weight',
    'lineHeight': 'line-height'
};
function findOrCreactStyle(id) {
    var varStyleTag = document.getElementById(id);
    if (!varStyleTag) {
        varStyleTag = document.createElement('style');
        varStyleTag.id = id;
        document.body.appendChild(varStyleTag);
    }
    return varStyleTag;
}
function insertStyle(style, id) {
    var varStyleTag = findOrCreactStyle(id);
    // bca-disable-line
    varStyleTag.innerHTML = style;
}
function addStyle(style, id) {
    var varStyleTag = findOrCreactStyle(id);
    // bca-disable-line
    varStyleTag.innerHTML += style;
}
function formatStyle(css, classNames, id) {
    var e_1, _a;
    var _b;
    if (!css) {
        return { value: '', origin: [] };
    }
    var res = [];
    var status2string = {
        default: '',
        hover: ':hover',
        active: ':hover:active',
        disabled: '.is-disabled'
    };
    var _loop_1 = function (item) {
        var e_2, _c;
        var body = css[item.key];
        var list = (_b = item.value) === null || _b === void 0 ? void 0 : _b.split(' ');
        var classNameList = [];
        if (!body) {
            return "continue";
        }
        list === null || list === void 0 ? void 0 : list.forEach(function (n) {
            if (/(\S*[C|c]lassName-\S*)/.test(n) &&
                !!~n.indexOf((id === null || id === void 0 ? void 0 : id.replace('u:', '').replace('-label', '').replace('-description', '').replace('-addOn', '')) || '')) {
                classNameList.push(n);
            }
        });
        var weightsList = item.weights || {};
        try {
            for (var classNameList_1 = (e_2 = void 0, tslib.__values(classNameList)), classNameList_1_1 = classNameList_1.next(); !classNameList_1_1.done; classNameList_1_1 = classNameList_1.next()) {
                var className = classNameList_1_1.value;
                // 没有具体的样式，或者没有对应的classname
                var statusMap = {
                    default: {},
                    hover: {},
                    active: {},
                    disabled: {}
                };
                for (var key in body) {
                    if (key === '$$id') {
                        continue;
                    }
                    if (!!~key.indexOf(':default')) {
                        statusMap.default[key.replace(':default', '')] = body[key];
                    }
                    else if (!!~key.indexOf(':hover')) {
                        statusMap.hover[key.replace(':hover', '')] = body[key];
                    }
                    else if (!!~key.indexOf(':active')) {
                        statusMap.active[key.replace(':active', '')] = body[key];
                    }
                    else if (!!~key.indexOf(':disabled')) {
                        statusMap.disabled[key.replace(':disabled', '')] = body[key];
                    }
                    else {
                        statusMap.default[key] = body[key];
                    }
                }
                var _loop_2 = function (status_1) {
                    var weights = weightsList[status_1];
                    var styles = [];
                    var fn = function (key, value) {
                        key = valueMap[key] || key;
                        styles.push("".concat(key, ": ").concat(value, ";"));
                    };
                    for (var key in statusMap[status_1]) {
                        if (key === '$$id') {
                            continue;
                        }
                        var style = statusMap[status_1][key];
                        if (typeof style === 'object') {
                            // 圆角特殊处理
                            if (key === 'radius') {
                                fn('border-radius', [
                                    style['top-left-border-radius'],
                                    style['top-right-border-radius'],
                                    style['bottom-right-border-radius'],
                                    style['bottom-left-border-radius']
                                ].join(' '));
                            }
                            else {
                                for (var k in style) {
                                    if (k === '$$id') {
                                        continue;
                                    }
                                    var value = style[k];
                                    value && fn(k, value);
                                }
                            }
                        }
                        else {
                            var value = style;
                            value && fn(key, value);
                        }
                    }
                    if (styles.length > 0) {
                        var cx = ((weights === null || weights === void 0 ? void 0 : weights.pre) || '') + className + ((weights === null || weights === void 0 ? void 0 : weights.suf) || '');
                        res.push({
                            className: cx + status2string[status_1],
                            content: ".".concat(cx + status2string[status_1], " {\n  ").concat(styles.join('\n  '), "\n}")
                        });
                        if (['hover', 'active', 'disabled'].includes(status_1)) {
                            res.push({
                                className: cx + '.' + status_1,
                                content: ".".concat(cx, ".").concat(status_1, " {\n  ").concat(styles.join('\n  '), "\n}")
                            });
                        }
                    }
                };
                for (var status_1 in statusMap) {
                    _loop_2(status_1);
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (classNameList_1_1 && !classNameList_1_1.done && (_c = classNameList_1.return)) _c.call(classNameList_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    try {
        for (var classNames_1 = tslib.__values(classNames), classNames_1_1 = classNames_1.next(); !classNames_1_1.done; classNames_1_1 = classNames_1.next()) {
            var item = classNames_1_1.value;
            _loop_1(item);
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (classNames_1_1 && !classNames_1_1.done && (_a = classNames_1.return)) _a.call(classNames_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return {
        value: res.map(function (n) { return n.content; }).join('\n'),
        origin: res
    };
}
function insertCustomStyle(css, classNames, id) {
    if (!css) {
        return;
    }
    var value = formatStyle(css, classNames, id).value;
    insertStyle(value, (id === null || id === void 0 ? void 0 : id.replace('u:', '')) || helper.uuid());
}

exports.addStyle = addStyle;
exports.findOrCreactStyle = findOrCreactStyle;
exports.formatStyle = formatStyle;
exports.insertCustomStyle = insertCustomStyle;
exports.insertStyle = insertStyle;
exports.valueMap = valueMap;
